From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 13 Jan 2021 03:54:01 +0900
Subject: [PATCH] POM Changes


diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000000000000000000000000000000000000..93d79a837f61bb4b17c531702c0ead10726b29dc
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,3 @@
++target/
++*.iml
++build/
diff --git a/src/main/java/net/minecraft/client/ClientBrandRetriever.java b/src/main/java/net/minecraft/client/ClientBrandRetriever.java
index fb84bbcd532981037e97b33e7b8bdef539dbbd2c..871ca99691b10950eeeb6d4b7630708f34f3bef7 100644
--- a/src/main/java/net/minecraft/client/ClientBrandRetriever.java
+++ b/src/main/java/net/minecraft/client/ClientBrandRetriever.java
@@ -3,7 +3,7 @@ package net.minecraft.client;
 import net.minecraft.obfuscate.DontObfuscate;
 
 public class ClientBrandRetriever {
-    public static final String VANILLA_NAME = "magmacube"; // MagmaCube > vanilla!
+    public static final String VANILLA_NAME = "blueberry"; // Blueberry > // MagmaCube > vanilla!;
 
     @DontObfuscate
     public static String getClientModName() {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 6907041caf3253dec59edac9b2dfba26127f9e33..515b94a262495060e46b8a7515d5d94ee380d2b1 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -925,7 +925,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return "magmacube"; // MagmaCube > vanilla!
+        return "blueberry"; // Blueberry > // MagmaCube > vanilla!
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
