From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:21 +0900
Subject: [PATCH] Get our version instead of magmacube's


diff --git a/src/main/java/net/minecraft/DetectedVersion.java b/src/main/java/net/minecraft/DetectedVersion.java
index fac7c2e012d25dd3c29e01dc08def7934af0c19b..9d9151ca97bc4a4003e8e55375e8854a3e768e0f 100644
--- a/src/main/java/net/minecraft/DetectedVersion.java
+++ b/src/main/java/net/minecraft/DetectedVersion.java
@@ -26,7 +26,7 @@ public class DetectedVersion implements WorldVersion {
     private final int dataPackVersion;
     private final Date buildTime;
     private final String releaseTarget;
-    @javax.annotation.Nullable public final String magmaCubeVersion; // MagmaCube
+    @javax.annotation.Nullable public final String magmaCubeVersion = net.blueberrymc.common.util.Versioning.getVersion().getMagmaCubeCommit(); // MagmaCube // Blueberry
 
     private DetectedVersion() {
         this.id = UUID.randomUUID().toString().replaceAll("-", "");
@@ -38,7 +38,7 @@ public class DetectedVersion implements WorldVersion {
         this.dataPackVersion = 8;
         this.buildTime = new Date();
         this.releaseTarget = "1.18";
-        this.magmaCubeVersion = "debug"; // MagmaCube
+        // this.magmaCubeVersion = "debug"; // MagmaCube // Blueberry
     }
 
     private DetectedVersion(JsonObject json) {
@@ -52,7 +52,7 @@ public class DetectedVersion implements WorldVersion {
         this.resourcePackVersion = GsonHelper.getAsInt(jsonObject2, "resource");
         this.dataPackVersion = GsonHelper.getAsInt(jsonObject2, "data");
         this.buildTime = Date.from(ZonedDateTime.parse(GsonHelper.getAsString(json, "build_time")).toInstant());
-        this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug"; // MagmaCube
+        // this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug"; // MagmaCube // Blueberry
     }
 
     public static WorldVersion tryDetectVersion() {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 2d44a98b4c6b3ffa6f64b406bf9fbe07ccb64b5c..7d82002a71c0a9c081d271b0753402d188918417 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -717,7 +717,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
     }
 
     public String getLaunchedVersion() {
-        return this.launchedVersion;
+        return net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion(); // Blueberry
     }
 
     public String getVersionType() {
