From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <ilyy@outlook.jp>
Date: Thu, 14 Jan 2021 01:44:21 +0900
Subject: [PATCH] Get our version instead of magmacube's


diff --git a/src/main/java/net/minecraft/DetectedVersion.java b/src/main/java/net/minecraft/DetectedVersion.java
index 5d021ed19b4d3fc25edaf12356692f5453b08eef..2cac7a874fa4a86bf35c971fe5e5b63dea2e1451 100644
--- a/src/main/java/net/minecraft/DetectedVersion.java
+++ b/src/main/java/net/minecraft/DetectedVersion.java
@@ -26,7 +26,7 @@ public class DetectedVersion implements GameVersion {
    private final int dataPackVersion;
    private final Date buildTime;
    private final String releaseTarget;
-   @javax.annotation.Nullable public final String magmaCubeVersion; // MagmaCube
+   @javax.annotation.Nullable public final String magmaCubeVersion = net.blueberrymc.common.util.Versioning.getVersion().getMagmaCubeCommit(); // MagmaCube // Blueberry
 
    private DetectedVersion() {
       this.id = UUID.randomUUID().toString().replaceAll("-", "");
@@ -38,7 +38,7 @@ public class DetectedVersion implements GameVersion {
       this.dataPackVersion = 7;
       this.buildTime = new Date();
       this.releaseTarget = "1.17";
-      this.magmaCubeVersion = "debug"; // MagmaCube
+      // this.magmaCubeVersion = "debug"; // MagmaCube // Blueberry
    }
 
    private DetectedVersion(JsonObject jsonObject) {
@@ -52,7 +52,7 @@ public class DetectedVersion implements GameVersion {
       this.resourcePackVersion = GsonHelper.getAsInt(jsonObject2, "resource");
       this.dataPackVersion = GsonHelper.getAsInt(jsonObject2, "data");
       this.buildTime = Date.from(ZonedDateTime.parse(GsonHelper.getAsString(jsonObject, "build_time")).toInstant());
-      this.magmaCubeVersion = jsonObject.has("magmacube_version") ? GsonHelper.getAsString(jsonObject, "magmacube_version") : "debug"; // MagmaCube
+      // this.magmaCubeVersion = jsonObject.has("magmacube_version") ? GsonHelper.getAsString(jsonObject, "magmacube_version") : "debug"; // MagmaCube // Blueberry
    }
 
    public static GameVersion tryDetectVersion() {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 8f9e803127d6dac54d99d91c7888ae1bcb3a1302..eb7dec3dca2490d9c69c7e2fc342f97f19b5e165 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -667,7 +667,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
    }
 
    public String getLaunchedVersion() {
-      return this.launchedVersion;
+      return net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion(); // Blueberry
    }
 
    public String getVersionType() {
