From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:21 +0900
Subject: [PATCH] Get our version instead of magmacube's


diff --git a/src/main/java/net/minecraft/DetectedVersion.java b/src/main/java/net/minecraft/DetectedVersion.java
index 7aff0d1f8b6e1335711de5a7c9c31481252c5d1a..7e16263868f665b6f9f1db4484c382a9e85a0b5f 100644
--- a/src/main/java/net/minecraft/DetectedVersion.java
+++ b/src/main/java/net/minecraft/DetectedVersion.java
@@ -26,7 +26,7 @@ public class DetectedVersion implements GameVersion {
    private final int dataPackVersion;
    private final Date buildTime;
    private final String releaseTarget;
-   @javax.annotation.Nullable public final String magmaCubeVersion; // MagmaCube
+   @javax.annotation.Nullable public final String magmaCubeVersion = net.blueberrymc.common.util.Versioning.getVersion().getMagmaCubeCommit(); // MagmaCube // Blueberry
 
    private DetectedVersion() {
       this.id = UUID.randomUUID().toString().replaceAll("-", "");
@@ -38,7 +38,7 @@ public class DetectedVersion implements GameVersion {
       this.dataPackVersion = 7;
       this.buildTime = new Date();
       this.releaseTarget = "1.17";
-      this.magmaCubeVersion = "debug"; // MagmaCube
+      // this.magmaCubeVersion = "debug"; // MagmaCube // Blueberry
    }
 
    private DetectedVersion(JsonObject json) {
@@ -52,7 +52,7 @@ public class DetectedVersion implements GameVersion {
       this.resourcePackVersion = GsonHelper.getAsInt(jsonObject2, "resource");
       this.dataPackVersion = GsonHelper.getAsInt(jsonObject2, "data");
       this.buildTime = Date.from(ZonedDateTime.parse(GsonHelper.getAsString(json, "build_time")).toInstant());
-      this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug";
+      // this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug"; // Blueberry
    }
 
    public static GameVersion tryDetectVersion() {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 997a39b0768e92858fd19a938e2cc670bd2a0f68..57aa259dc751e86f036b572f5eb83297623f9a1b 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -691,7 +691,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
    }
 
    public String getLaunchedVersion() {
-      return this.launchedVersion;
+      return net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion(); // Blueberry
    }
 
    public String getVersionType() {
