From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:21 +0900
Subject: [PATCH] Get our version instead of magmacube's


diff --git a/src/main/java/net/minecraft/DetectedVersion.java b/src/main/java/net/minecraft/DetectedVersion.java
index be83adb55afb3c52fe2f9ec505dfd6fb88907d6a..6ebde913188ef2fc5f4460e6c025f9ec5c55a665 100644
--- a/src/main/java/net/minecraft/DetectedVersion.java
+++ b/src/main/java/net/minecraft/DetectedVersion.java
@@ -26,7 +26,7 @@ public class DetectedVersion implements WorldVersion {
     private final int dataPackVersion;
     private final Date buildTime;
     private final String releaseTarget;
-    @javax.annotation.Nullable public final String magmaCubeVersion; // MagmaCube
+    @javax.annotation.Nullable public final String magmaCubeVersion = net.blueberrymc.common.util.Versioning.getVersion().getMagmaCubeCommit(); // MagmaCube // Blueberry
 
     private DetectedVersion() {
         this.id = UUID.randomUUID().toString().replaceAll("-", "");
@@ -38,7 +38,7 @@ public class DetectedVersion implements WorldVersion {
         this.dataPackVersion = 8;
         this.buildTime = new Date();
         this.releaseTarget = "1.18";
-        this.magmaCubeVersion = "debug"; // MagmaCube
+        // this.magmaCubeVersion = "debug"; // MagmaCube // Blueberry
     }
 
     private DetectedVersion(JsonObject json) {
@@ -52,7 +52,7 @@ public class DetectedVersion implements WorldVersion {
         this.resourcePackVersion = GsonHelper.getAsInt(jsonObject2, "resource");
         this.dataPackVersion = GsonHelper.getAsInt(jsonObject2, "data");
         this.buildTime = Date.from(ZonedDateTime.parse(GsonHelper.getAsString(json, "build_time")).toInstant());
-        this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug"; // MagmaCube
+        // this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug"; // MagmaCube // Blueberry
     }
 
     public static WorldVersion tryDetectVersion() {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 7ef68b3761ca17361b6280f68ee50d6b05f39836..2fbe4f3c8ba1a139715e7d7998cae03b1908fa41 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -716,7 +716,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
     }
 
     public String getLaunchedVersion() {
-        return this.launchedVersion;
+        return net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion(); // Blueberry
     }
 
     public String getVersionType() {
