From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:21 +0900
Subject: [PATCH] Get our version instead of magmacube's


diff --git a/src/main/java/net/minecraft/DetectedVersion.java b/src/main/java/net/minecraft/DetectedVersion.java
index 0439b7b1a1cf9d35b7a5f2e9cd3b0460c24233de..d248fce9c3c793b7fb04904b70be356a6526854e 100644
--- a/src/main/java/net/minecraft/DetectedVersion.java
+++ b/src/main/java/net/minecraft/DetectedVersion.java
@@ -25,7 +25,7 @@ public class DetectedVersion implements WorldVersion {
     private final int resourcePackVersion;
     private final int dataPackVersion;
     private final Date buildTime;
-    @javax.annotation.Nullable public final String magmaCubeVersion; // MagmaCube
+    @javax.annotation.Nullable public final String magmaCubeVersion = net.blueberrymc.common.util.Versioning.getVersion().getMagmaCubeCommit(); // MagmaCube // Blueberry
 
     private DetectedVersion() {
         this.id = UUID.randomUUID().toString().replaceAll("-", "");
@@ -36,7 +36,7 @@ public class DetectedVersion implements WorldVersion {
         this.resourcePackVersion = 13;
         this.dataPackVersion = 12;
         this.buildTime = new Date();
-        this.magmaCubeVersion = "debug"; // MagmaCube
+        // this.magmaCubeVersion = "debug"; // MagmaCube // Blueberry
     }
 
     private DetectedVersion(JsonObject json) {
@@ -49,7 +49,7 @@ public class DetectedVersion implements WorldVersion {
         this.resourcePackVersion = GsonHelper.getAsInt(jsonObject2, "resource");
         this.dataPackVersion = GsonHelper.getAsInt(jsonObject2, "data");
         this.buildTime = Date.from(ZonedDateTime.parse(GsonHelper.getAsString(json, "build_time")).toInstant());
-        this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug"; // MagmaCube
+        // this.magmaCubeVersion = json.has("magmacube_version") ? GsonHelper.getAsString(json, "magmacube_version") : "debug"; // MagmaCube // Blueberry
     }
 
     public static WorldVersion tryDetectVersion() {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 6c71ca2b915b7c779510912eacfab691fce395f7..f9d4c4d75f4d539f750b8de0c5af16dddb15f3cd 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -766,7 +766,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
     }
 
     public String getLaunchedVersion() {
-        return this.launchedVersion;
+        return net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion(); // Blueberry
     }
 
     public String getVersionType() {
