From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:46 +0900
Subject: [PATCH] Mods support


diff --git a/src/main/java/net/minecraft/CrashReport.java b/src/main/java/net/minecraft/CrashReport.java
index 5203f9f6202a4824a4f96607a3466a5d804d4df7..8cd04bbd0777acfd4cbb6a4621ee98db9d88f97d 100644
--- a/src/main/java/net/minecraft/CrashReport.java
+++ b/src/main/java/net/minecraft/CrashReport.java
@@ -228,6 +228,6 @@ public class CrashReport {
 
     public static void preload() {
         MemoryReserve.allocate();
-        (new CrashReport("Don't panic!", new Throwable())).getFriendlyReport();
+        LOGGER.info((new CrashReport("Don't panic!", new Throwable())).getFriendlyReport()); // Blueberry
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/SystemReport.java b/src/main/java/net/minecraft/SystemReport.java
index 74e8f78d68f387fc5d99e42999b7a63a8103f77c..ab536263910ad9aa825fbf2e015bf47582f8f2c0 100644
--- a/src/main/java/net/minecraft/SystemReport.java
+++ b/src/main/java/net/minecraft/SystemReport.java
@@ -48,6 +48,25 @@ public class SystemReport {
             List<String> list = Util.getVmArguments().collect(Collectors.toList());
             return String.format("%d total; %s", list.size(), String.join(" ", list));
         }));
+        // Blueberry start
+        // TODO: we need better layout
+        this.setDetail("Blueberry Version", net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion());
+        this.setDetail("Mods", () -> {
+            StringBuilder sb = new StringBuilder("\n");
+            sb.append("      Status:\n");
+            sb.append("        L = Loaded\n");
+            sb.append("        P = Pre Init\n");
+            sb.append("        I = Init\n");
+            sb.append("        J = Post Init\n");
+            sb.append("        A = Available\n");
+            sb.append("        E = Errored\n");
+            sb.append("        U = Unloaded\n");
+            for (net.blueberrymc.common.bml.BlueberryMod mod : net.blueberrymc.common.Blueberry.getModLoader().getLoadedMods()) {
+                sb.append("      ").append(mod.getName()).append(" (").append(mod.getDescription().getModId()).append(") [").append(mod.getDescription().getVersion()).append("] - ").append(mod.getStateList().toString()).append("\n");
+            }
+            return sb.toString();
+        });
+        // Blueberry end
     }
 
     public void setDetail(String s, String s2) {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index f8eba59cc141308754dca739687348775756b3d6..8baad70eeea67b51f8ab96bd3b68468d2b8c7837 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -458,6 +458,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
         this.mainRenderTarget.setClearColor(0.0F, 0.0F, 0.0F, 0.0F);
         this.mainRenderTarget.clear(ON_OSX);
         this.resourceManager = new SimpleReloadableResourceManager(PackType.CLIENT_RESOURCES);
+        net.blueberrymc.common.Blueberry.getModLoader().callPreInit(); // Blueberry
         this.resourcePackRepository.reload();
         this.options.loadSelectedResourcePacks(this.resourcePackRepository);
         this.languageManager = new LanguageManager(this.options.languageCode);
@@ -535,9 +536,15 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
         this.window.updateRawMouseInput(this.options.rawMouseInput);
         this.window.setDefaultErrorCallback();
         this.resizeDisplay();
+        if (s != null) {
+            this.setScreen(new net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen(() -> ConnectScreen.startConnecting(new TitleScreen(), this, new ServerAddress(s, i), (ServerData)null))); // Blueberry
+        } else {
+            this.setScreen(new net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen(new TitleScreen())); // Blueberry
+        }
         this.gameRenderer.preloadUiShader(this.getClientPackSource().getVanillaPack());
         LoadingOverlay.registerTextures(this);
         List<PackResources> list = this.resourcePackRepository.openAllSelected();
+        net.blueberrymc.common.Blueberry.getModLoader().callInit(); // Blueberry
         this.reloadStateTracker.startReload(ResourceLoadStateTracker.ReloadReason.INITIAL, list);
         this.setOverlay(new LoadingOverlay(this, this.resourceManager.createReload(Util.backgroundExecutor(), this, RESOURCE_RELOAD_INITIAL_TASK, list), (optional) -> Util.ifElse(optional, this::rollbackResourcePacks, () -> {
                 if (SharedConstants.IS_RUNNING_IN_IDE) {
@@ -734,9 +741,10 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
     }
 
     public static void crash(CrashReport crashReport) {
-        File crashReportsDir = new File(getInstance().gameDirectory, "crash-reports");
+        File crashReportsDir = new File(instance == null ? new File(".") : getInstance().gameDirectory, "crash-reports"); // Blueberry
         File crashReportFile = new File(crashReportsDir, "crash-" + (new SimpleDateFormat("yyyy-MM-dd_HH.mm.ss")).format(new Date()) + "-client.txt");
         Bootstrap.realStdoutPrintln(crashReport.getFriendlyReport());
+        net.blueberrymc.common.util.DiscordRPCTaskExecutor.shutdownNow(); // Blueberry
         if (crashReport.getSaveFile() != null) {
             Bootstrap.realStdoutPrintln("#@!@# Game crashed! Crash report saved to: #@!@# " + crashReport.getSaveFile());
             System.exit(-1);
@@ -875,6 +883,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 
         this.screen = screen;
         BufferUploader.reset();
+        net.blueberrymc.client.event.ClientEventFactory.callScreenChangedEvent(this.screen); // Blueberry
         if (screen != null) {
             this.mouseHandler.releaseMouse();
             KeyMapping.releaseAll();
@@ -890,6 +899,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 
     public void setOverlay(@Nullable Overlay overlay) {
         this.overlay = overlay;
+        net.blueberrymc.client.event.ClientEventFactory.callOverlayChangedEvent(overlay); // Blueberry
     }
 
     public void destroy() {
@@ -914,6 +924,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
                 this.screen.removed();
             }
 
+            net.blueberrymc.common.Blueberry.shutdown();
             this.close();
         } finally {
             Util.timeSource = System::nanoTime;
@@ -1646,6 +1657,10 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
             this.pendingConnection.tick();
         }
 
+        // Blueberry start
+        this.profiler.popPush("blueberryClientScheduler");
+        net.blueberrymc.common.Blueberry.getUtil().getClientScheduler().tick();
+        // Blueberry end
         this.profiler.popPush("keyboard");
         this.keyboardHandler.tick();
         this.profiler.pop();
diff --git a/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java b/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
index 47c955961f3166ce741773660c509b4154063c9e..83877228d193f30ed98b20761623b058fab0f316 100644
--- a/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
+++ b/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
@@ -28,6 +28,7 @@ import java.util.concurrent.CompletableFuture;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+import net.blueberrymc.common.Blueberry; // Blueberry
 import net.minecraft.ChatFormatting;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
@@ -337,6 +338,12 @@ public class DebugScreenOverlay extends GuiComponent {
         long l3 = Runtime.getRuntime().freeMemory();
         long l4 = l2 - l3;
         List<String> list = Lists.newArrayList(new String[]{String.format("Java: %s %dbit", System.getProperty("java.version"), this.minecraft.is64Bit() ? 64 : 32), String.format("Mem: % 2d%% %03d/%03dMB", l4 * 100L / l, bytesToMegabytes(l4), bytesToMegabytes(l)), String.format("Allocated: % 2d%% %03dMB", l2 * 100L / l, bytesToMegabytes(l2)), "", String.format("CPU: %s", GlUtil.getCpuInfo()), "", String.format("Display: %dx%d (%s)", Minecraft.getInstance().getWindow().getWidth(), Minecraft.getInstance().getWindow().getHeight(), GlUtil.getVendor()), GlUtil.getRenderer(), GlUtil.getOpenGLVersion()});
+        // Blueberry start
+        list.add("");
+        list.add("MagmaCube " + Blueberry.getVersion().getMagmaCubeCommit());
+        list.add("Blueberry " + Blueberry.getVersion().getFullyQualifiedVersion());
+        list.add(Blueberry.getModLoader().getLoadedMods().size() + " mods loaded, " + Blueberry.getModLoader().getActiveMods().size() + " mods active");
+        // Blueberry end
         if (this.minecraft.showOnlyReducedInfo()) {
             return list;
         } else {
@@ -344,7 +351,7 @@ public class DebugScreenOverlay extends GuiComponent {
                 BlockPos blockPos = ((BlockHitResult)this.block).getBlockPos();
                 BlockState blockState = this.minecraft.level.getBlockState(blockPos);
                 list.add("");
-                list.add(ChatFormatting.UNDERLINE + "Targeted Block: " + blockPos.getX() + ", " + blockPos.getY() + ", " + blockPos.getZ());
+                list.add("" + ChatFormatting.UNDERLINE + "Targeted Block: " + blockPos.getX() + ", " + blockPos.getY() + ", " + blockPos.getZ());
                 list.add(String.valueOf(Registry.BLOCK.getKey(blockState.getBlock())));
                 UnmodifiableIterator var12 = blockState.getValues().entrySet().iterator();
 
diff --git a/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java b/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
index 091b30e0d29b85dd79b473d3419d741238a1f0a2..6467fcb548bda5a1a060ef8b0bf23460c4376ae1 100644
--- a/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -119,6 +119,7 @@ public class LoadingOverlay extends Overlay {
         }
 
         if (f2 >= 2.0F) {
+            net.blueberrymc.common.Blueberry.getModLoader().callPostInit(); // Blueberry
             this.minecraft.setOverlay((Overlay)null);
         }
 
@@ -132,6 +133,7 @@ public class LoadingOverlay extends Overlay {
 
             this.fadeOutStart = Util.getMillis();
             if (this.minecraft.screen != null) {
+                if (this.minecraft.screen instanceof net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen) ((net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen) minecraft.screen).refresh(); // Blueberry
                 this.minecraft.screen.init(this.minecraft, this.minecraft.getWindow().getGuiScaledWidth(), this.minecraft.getWindow().getGuiScaledHeight());
             }
         }
diff --git a/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java b/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
index 3718f7005844bf6ae52680838ca9f5925c14eecb..7f70bd20be52d13f170a7832236e75d22ed01647 100644
--- a/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
@@ -78,6 +78,7 @@ public class PauseScreen extends Screen {
             }
 
         }));
+        this.addRenderableWidget(new Button(this.width / 2 - 102, this.height / 4 + 144 + -16, 204, 20, new net.blueberrymc.client.resources.BlueberryText("blueberry", "gui.screens.mods"), (buttonx) -> net.blueberrymc.client.gui.screens.ModListScreen.switchToModListScreen())); // Blueberry
     }
 
     public void tick() {
diff --git a/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java b/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
index 4b8d68438c9ead9ac3306a75ffa19e478882d323..20ad0c5987999dda6f08f00f4e855e3dca8381d6 100644
--- a/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
@@ -146,7 +146,8 @@ public class TitleScreen extends Screen {
             Screen screen = (Screen)(this.minecraft.options.skipMultiplayerWarning ? new JoinMultiplayerScreen(this) : new SafetyScreen(this));
             this.minecraft.setScreen(screen);
         }, onTooltip))).active = flag;
-        (this.addRenderableWidget(new Button(this.width / 2 - 100, i + i2 * 2, 200, 20, new TranslatableComponent("menu.online"), (button) -> this.realmsButtonClicked(), onTooltip))).active = flag;
+        (this.addRenderableWidget(new Button(this.width / 2 - 100, i + i2 * 2, 98, 20, new net.blueberrymc.client.resources.BlueberryText("blueberry", "gui.screens.mods"), (button) -> net.blueberrymc.client.gui.screens.ModListScreen.switchToModListScreen(), onTooltip))).active = flag;
+        (this.addRenderableWidget(new Button(this.width / 2 + 2, i + i2 * 2, 98, 20, new TranslatableComponent("menu.online"), (button) -> this.realmsButtonClicked(), onTooltip))).active = flag;
     }
 
     private void createDemoMenuOptions(int i, int i2) {
@@ -291,8 +292,10 @@ public class TitleScreen extends Screen {
                 s = s + I18n.get("menu.modded");
             }
 
-            drawString(poseStack, this.font, "MagmaCube " + ((net.minecraft.DetectedVersion) SharedConstants.getCurrentVersion()).magmaCubeVersion, 2, this.height - 20, 0xFFFFFF | i6); // MagmaCube
-            drawString(poseStack, this.font, s, 2, this.height - 10, 16777215 | i6);
+            drawString(poseStack, this.font, "Blueberry " + net.blueberrymc.common.Blueberry.getVersion().getFullyQualifiedVersion(), 2, this.height - 40, 0xFFFFFF | i6); // Blueberry
+            drawString(poseStack, this.font, "MagmaCube " + ((net.minecraft.DetectedVersion) SharedConstants.getCurrentVersion()).magmaCubeVersion, 2, this.height - 30, 0xFFFFFF | i6); // MagmaCube // Blueberry - offset 20 -> 30
+            drawString(poseStack, this.font, s, 2, this.height - 20, 16777215 | i6); // Blueberry - offset 10 -> 20
+            drawString(poseStack, this.font, net.blueberrymc.common.Blueberry.getModLoader().getLoadedMods().size() + " mods loaded, " + net.blueberrymc.common.Blueberry.getModLoader().getActiveMods().size() + " mods active", 2, this.height - 10, 0xFFFFFF | i6); // Blueberry
             drawString(poseStack, this.font, COPYRIGHT_TEXT, this.copyrightX, this.height - 10, 16777215 | i6);
             if (i > this.copyrightX && i < this.copyrightX + this.copyrightWidth && i2 > this.height - 10 && i2 < this.height) {
                 fill(poseStack, this.copyrightX, this.height - 1, this.copyrightX + this.copyrightWidth, this.height, 16777215 | i6);
diff --git a/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java b/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
index b03005ddfe33c21b92de457f2da411ccb22a3c75..6e3d0298a8d2b61fc872fa8a81ea15db26285353 100644
--- a/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
+++ b/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
@@ -295,6 +295,13 @@ public class ServerSelectionList extends ObjectSelectionList<ServerSelectionList
             RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             RenderSystem.setShaderTexture(0, GuiComponent.GUI_ICONS_LOCATION);
             GuiComponent.blit(poseStack, i3 + i4 - 15, i2, (float)(i10 * 10), (float)(176 + i11 * 8), 10, 8, 256, 256);
+            // Blueberry start
+            if (serverData.serverType != null) {
+                RenderSystem.setShader(GameRenderer::getPositionTexColorShader);
+                RenderSystem.setShaderTexture(0, net.blueberrymc.client.gui.BlueberryGuiComponents.GUI_ICONS_LOCATION);
+                GuiComponent.blit(poseStack, i3 + i4 - 18, i2 + 8, (float) 0.0F, (float) (serverData.serverType.getOffset() * 16), 16, 16, 256, 256);
+            }
+            // Blueberry end
             String s = this.serverData.getIconB64();
             if (!Objects.equals(s, this.lastIconB64)) {
                 if (this.uploadServerIcon(s)) {
@@ -315,6 +322,12 @@ public class ServerSelectionList extends ObjectSelectionList<ServerSelectionList
             int i20 = i7 - i2;
             if (i19 >= i4 - 15 && i19 <= i4 - 5 && i20 >= 0 && i20 <= 8) {
                 this.screen.setToolTip(Collections.singletonList(component2));
+                // Blueberry start
+            } else if (serverData.serverType != null && i19 >= i4 - 18 && i19 <= i4 - 5 && i20 >= 0 && i20 <= 22) {
+                String text = serverData.serverType.getBlueberryText().getContents();
+                if (text.contains("%d")) text = String.format(text, serverData.modsCount);
+                this.screen.setToolTip(Collections.singletonList(new TextComponent(text)));
+                // Blueberry end
             } else if (i19 >= i4 - i9 - 15 - 2 && i19 <= i4 - 15 - 2 && i20 >= 0 && i20 <= 8) {
                 this.screen.setToolTip(list2);
             }
diff --git a/src/main/java/net/minecraft/client/main/Main.java b/src/main/java/net/minecraft/client/main/Main.java
index 510527ed39076ab18712bf4c1c7aa3e8b17c8d2d..eb65e1dc63645be01b7183df1c326b1e138fd8bd 100644
--- a/src/main/java/net/minecraft/client/main/Main.java
+++ b/src/main/java/net/minecraft/client/main/Main.java
@@ -47,6 +47,7 @@ public class Main {
     @DontObfuscate
     public static void main(String[] args) {
         SharedConstants.tryDetectVersion();
+        if (net.minecraft.SharedConstants.IS_RUNNING_IN_IDE) org.lwjgl.glfw.GLFW.glfwInit(); // MagmaCube - lwjgl debug agent causes glGetError to crash when we didn't call glfwInit
         OptionParser optionParser = new OptionParser();
         optionParser.allowsUnrecognizedOptions();
         optionParser.accepts("demo");
@@ -85,6 +86,8 @@ public class Main {
         if (!list.isEmpty()) {
             System.out.println("Completely ignored arguments: " + list);
         }
+        File file = parseArgument(optionSet, optionSpec4); // Blueberry - moved from below
+        net.blueberrymc.common.Blueberry.bootstrap(net.blueberrymc.common.Side.CLIENT, file);
 
         String s = parseArgument(optionSet, optionSpec7);
         Proxy proxy = Proxy.NO_PROXY;
@@ -118,7 +121,7 @@ public class Main {
         PropertyMap propertyMap = GsonHelper.fromJson(gson, parseArgument(optionSet, optionSpec21), PropertyMap.class);
         PropertyMap propertyMap2 = GsonHelper.fromJson(gson, parseArgument(optionSet, optionSpec22), PropertyMap.class);
         String s5 = parseArgument(optionSet, optionSpec25);
-        File file = parseArgument(optionSet, optionSpec4);
+        // File file = parseArgument(optionSet, optionSpec4); // Blueberry - moved up
         File file2 = optionSet.has(optionSpec5) ? parseArgument(optionSet, optionSpec5) : new File(file, "assets/");
         File file3 = optionSet.has(optionSpec6) ? parseArgument(optionSet, optionSpec6) : new File(file, "resourcepacks/");
         String s6 = optionSet.has(optionSpec12) ? (String)optionSpec12.value(optionSet) : Player.createPlayerUUID((String)optionSpec11.value(optionSet)).toString();
diff --git a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
index f6b64dc5d5f1d2f2ceefebe17760c56edaa13254..e74b8e18e74ea946a593371149ff830d860d46b3 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -278,7 +278,7 @@ import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class ClientPacketListener implements ClientGamePacketListener {
+public class ClientPacketListener implements ClientGamePacketListener, net.blueberrymc.network.client.ClientBlueberryPacketListener { // Blueberry
     private static final Logger LOGGER = LogManager.getLogger();
     private static final Component GENERIC_DISCONNECT_MESSAGE = new TranslatableComponent("disconnect.lost");
     private final Connection connection;
@@ -1574,6 +1574,7 @@ public class ClientPacketListener implements ClientGamePacketListener {
 
         try {
             friendlyByteBuf = clientboundCustomPayloadPacket.getData();
+            if (net.blueberrymc.network.client.ClientBlueberryPacketListener.super.handleBlueberryCustomPayload(clientboundCustomPayloadPacket)) return; // Blueberry
             if (ClientboundCustomPayloadPacket.BRAND.equals(resourceLocation)) {
                 String s = friendlyByteBuf.readUtf();
                 this.minecraft.player.setServerBrand(s);
diff --git a/src/main/java/net/minecraft/client/multiplayer/ServerData.java b/src/main/java/net/minecraft/client/multiplayer/ServerData.java
index dc6740ab13721dca12ae0a36e10859ba662eb0b3..75a31f723df73125e7eaac2775a4c3a36697951c 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ServerData.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ServerData.java
@@ -23,6 +23,8 @@ public class ServerData {
     @Nullable
     private String iconB64;
     private boolean lan;
+    public net.blueberrymc.network.ServerType serverType = null; // Blueberry
+    public int modsCount = 0; // Blueberry
 
     public ServerData(String s, String s2, boolean flag) {
         this.name = s;
diff --git a/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java b/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
index 72bed5ea13d2439632089bf4d99772dc99372dd9..fd0cec1d2aa8f5d9eadda6052a01a28a8b32c852 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
@@ -75,6 +75,7 @@ public class ServerStatusPinger {
                         connection.disconnect(new TranslatableComponent("multiplayer.status.unrequested"));
                     } else {
                         this.receivedPing = true;
+                        serverData.serverType = net.blueberrymc.network.ServerType.VANILLA; // Blueberry
                         ServerStatus serverStatus = clientboundStatusResponsePacket.getStatus();
                         if (serverStatus.getDescription() != null) {
                             serverData.motd = serverStatus.getDescription();
@@ -145,6 +146,19 @@ public class ServerStatusPinger {
 
                 }
 
+                // Blueberry start
+                @Override
+                public void handleBlueberryHandshakeResponse(net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket packet) {
+                    boolean compatible = net.blueberrymc.common.util.ListUtils.isCompatible(packet.getModInfos(), net.blueberrymc.common.Blueberry.getModManager().getModInfos());
+                    if (compatible) {
+                        serverData.serverType = net.blueberrymc.network.ServerType.BLUEBERRY_GOOD;
+                    } else {
+                        serverData.serverType = net.blueberrymc.network.ServerType.BLUEBERRY_BAD;
+                    }
+                    serverData.modsCount = packet.getModInfos().size();
+                }
+                // Blueberry end
+
                 public Connection getConnection() {
                     return connection;
                 }
diff --git a/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java b/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
index 5d647d8d4ead1486f9eb23a31033c9493c1d877a..6d39319fc7d7cbca7aa0516ba4c15f920909cd28 100644
--- a/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
+++ b/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
@@ -102,5 +102,6 @@ public class BlockRenderDispatcher implements ResourceManagerReloadListener {
 
     public void onResourceManagerReload(ResourceManager resourceManager) {
         this.liquidBlockRenderer.setupSprites();
+        net.blueberrymc.client.world.level.fluid.FluidSpriteManager.setupSprites(); // Blueberry
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java b/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
index 4191e7bc8d5b7c0b6abefd9674c947bd47becc4c..4ffe415377fda140d0b850944ce711cafe911cf8 100644
--- a/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -72,6 +72,10 @@ public class LiquidBlockRenderer {
         TextureAtlasSprite[] textureAtlasSprites = flag ? this.lavaIcons : this.waterIcons;
         BlockState blockState = blockAndTintGetter.getBlockState(blockPos);
         int i = flag ? 16777215 : BiomeColors.getAverageWaterColor(blockAndTintGetter, blockPos);
+        // Blueberry start
+        net.blueberrymc.client.event.render.LiquidBlockRenderEvent event = net.blueberrymc.client.event.ClientEventFactory.callLiquidBlockRenderEvent(fluidState, blockPos, i);
+        i = event.getColor();
+        // Blueberry end
         float f = (float)(i >> 16 & 255) / 255.0F;
         float f2 = (float)(i >> 8 & 255) / 255.0F;
         float f3 = (float)(i & 255) / 255.0F;
diff --git a/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java b/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
index a6f52260d9b45639757fcd5bd9c7008a704a9593..52ff44f5d4636bcabec3d4c1eb6c197d29262e79 100644
--- a/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
@@ -23,6 +23,7 @@ import net.minecraft.world.entity.projectile.ProjectileUtil;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.Vec3;
+import static net.blueberrymc.common.bml.InternalBlueberryMod.*; // Blueberry
 
 public class DebugRenderer {
     public final PathfindingRenderer pathfindingRenderer = new PathfindingRenderer();
@@ -93,28 +94,25 @@ public class DebugRenderer {
 
     public void render(PoseStack poseStack, MultiBufferSource.BufferSource bufferSource, double d, double d2, double d3) {
         if (this.renderChunkborder && !Minecraft.getInstance().showOnlyReducedInfo()) {
-            this.chunkBorderRenderer.render(poseStack, bufferSource, d, d2, d3);
-         // MagmaCube start - show debug renderer with F3+G
-         if (net.minecraft.SharedConstants.IS_RUNNING_IN_IDE) {
-            this.pathfindingRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.waterDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.chunkBorderRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.heightMapRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.collisionBoxRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.neighborsUpdateRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.structureRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.lightDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.worldGenAttemptRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.solidFaceRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.chunkRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.brainDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.villageSectionsDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.beeDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.raidDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.goalSelectorRenderer.render(poseStack, bufferSource, d, d2, d3);
-            this.gameTestDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
-         }
-         // MagmaCube end - show debug renderer with F3+G
+            // Blueberry start
+            if (showDRPathfinding) this.pathfindingRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRWaterDebug) this.waterDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRChunkBorder) this.chunkBorderRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRHeightMap) this.heightMapRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRCollisionBox) this.collisionBoxRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRNeighborsUpdate) this.neighborsUpdateRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRStructure) this.structureRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRLightDebug) this.lightDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRWorldGenAttempt) this.worldGenAttemptRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRSolidFace) this.solidFaceRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRChunk) this.chunkRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRBrainDebug) this.brainDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRVillageSectionsDebug) this.villageSectionsDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRBeeDebug) this.beeDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRRaidDebug) this.raidDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRGoalSelector) this.goalSelectorRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (showDRGameTestDebug) this.gameTestDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+         // Blueberry end
         }
 
         this.gameTestDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
diff --git a/src/main/java/net/minecraft/client/resources/ClientPackSource.java b/src/main/java/net/minecraft/client/resources/ClientPackSource.java
index 142989d3972771f04aaf6d6b672dac378c867481..7a9a0f85a797932b82d198b736123a7db3e55efd 100644
--- a/src/main/java/net/minecraft/client/resources/ClientPackSource.java
+++ b/src/main/java/net/minecraft/client/resources/ClientPackSource.java
@@ -86,6 +86,7 @@ public class ClientPackSource implements RepositorySource {
         if (pack2 != null) {
             consumer.accept(pack2);
         }
+        net.blueberrymc.common.Blueberry.getModManager().loadPacks(consumer, packConstructor); // Blueberry
 
     }
 
diff --git a/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java b/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
index 43f39473c5ef3de9e53e6722bea6a0a58382526e..f5254cecd40e9d643cb59316e929560adb08c85d 100644
--- a/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
+++ b/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
@@ -2,7 +2,7 @@ package net.minecraft.network.protocol.status;
 
 import net.minecraft.network.PacketListener;
 
-public interface ClientStatusPacketListener extends PacketListener {
+public interface ClientStatusPacketListener extends PacketListener, net.blueberrymc.network.client.handshake.ClientBlueberryHandshakePacketListener { // Blueberry
     void handleStatusResponse(ClientboundStatusResponsePacket var1);
 
     void handlePongResponse(ClientboundPongResponsePacket var1);
diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 8faa9f44e8354e77b4b8b784966480a7135c1b2d..a9ef6c346f966ff136845b3ae2fc7e3993003936 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -86,6 +86,8 @@ public class Main {
                 optionParser.printHelpOn(System.err);
                 return;
             }
+            File file = new File((String)optionSet.valueOf(optionSpec10)); // Blueberry - moved from below
+            net.blueberrymc.common.Blueberry.bootstrap(net.blueberrymc.common.Side.SERVER, file); // Blueberry
 
             JvmProfiler.INSTANCE.initialize();
             CrashReport.preload();
@@ -112,7 +114,7 @@ public class Main {
                 return;
             }
 
-            File file = new File((String)optionSet.valueOf(optionSpec10));
+            // File file = new File((String)optionSet.valueOf(optionSpec10)); // Blueberry - moved up
             YggdrasilAuthenticationService yggdrasilAuthenticationService = new YggdrasilAuthenticationService(Proxy.NO_PROXY);
             MinecraftSessionService minecraftSessionService = yggdrasilAuthenticationService.createMinecraftSessionService();
             GameProfileRepository gameProfileRepository = yggdrasilAuthenticationService.createProfileRepository();
@@ -179,6 +181,7 @@ public class Main {
             WorldData worldData2 = worldData;
             final DedicatedServer dedicatedServer = MinecraftServer.spin((threadx) -> {
                 DedicatedServer dedicatedServerIn = new DedicatedServer(threadx, registryHolder, levelStorageAccess, packRepository, serverResources, worldData2, dedicatedServerSettings, DataFixers.getDataFixer(), minecraftSessionService, gameProfileRepository, gameProfileCache, LoggerChunkProgressListener::new);
+                ((net.blueberrymc.server.BlueberryServer) net.blueberrymc.common.Blueberry.getUtil()).setServer(dedicatedServerIn); // Blueberry
                 dedicatedServerIn.setSingleplayerName((String)optionSet.valueOf(optionSpec9));
                 dedicatedServerIn.setPort(optionSet.valueOf(optionSpec12));
                 dedicatedServerIn.setDemo(optionSet.has(optionSpec3));
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index bb7c295ac4aa0104941882c8b143f4f213c412e9..2a04dfad7d51c65f9a4a7768945bb1a82fa8f57f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -276,6 +276,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.proxy = proxy;
         this.packRepository = packRepository;
         this.resources = serverResources;
+        if (net.blueberrymc.common.Blueberry.isServer()) net.blueberrymc.common.Blueberry.getModLoader().callPreInit(); // Blueberry
         this.sessionService = minecraftSessionService;
         this.profileRepository = gameProfileRepository;
         this.profileCache = gameProfileCache;
@@ -292,6 +293,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.structureManager = new StructureManager(serverResources.getResourceManager(), levelStorageAccess, dataFixer);
         this.serverThread = thread;
         this.executor = Util.backgroundExecutor();
+        if (net.blueberrymc.common.Blueberry.isServer()) net.blueberrymc.common.Blueberry.getModLoader().callInit(); // Blueberry
     }
 
     private void readScoreboard(DimensionDataStorage dimensionDataStorage) {
@@ -643,6 +645,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 this.status.setDescription(new TextComponent(this.motd));
                 this.status.setVersion(new ServerStatus.Version(SharedConstants.getCurrentVersion().getName(), SharedConstants.getCurrentVersion().getProtocolVersion()));
                 this.updateStatusIcon(this.status);
+                if (net.blueberrymc.common.Blueberry.isServer()) net.blueberrymc.common.Blueberry.getModLoader().callPostInit(); // Blueberry
 
                 while(this.running) {
                     long l = Util.getMillis() - this.nextTickTime;
@@ -822,6 +825,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
 
         this.profiler.pop();
+        // Blueberry start
+        this.profiler.push("blueberryServerScheduler");
+        net.blueberrymc.common.Blueberry.getUtil().getServerScheduler().tick();
+        this.profiler.pop();
+        // Blueberry end
         this.profiler.push("tallying");
         long l2 = this.tickTimes[this.tickCount % TICK_STATS_SPAN] = Util.getNanos() - l;
         this.averageTickTime = this.averageTickTime * AVERAGE_TICK_TIME_SMOOTHING + (float)l2 / 1000000.0F * 0.19999999F;
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index f6a1df7f782e797937a47bf467698a324ff3e5ca..4943553c656953fc19bb0a70f569ff3131dd8400 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -538,6 +538,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
     }
 
     public void stopServer() {
+        net.blueberrymc.common.Blueberry.shutdown(); // Blueberry
         super.stopServer();
         Util.shutdownExecutors();
     }
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 11a917088124b7bab86d4afcca19d96263c399f2..e285ced6ca9c41992855088e470fe84c8f15335d 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -153,7 +153,7 @@ import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class ServerGamePacketListenerImpl implements ServerPlayerConnection, ServerGamePacketListener {
+public class ServerGamePacketListenerImpl implements ServerPlayerConnection, ServerGamePacketListener, net.blueberrymc.network.server.ServerBlueberryPacketListener { // Blueberry
     static final Logger LOGGER = LogManager.getLogger();
     private static final int LATENCY_CHECK_INTERVAL = 15000;
     public final Connection connection;
@@ -1375,6 +1375,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
     }
 
     public void handleCustomPayload(ServerboundCustomPayloadPacket serverboundCustomPayloadPacket) {
+        if (net.blueberrymc.network.server.ServerBlueberryPacketListener.super.handleBlueberryCustomPayload(serverboundCustomPayloadPacket)) return; // Blueberry
     }
 
     public void handleChangeDifficulty(ServerboundChangeDifficultyPacket serverboundChangeDifficultyPacket) {
diff --git a/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
index 6e66f629dde7defbdac6fd3483e5ff0310c6de43..19c52d8d898a9a60eb75575785f24c20b95a1ff9 100644
--- a/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -34,6 +34,7 @@ public class ServerStatusPacketListenerImpl implements ServerStatusPacketListene
         } else {
             this.hasRequestedStatus = true;
             this.connection.send(new ClientboundStatusResponsePacket(this.server.getStatus()));
+            this.connection.send(new net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket(net.blueberrymc.common.Blueberry.getModManager().getModInfos())); // Blueberry
         }
     }
 
diff --git a/src/main/resources/log4j2-debug.xml b/src/main/resources/log4j2-debug.xml
index aea01980678d1a4f74982b0c08f42456d557b124..cabc65a1a5255cb445e4e2ac59f83b759eea9979 100644
--- a/src/main/resources/log4j2-debug.xml
+++ b/src/main/resources/log4j2-debug.xml
@@ -2,13 +2,13 @@
 <Configuration status="WARN" packages="com.mojang.util">
     <Appenders>
         <Console name="SysOut" target="SYSTEM_OUT">
-            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
         </Console>
         <Queue name="ServerGuiConsole">
-            <PatternLayout pattern="[%d{HH:mm:ss} %level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss} %level] [%c]: %msg%n" />
         </Queue>
         <RollingRandomAccessFile name="File" fileName="logs/latest.log" filePattern="logs/%d{yyyy-MM-dd}-%i.log.gz">
-            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
             <Policies>
                 <TimeBasedTriggeringPolicy />
                 <OnStartupTriggeringPolicy />
diff --git a/src/main/resources/log4j2.xml b/src/main/resources/log4j2.xml
index bc53e32751025a515bdb08ea587f6b30202ed095..8b00a0d606dae68533536e2eb8a6b10030a9e098 100644
--- a/src/main/resources/log4j2.xml
+++ b/src/main/resources/log4j2.xml
@@ -2,13 +2,13 @@
 <Configuration status="WARN">
     <Appenders>
         <Console name="SysOut" target="SYSTEM_OUT">
-            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
         </Console>
         <Queue name="ServerGuiConsole">
-            <PatternLayout pattern="[%d{HH:mm:ss} %level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss} %level] [%c]: %msg%n" />
         </Queue>
         <RollingRandomAccessFile name="File" fileName="logs/latest.log" filePattern="logs/%d{yyyy-MM-dd}-%i.log.gz">
-            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
             <Policies>
                 <TimeBasedTriggeringPolicy />
                 <OnStartupTriggeringPolicy />
