From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:21 +0900
Subject: [PATCH] Get our version instead of magmacube's


diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 797c8530d79532afc8b12a173ba58392cb234bfc..0c1d00140bdfebc8a65735f686332906d16dbac1 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -873,7 +873,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 	}
 
 	public String getLaunchedVersion() {
-		return this.launchedVersion;
+		return net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion(); // Blueberry
 	}
 
 	public String getVersionType() {
