From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 Jan 2021 19:48:53 +0900
Subject: [PATCH] Client-side command support


diff --git a/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java b/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java
index b978ac67055c4a5a00f123e35d613f8d0f97caad..c7acfaddbe5542dc3defb6b209a0e4f8a8db4206 100644
--- a/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java
+++ b/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java
@@ -170,7 +170,13 @@ public class CommandSuggestions {
         boolean flag2 = this.commandsOnly || flag;
         int i = this.input.getCursorPosition();
         if (flag2) {
-            CommandDispatcher<SharedSuggestionProvider> commandDispatcher = this.minecraft.player.connection.getCommands();
+            // Blueberry start - client-side command completion
+            CommandDispatcher<SharedSuggestionProvider> commandDispatcher = net.blueberrymc.client.commands.ClientCommandManager.getRoot(this.minecraft.player);
+            CommandDispatcher<SharedSuggestionProvider> mojangCommandDispatcher = this.minecraft.player.connection.getCommands();
+            for (CommandNode<SharedSuggestionProvider> child : mojangCommandDispatcher.getRoot().getChildren()) {
+                commandDispatcher.getRoot().addChild(child);
+            }
+            // Blueberry end - client-side command completion
             if (this.currentParse == null) {
                 this.currentParse = commandDispatcher.parse(stringReader, this.minecraft.player.connection.getSuggestionsProvider());
             }
diff --git a/src/main/java/net/minecraft/client/gui/screens/Screen.java b/src/main/java/net/minecraft/client/gui/screens/Screen.java
index 3f3bf962b7373747246774fcd4467c39271ddfc4..8e7452d64996494f06f836057208893a396c820e 100644
--- a/src/main/java/net/minecraft/client/gui/screens/Screen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/Screen.java
@@ -355,6 +355,12 @@ public abstract class Screen extends AbstractContainerEventHandler implements Wi
             this.minecraft.gui.getChat().addRecentChat(s);
         }
 
+        // Blueberry start
+        if (net.blueberrymc.client.commands.ClientCommandManager.hasCommand(s)) {
+            net.blueberrymc.client.commands.ClientCommandManager.performCommand(this.minecraft.player.createCommandSourceStack(), s);
+            return;
+        }
+        // Blueberry end
         this.minecraft.player.chat(s);
     }
 
