From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <ilyy@outlook.jp>
Date: Fri, 15 Jan 2021 19:48:53 +0900
Subject: [PATCH] Client-side command support


diff --git a/src/main/java/net/minecraft/client/player/LocalPlayer.java b/src/main/java/net/minecraft/client/player/LocalPlayer.java
index 04b3e2a2c6c1315eb830dedad476f506d30a2e92..08bab82a593db19053e91ee2d6b645f4c77c78b9 100644
--- a/src/main/java/net/minecraft/client/player/LocalPlayer.java
+++ b/src/main/java/net/minecraft/client/player/LocalPlayer.java
@@ -267,6 +267,7 @@ public class LocalPlayer extends AbstractClientPlayer {
    }
 
    public void chat(String s) {
+      if (net.blueberrymc.client.commands.ClientCommandManager.hasCommand(s)) net.blueberrymc.client.commands.ClientCommandManager.performCommand(this.createCommandSourceStack(), s); // Blueberry
       this.connection.send(new ServerboundChatPacket(s));
    }
 
