From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 Jan 2021 19:48:53 +0900
Subject: [PATCH] Client-side command support


diff --git a/src/main/java/net/minecraft/client/player/LocalPlayer.java b/src/main/java/net/minecraft/client/player/LocalPlayer.java
index 9469862fb83d5b95bce0e97de50d7ee7e72c0581..c5121bef90d9e7a36cce2632a4c31165d84b3f98 100644
--- a/src/main/java/net/minecraft/client/player/LocalPlayer.java
+++ b/src/main/java/net/minecraft/client/player/LocalPlayer.java
@@ -274,6 +274,7 @@ public class LocalPlayer extends AbstractClientPlayer {
     }
 
     public void chat(String s) {
+        if (net.blueberrymc.client.commands.ClientCommandManager.hasCommand(s)) net.blueberrymc.client.commands.ClientCommandManager.performCommand(this.createCommandSourceStack(), s); // Blueberry
         this.connection.send(new ServerboundChatPacket(s));
     }
 
