From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:46 +0900
Subject: [PATCH] Mods support


diff --git a/src/main/java/com/mojang/blaze3d/platform/Window.java b/src/main/java/com/mojang/blaze3d/platform/Window.java
index 05575121e5b84eeb54e5ec2add4cf31e26b7c193..c6ca0edf81e87c3f22276953d0fb6533d6dd4b09 100644
--- a/src/main/java/com/mojang/blaze3d/platform/Window.java
+++ b/src/main/java/com/mojang/blaze3d/platform/Window.java
@@ -80,7 +80,7 @@ public final class Window implements AutoCloseable {
         GLFW.glfwWindowHint(139267, 2);
         GLFW.glfwWindowHint(139272, 204801);
         GLFW.glfwWindowHint(139270, 1);
-        this.window = GLFW.glfwCreateWindow(this.width, this.height, s2, this.fullscreen && monitor != null ? monitor.getMonitor() : 0L, 0L);
+        this.window = net.blueberrymc.client.EarlyLoadingScreen.getInstance().acquireWindowOrGet(() -> GLFW.glfwCreateWindow(this.width, this.height, s2, this.fullscreen && monitor != null ? monitor.getMonitor() : 0L, 0L));
         if (monitor != null) {
             VideoMode videoMode = monitor.getPreferredVidMode(this.fullscreen ? this.preferredFullscreenVideoMode : Optional.empty());
             this.windowedX = this.x = monitor.getX() + videoMode.getWidth() / 2 - this.width / 2;
diff --git a/src/main/java/net/blueberrymc/client/BlueberryClientImpl.java b/src/main/java/net/blueberrymc/client/BlueberryClientImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..a7c3600502f435aa0814ffe9a3454309682b6fd8
--- /dev/null
+++ b/src/main/java/net/blueberrymc/client/BlueberryClientImpl.java
@@ -0,0 +1,29 @@
+package net.blueberrymc.client;
+
+import com.mojang.blaze3d.systems.RenderSystem;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.screens.MenuScreens;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.gui.screens.inventory.MenuAccess;
+import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.inventory.MenuType;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import org.jetbrains.annotations.NotNull;
+
+public class BlueberryClientImpl extends BlueberryClient {
+    @Override
+    public boolean isOnGameThread() {
+        return RenderSystem.isOnRenderThread()
+                || Thread.currentThread().getName().equals("main")
+                || Thread.currentThread().getName().equals("Server thread"); // may be called from integrated server (logical server)
+    }
+
+    public void registerSpecialBlockEntityRenderer(@NotNull BlockEntityType<?> blockEntityType, @NotNull BlockEntityRenderer<?> blockEntityRenderer) {
+        Minecraft.getInstance().getBlockEntityRenderDispatcher().registerSpecialRenderer(blockEntityType, blockEntityRenderer);
+    }
+
+    public <M extends AbstractContainerMenu, U extends Screen & MenuAccess<M>> void registerMenuScreensFactory(@NotNull MenuType<? extends M> menuType, @NotNull ScreenConstructor<M, U> screenConstructor) {
+        MenuScreens.register(menuType, screenConstructor::create);
+    }
+}
diff --git a/src/main/java/net/blueberrymc/server/BlueberryServerImpl.java b/src/main/java/net/blueberrymc/server/BlueberryServerImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..7b1e2686690e9d7df32131010891223b24eb568f
--- /dev/null
+++ b/src/main/java/net/blueberrymc/server/BlueberryServerImpl.java
@@ -0,0 +1,4 @@
+package net.blueberrymc.server;
+
+public class BlueberryServerImpl extends BlueberryServer {
+}
diff --git a/src/main/java/net/minecraft/CrashReport.java b/src/main/java/net/minecraft/CrashReport.java
index 18d790873a57c1a38abca9874d170eee7d4abfd4..c330c712b76b836eb964d62914095364318900fd 100644
--- a/src/main/java/net/minecraft/CrashReport.java
+++ b/src/main/java/net/minecraft/CrashReport.java
@@ -227,6 +227,6 @@ public class CrashReport {
 
     public static void preload() {
         MemoryReserve.allocate();
-        (new CrashReport("Don't panic!", new Throwable())).getFriendlyReport();
+        LOGGER.info((new CrashReport("Don't panic!", new Throwable())).getFriendlyReport()); // Blueberry
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/SystemReport.java b/src/main/java/net/minecraft/SystemReport.java
index d407afcacaad89de4836abdc8a975f73570614ef..e87e91bc99e99b69b7f319f3bf3f4d2fc55beaef 100644
--- a/src/main/java/net/minecraft/SystemReport.java
+++ b/src/main/java/net/minecraft/SystemReport.java
@@ -49,6 +49,26 @@ public class SystemReport {
             List<String> list = Util.getVmArguments().collect(Collectors.toList());
             return String.format(Locale.ROOT, "%d total; %s", list.size(), String.join(" ", list));
         }));
+        // Blueberry start
+        this.setDetail("Blueberry Version", net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion());
+        this.setDetail("Blueberry commit", net.blueberrymc.common.util.Versioning.getVersion().getCommit());
+        // TODO: we need better layout
+        this.setDetail("Mods", () -> {
+            StringBuilder sb = new StringBuilder("\n");
+            sb.append("      Status:\n");
+            sb.append("        L = Loaded\n");
+            sb.append("        P = Pre Init\n");
+            sb.append("        I = Init\n");
+            sb.append("        J = Post Init\n");
+            sb.append("        A = Available\n");
+            sb.append("        E = Errored\n");
+            sb.append("        U = Unloaded\n");
+            for (net.blueberrymc.common.bml.BlueberryMod mod : net.blueberrymc.common.Blueberry.getModLoader().getLoadedMods()) {
+                sb.append("      ").append(mod.getName()).append(" (").append(mod.getDescription().getModId()).append(") [").append(mod.getDescription().getVersion()).append("] - ").append(mod.getStateList().toString()).append("\n");
+            }
+            return sb.toString();
+        });
+        // Blueberry end
     }
 
     public void setDetail(String s, String s2) {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 4ee819d11a41f718d3437e3c3766039b627d3ac4..2629c0a64e5f4c72d71ed916f41715023f5954cc 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -463,7 +463,9 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
         this.mainRenderTarget = new MainTarget(this.window.getWidth(), this.window.getHeight());
         this.mainRenderTarget.setClearColor(0.0F, 0.0F, 0.0F, 0.0F);
         this.mainRenderTarget.clear(ON_OSX);
+        net.blueberrymc.client.EarlyLoadingScreen.getInstance().startRender(false); // Blueberry
         this.resourceManager = new ReloadableResourceManager(PackType.CLIENT_RESOURCES);
+        net.blueberrymc.common.Blueberry.getModLoader().callPreInit(); // Blueberry
         this.resourcePackRepository.reload();
         this.options.loadSelectedResourcePacks(this.resourcePackRepository);
         this.languageManager = new LanguageManager(this.options.languageCode);
@@ -477,6 +479,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
         this.splashManager = new SplashManager(this.user);
         this.resourceManager.registerReloadListener(this.splashManager);
         this.musicManager = new MusicManager(this);
+        net.blueberrymc.client.EarlyLoadingScreen.getInstance().blockUntilFinish(); // Blueberry
         this.fontManager = new FontManager(this.textureManager);
         this.font = this.fontManager.createFont();
         this.fontFilterFishy = this.fontManager.createFontFilterFishy();
@@ -496,6 +499,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
         this.blockEntityRenderDispatcher = new BlockEntityRenderDispatcher(this.font, this.entityModels, this::getBlockRenderer, this::getItemRenderer, this::getEntityRenderDispatcher);
         this.resourceManager.registerReloadListener(this.blockEntityRenderDispatcher);
         BlockEntityWithoutLevelRenderer blockEntityWithoutLevelRenderer = new BlockEntityWithoutLevelRenderer(this.blockEntityRenderDispatcher, this.entityModels);
+        BlockEntityWithoutLevelRenderer.instance = blockEntityWithoutLevelRenderer; // Blueberry
         this.resourceManager.registerReloadListener(blockEntityWithoutLevelRenderer);
         this.itemRenderer = new ItemRenderer(this.textureManager, this.modelManager, this.itemColors, blockEntityWithoutLevelRenderer);
         this.resourceManager.registerReloadListener(this.itemRenderer);
@@ -553,6 +557,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
         this.reportingContext = ReportingContext.create(ReportEnvironment.local(), this.userApiService);
         LoadingOverlay.registerTextures(this);
         List<PackResources> list = this.resourcePackRepository.openAllSelected();
+        net.blueberrymc.common.Blueberry.getModLoader().callInit(); // Blueberry
         this.reloadStateTracker.startReload(ResourceLoadStateTracker.ReloadReason.INITIAL, list);
         ReloadInstance reloadInstance = this.resourceManager.createReload(Util.backgroundExecutor(), this, RESOURCE_RELOAD_INITIAL_TASK, list);
         this.setOverlay(new LoadingOverlay(this, reloadInstance, (optional) -> Util.ifElse(optional, this::rollbackResourcePacks, () -> {
@@ -564,8 +569,9 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
             }), false));
         if (s != null) {
             ServerAddress serverAddress = new ServerAddress(s, i);
-            reloadInstance.done().thenRunAsync(() -> ConnectScreen.startConnecting(new TitleScreen(), this, serverAddress, new ServerData(I18n.get("selectServer.defaultName"), serverAddress.toString(), false)), this);
+            reloadInstance.done().thenRunAsync(() -> new net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen(() -> ConnectScreen.startConnecting(new TitleScreen(), this, serverAddress, new ServerData(I18n.get("selectServer.defaultName"), serverAddress.toString(), false))), this); // Blueberry
         } else if (this.shouldShowBanNotice()) {
+            this.setScreen(new net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen(() -> { // Blueberry
             this.setScreen(BanNoticeScreen.create((flag) -> {
                 if (flag) {
                     Util.getPlatform().openUri("https://aka.ms/mcjavamoderation");
@@ -573,8 +579,9 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 
                 this.setScreen(new TitleScreen(true));
             }, this.multiplayerBan()));
+            })); // Blueberry
         } else {
-            this.setScreen(new TitleScreen(true));
+            this.setScreen(new net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen(new TitleScreen())); // Blueberry
         }
 
     }
@@ -764,9 +771,10 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
     }
 
     public static void crash(CrashReport crashReport) {
-        File crashReportsDir = new File(getInstance().gameDirectory, "crash-reports");
+        File crashReportsDir = new File(instance == null ? new File(".") : getInstance().gameDirectory, "crash-reports"); // Blueberry
         File crashReportFile = new File(crashReportsDir, "crash-" + Util.getFilenameFormattedDateTime() + "-client.txt");
         Bootstrap.realStdoutPrintln(crashReport.getFriendlyReport());
+        net.blueberrymc.common.util.DiscordRPCTaskExecutor.shutdownNow(); // Blueberry
         if (crashReport.getSaveFile() != null) {
             Bootstrap.realStdoutPrintln("#@!@# Game crashed! Crash report saved to: #@!@# " + crashReport.getSaveFile());
             System.exit(-1);
@@ -913,6 +921,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 
         this.screen = screen;
         BufferUploader.reset();
+        net.blueberrymc.client.event.ClientEventFactory.callScreenChangedEvent(this.screen); // Blueberry
         if (screen != null) {
             this.mouseHandler.releaseMouse();
             KeyMapping.releaseAll();
@@ -928,6 +937,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 
     public void setOverlay(@Nullable Overlay overlay) {
         this.overlay = overlay;
+        net.blueberrymc.client.event.ClientEventFactory.callOverlayChangedEvent(overlay); // Blueberry
     }
 
     public void destroy() {
@@ -952,6 +962,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
                 this.screen.removed();
             }
 
+            net.blueberrymc.common.Blueberry.shutdown();
             this.close();
         } finally {
             Util.timeSource = System::nanoTime;
@@ -1758,6 +1769,10 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
             this.pendingConnection.tick();
         }
 
+        // Blueberry start
+        this.profiler.popPush("blueberryClientScheduler");
+        net.blueberrymc.common.Blueberry.getUtil().getClientScheduler().tick();
+        // Blueberry end
         this.profiler.popPush("keyboard");
         this.keyboardHandler.tick();
         this.profiler.pop();
diff --git a/src/main/java/net/minecraft/client/gui/components/ChatComponent.java b/src/main/java/net/minecraft/client/gui/components/ChatComponent.java
index c8ead3162023dd2a3b145b30976ffa93621835b9..2236abca06310cd283b108158c0c4f48bbedf779 100644
--- a/src/main/java/net/minecraft/client/gui/components/ChatComponent.java
+++ b/src/main/java/net/minecraft/client/gui/components/ChatComponent.java
@@ -448,15 +448,27 @@ public class ChatComponent extends GuiComponent {
     }
 
     public static int getWidth(double d) {
+        // Blueberry start
+        int width = 320;
+        if (net.blueberrymc.common.bml.InternalBlueberryModConfig.Misc.ChatSettings.extendedWidth && Minecraft.getInstance().getWindow() != null) {
+            width = Minecraft.getInstance().getWindow().getWidth() / 2 - 8;
+        }
+        // Blueberry end
         int i = 320;
         int i2 = 40;
-        return Mth.floor(d * 280.0D + 40.0D);
+        return Mth.floor(d * width); // Blueberry
     }
 
     public static int getHeight(double d) {
+        // Blueberry start
+        int height = 180;
+        if (net.blueberrymc.common.bml.InternalBlueberryModConfig.Misc.ChatSettings.extendedHeight && Minecraft.getInstance().getWindow() != null) {
+            height = Minecraft.getInstance().getWindow().getHeight() / 2 - 40;
+        }
+        // Blueberry end
         int i = 180;
         int i2 = 20;
-        return Mth.floor(d * 160.0D + 20.0D);
+        return Mth.floor(d * height); // Blueberry
     }
 
     public static double defaultUnfocusedPct() {
diff --git a/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java b/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
index 546720f0922a048fce26cdcc7039be75ff38436c..32a2cb4747a1baec11adb88a7690d02a70ba5893 100644
--- a/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
+++ b/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
@@ -71,6 +71,7 @@ import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import org.joml.Matrix4f;
+import net.blueberrymc.common.Blueberry; // Blueberry
 
 public class DebugScreenOverlay extends GuiComponent {
     private static final int COLOR_GREY = 14737632;
@@ -354,6 +355,12 @@ public class DebugScreenOverlay extends GuiComponent {
         long l3 = Runtime.getRuntime().freeMemory();
         long l4 = l2 - l3;
         List<String> list = Lists.newArrayList(new String[]{String.format(Locale.ROOT, "Java: %s %dbit", System.getProperty("java.version"), this.minecraft.is64Bit() ? 64 : 32), String.format(Locale.ROOT, "Mem: % 2d%% %03d/%03dMB", l4 * 100L / l, bytesToMegabytes(l4), bytesToMegabytes(l)), String.format(Locale.ROOT, "Allocation rate: %03dMB /s", bytesToMegabytes(this.allocationRateCalculator.bytesAllocatedPerSecond(l4))), String.format(Locale.ROOT, "Allocated: % 2d%% %03dMB", l2 * 100L / l, bytesToMegabytes(l2)), "", String.format(Locale.ROOT, "CPU: %s", GlUtil.getCpuInfo()), "", String.format(Locale.ROOT, "Display: %dx%d (%s)", Minecraft.getInstance().getWindow().getWidth(), Minecraft.getInstance().getWindow().getHeight(), GlUtil.getVendor()), GlUtil.getRenderer(), GlUtil.getOpenGLVersion()});
+        // Blueberry start
+        list.add("");
+        list.add("MagmaCube " + Blueberry.getVersion().getShortMagmaCubeCommit());
+        list.add("Blueberry " + Blueberry.getVersion().getFullyQualifiedVersion());
+        list.add(Blueberry.getModLoader().getLoadedMods().size() + " mods loaded, " + Blueberry.getModLoader().getActiveMods().size() + " mods active");
+        // Blueberry end
         if (this.minecraft.showOnlyReducedInfo()) {
             return list;
         } else {
diff --git a/src/main/java/net/minecraft/client/gui/font/FontManager.java b/src/main/java/net/minecraft/client/gui/font/FontManager.java
index bbae7d696df930a090c530c278eee0526b22e4f9..ff100e69e1271687158fe33d4a032c1e50c64d6e 100644
--- a/src/main/java/net/minecraft/client/gui/font/FontManager.java
+++ b/src/main/java/net/minecraft/client/gui/font/FontManager.java
@@ -140,6 +140,7 @@ public class FontManager implements AutoCloseable {
             });
             profilerFiller.pop();
             profilerFiller.endTick();
+            net.minecraft.client.gui.screens.LoadingOverlay.isFontReady = true; // Blueberry
         }
 
         public String getName() {
diff --git a/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java b/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
index b719387c9ed89c1937e258c82a80acad8394bded..f003c460d8db61e8c2ac88102b01ca35ca01a125 100644
--- a/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -25,6 +25,7 @@ import net.minecraft.util.FastColor;
 import net.minecraft.util.Mth;
 
 public class LoadingOverlay extends Overlay {
+    public static boolean isFontReady = false; // Blueberry
     static final ResourceLocation MOJANG_STUDIOS_LOGO_LOCATION = new ResourceLocation("textures/gui/title/mojangstudios.png");
     private static final int LOGO_BACKGROUND_COLOR = FastColor.ARGB32.color(255, 239, 50, 61);
     private static final int LOGO_BACKGROUND_COLOR_DARK = FastColor.ARGB32.color(255, 0, 0, 0);
@@ -113,6 +114,7 @@ public class LoadingOverlay extends Overlay {
         blit(poseStack, i8, i9 - i10, i11, (int)d, 0.0625F, 60.0F, 120, 60, 120, 120);
         RenderSystem.defaultBlendFunc();
         RenderSystem.disableBlend();
+        net.blueberrymc.client.EarlyLoadingScreen.getInstance().renderMessagesFromGUI(poseStack); // Blueberry
         int i12 = (int)((double)this.minecraft.getWindow().getGuiScaledHeight() * 0.8325D);
         float f10 = this.reload.getActualProgress();
         this.currentProgress = Mth.clamp(this.currentProgress * 0.95F + f10 * 0.050000012F, 0.0F, 1.0F);
@@ -121,6 +123,7 @@ public class LoadingOverlay extends Overlay {
         }
 
         if (f2 >= 2.0F) {
+            net.blueberrymc.common.Blueberry.getModLoader().callPostInit(); // Blueberry
             this.minecraft.setOverlay((Overlay)null);
         }
 
@@ -134,6 +137,7 @@ public class LoadingOverlay extends Overlay {
 
             this.fadeOutStart = Util.getMillis();
             if (this.minecraft.screen != null) {
+                if (this.minecraft.screen instanceof net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen) ((net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen) minecraft.screen).refresh(); // Blueberry
                 this.minecraft.screen.init(this.minecraft, this.minecraft.getWindow().getGuiScaledWidth(), this.minecraft.getWindow().getGuiScaledHeight());
             }
         }
diff --git a/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java b/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
index 75bccbd8046e6f31050fb9a9a52c192bad5a3159..0a5d7eb490599cd6fa76e5a6d93312722ffe230e 100644
--- a/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
@@ -84,6 +84,7 @@ public class PauseScreen extends Screen {
         gridWidget.pack();
         FrameWidget.alignInRectangle(gridWidget, 0, 0, this.width, this.height, 0.5F, 0.25F);
         this.addRenderableWidget(gridWidget);
+        this.addRenderableWidget(Button.builder(net.blueberrymc.common.resources.BlueberryText.text("blueberry", "gui.screens.mods"), (buttonx) -> net.blueberrymc.client.gui.screens.ModListScreen.switchToModListScreen()).bounds(this.width / 2 - 102, this.height / 4 + 144 - 16, 204, 20).build()); // Blueberry (TODO)
     }
 
     private void onDisconnect() {
diff --git a/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java b/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
index cd03142651556e2c0c38f3405e541e8ffd8ba408..5912e29963717e4134990eca4a42d18b81fc2d44 100644
--- a/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
@@ -13,6 +13,9 @@ import java.io.IOException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
 import javax.annotation.Nullable;
+
+import net.blueberrymc.client.gui.screens.ModListScreen;
+import net.blueberrymc.common.resources.BlueberryText;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.client.Minecraft;
@@ -144,7 +147,10 @@ public class TitleScreen extends Screen {
             Screen screen = (Screen)(this.minecraft.options.skipMultiplayerWarning ? new JoinMultiplayerScreen(this) : new SafetyScreen(this));
             this.minecraft.setScreen(screen);
         }).bounds(this.width / 2 - 100, i + i2 * 1, 200, 20).tooltip(tooltip).build())).active = flag;
-        (this.addRenderableWidget(Button.builder(Component.translatable("menu.online"), (button) -> this.realmsButtonClicked()).bounds(this.width / 2 - 100, i + i2 * 2, 200, 20).tooltip(tooltip).build())).active = flag;
+        // Blueberry start
+        (this.addRenderableWidget(Button.builder(BlueberryText.text("blueberry", "gui.screens.mods"), (button) -> ModListScreen.switchToModListScreen()).bounds(this.width / 2 - 100, i + i2 * 2, 98, 20).tooltip(tooltip).build())).active = flag;
+        (this.addRenderableWidget(Button.builder(Component.translatable("menu.online"), (button) -> this.realmsButtonClicked()).bounds(this.width / 2 + 2, i + i2 * 2, 98, 20).tooltip(tooltip).build())).active = flag;
+        // Blueberry end
     }
 
     @Nullable
@@ -306,8 +312,12 @@ public class TitleScreen extends Screen {
                 s = s + I18n.get("menu.modded");
             }
 
-            drawString(poseStack, this.font, "MagmaCube " + ((net.minecraft.DetectedVersion) SharedConstants.getCurrentVersion()).magmaCubeVersion, 2, this.height - 20, 0xFFFFFF | i6); // MagmaCube
-            drawString(poseStack, this.font, s, 2, this.height - 10, 16777215 | i6);
+            // Blueberry start
+            drawString(poseStack, this.font, "Blueberry " + net.blueberrymc.common.Blueberry.getVersion().getFullyQualifiedVersion(), 2, this.height - 40, 0xFFFFFF | i6);
+            drawString(poseStack, this.font, "MagmaCube " + net.blueberrymc.common.util.Versioning.getVersion().getShortMagmaCubeCommit(), 2, this.height - 30, 0xFFFFFF | i6); // MagmaCube // Blueberry - offset 20 -> 30
+            drawString(poseStack, this.font, s, 2, this.height - 20, 16777215 | i6); // Blueberry - offset 10 -> 20
+            drawString(poseStack, this.font, net.blueberrymc.common.Blueberry.getModLoader().getLoadedMods().size() + " mods loaded, " + net.blueberrymc.common.Blueberry.getModLoader().getActiveMods().size() + " mods active", 2, this.height - 10, 0xFFFFFF | i6);
+            // Blueberry end
 
             for(GuiEventListener guiEventListener : this.children()) {
                 if (guiEventListener instanceof AbstractWidget) {
diff --git a/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java b/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
index aac499837aba5984508919059b2e775f5c1549e3..ec74c3840a19b62475ae3e02e2f13d0483c5f502 100644
--- a/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
+++ b/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
@@ -282,6 +282,13 @@ public class ServerSelectionList extends ObjectSelectionList<ServerSelectionList
             RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             RenderSystem.setShaderTexture(0, GuiComponent.GUI_ICONS_LOCATION);
             GuiComponent.blit(poseStack, i3 + i4 - 15, i2, (float)(i10 * 10), (float)(176 + i11 * 8), 10, 8, 256, 256);
+            // Blueberry start
+            if (serverData.serverType != null) {
+                RenderSystem.setShader(GameRenderer::getPositionTexColorShader);
+                RenderSystem.setShaderTexture(0, net.blueberrymc.client.gui.BlueberryGuiComponents.GUI_ICONS_LOCATION);
+                GuiComponent.blit(poseStack, i3 + i4 - 18, i2 + 8, (float) 0.0F, (float) (serverData.serverType.getOffset() * 16), 16, 16, 256, 256);
+            }
+            // Blueberry end
             String s = this.serverData.getIconB64();
             if (!Objects.equals(s, this.lastIconB64)) {
                 if (this.uploadServerIcon(s)) {
@@ -302,6 +309,12 @@ public class ServerSelectionList extends ObjectSelectionList<ServerSelectionList
             int i20 = i7 - i2;
             if (i19 >= i4 - 15 && i19 <= i4 - 5 && i20 >= 0 && i20 <= 8) {
                 this.screen.setToolTip(Collections.singletonList(component2));
+                // Blueberry start
+            } else if (serverData.serverType != null && i19 >= i4 - 18 && i19 <= i4 - 5 && i20 >= 0 && i20 <= 22) {
+                String text = serverData.serverType.getBlueberryText().getContents();
+                if (text.contains("%d")) text = String.format(text, serverData.modsCount);
+                this.screen.setToolTip(Collections.singletonList(Component.literal(text)));
+                // Blueberry end
             } else if (i19 >= i4 - i9 - 15 - 2 && i19 <= i4 - 15 - 2 && i20 >= 0 && i20 <= 8) {
                 this.screen.setToolTip(list2);
             }
diff --git a/src/main/java/net/minecraft/client/main/Main.java b/src/main/java/net/minecraft/client/main/Main.java
index 6813cc831731506dedf18d0b475918c149ffb8b8..12d7e2656e89ead6bd08958757328d76fc32d1c8 100644
--- a/src/main/java/net/minecraft/client/main/Main.java
+++ b/src/main/java/net/minecraft/client/main/Main.java
@@ -50,9 +50,9 @@ public class Main {
     }
 
     public static void run(String[] strings, boolean flag) {
-        SharedConstants.tryDetectVersion();
+        //SharedConstants.tryDetectVersion(); // Blueberry - moved below
         if (flag) {
-            SharedConstants.enableDataFixerOptimizations();
+            //SharedConstants.enableDataFixerOptimizations(); // Blueberry - moved below
         }
 
         OptionParser optionParser = new OptionParser();
@@ -93,6 +93,12 @@ public class Main {
         if (!list.isEmpty()) {
             System.out.println("Completely ignored arguments: " + list);
         }
+        // Blueberry start
+        net.blueberrymc.common.Blueberry.preBootstrap();
+        net.blueberrymc.common.Blueberry.bootstrap(new net.blueberrymc.client.BlueberryClientImpl());
+        SharedConstants.tryDetectVersion();
+        if (flag) SharedConstants.enableDataFixerOptimizations();
+        // Blueberry end
 
         String s = parseArgument(optionSet, optionSpec7);
         Proxy proxy = Proxy.NO_PROXY;
diff --git a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
index e302d3dba0327b9c1d9896fb13c306e85549645d..d0566b508c7ad8aa0b27172142ff0e0eb0c5af0e 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -304,7 +304,7 @@ import net.minecraft.world.scores.Team;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 import org.slf4j.Logger;
 
-public class ClientPacketListener implements TickablePacketListener, ClientGamePacketListener {
+public class ClientPacketListener implements TickablePacketListener, ClientGamePacketListener, net.blueberrymc.network.client.ClientBlueberryPacketListener { // Blueberry
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Component GENERIC_DISCONNECT_MESSAGE = Component.translatable("disconnect.lost");
     private static final Component UNSECURE_SERVER_TOAST_TITLE = Component.translatable("multiplayer.unsecureserver.toast.title");
@@ -1738,6 +1738,7 @@ public class ClientPacketListener implements TickablePacketListener, ClientGameP
 
         try {
             friendlyByteBuf = clientboundCustomPayloadPacket.getData();
+            if (net.blueberrymc.network.client.ClientBlueberryPacketListener.super.handleBlueberryCustomPayload(clientboundCustomPayloadPacket)) return; // Blueberry
             if (ClientboundCustomPayloadPacket.BRAND.equals(resourceLocation)) {
                 String s = friendlyByteBuf.readUtf();
                 this.minecraft.player.setServerBrand(s);
diff --git a/src/main/java/net/minecraft/client/multiplayer/ServerData.java b/src/main/java/net/minecraft/client/multiplayer/ServerData.java
index a6c32fd8710fb84e689230d6034d37d8352edb48..54d9e206733394eddc32cf3ca52fae3ca952e604 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ServerData.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ServerData.java
@@ -23,6 +23,10 @@ public class ServerData {
     private String iconB64;
     private boolean lan;
     private boolean enforcesSecureChat;
+    // Blueberry start
+    public net.blueberrymc.network.ServerType serverType = null;
+    public int modsCount = 0;
+    // Blueberry end
 
     public ServerData(String s, String s2, boolean flag) {
         this.name = s;
diff --git a/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java b/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
index 2d566c70f42c5df2c6a23d4f8d6b654cee130a1b..9ef5347c7e33813a00599375e41959d723bd20ac 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
@@ -75,6 +75,7 @@ public class ServerStatusPinger {
                         connection.disconnect(Component.translatable("multiplayer.status.unrequested"));
                     } else {
                         this.receivedPing = true;
+                        serverData.serverType = net.blueberrymc.network.ServerType.VANILLA; // Blueberry
                         ServerStatus serverStatus = clientboundStatusResponsePacket.getStatus();
                         if (serverStatus.getDescription() != null) {
                             serverData.motd = serverStatus.getDescription();
@@ -144,6 +145,19 @@ public class ServerStatusPinger {
 
                 }
 
+                // Blueberry start
+                @Override
+                public void handleBlueberryHandshakeResponse(net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket packet) {
+                    boolean compatible = net.blueberrymc.common.util.ListUtils.isCompatible(packet.getModInfos(), net.blueberrymc.common.Blueberry.getModLoader().getModInfos());
+                    if (compatible) {
+                        serverData.serverType = net.blueberrymc.network.ServerType.BLUEBERRY_GOOD;
+                    } else {
+                        serverData.serverType = net.blueberrymc.network.ServerType.BLUEBERRY_BAD;
+                    }
+                    serverData.modsCount = packet.getModInfos().size();
+                }
+                // Blueberry end
+
                 public Connection getConnection() {
                     return connection;
                 }
diff --git a/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java b/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java
index f3d0cc2722ccb4cd4f830a0defeb13ce5e4d7969..c036ba38f9a3b2c81cbd42ff01e3a461c22b793d 100644
--- a/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java
@@ -57,6 +57,7 @@ import org.apache.commons.lang3.StringUtils;
 public class BlockEntityWithoutLevelRenderer implements ResourceManagerReloadListener {
     private static final ShulkerBoxBlockEntity[] SHULKER_BOXES = (ShulkerBoxBlockEntity[])Arrays.stream(DyeColor.values()).sorted(Comparator.comparingInt(DyeColor::getId)).map((dyeColor) -> new ShulkerBoxBlockEntity(dyeColor, BlockPos.ZERO, Blocks.SHULKER_BOX.defaultBlockState())).toArray((i) -> new ShulkerBoxBlockEntity[i]);
     private static final ShulkerBoxBlockEntity DEFAULT_SHULKER_BOX = new ShulkerBoxBlockEntity(BlockPos.ZERO, Blocks.SHULKER_BOX.defaultBlockState());
+    public static BlockEntityWithoutLevelRenderer instance; // Blueberry
     private final ChestBlockEntity chest = new ChestBlockEntity(BlockPos.ZERO, Blocks.CHEST.defaultBlockState());
     private final ChestBlockEntity trappedChest = new TrappedChestBlockEntity(BlockPos.ZERO, Blocks.TRAPPED_CHEST.defaultBlockState());
     private final EnderChestBlockEntity enderChest = new EnderChestBlockEntity(BlockPos.ZERO, Blocks.ENDER_CHEST.defaultBlockState());
diff --git a/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java b/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
index 7db4ec20b10132dfd7796efe15687779fcedd89e..13eeac1d2b838db19055d77f684fd17c8b51dd3a 100644
--- a/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
+++ b/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
@@ -97,7 +97,10 @@ public class BlockRenderDispatcher implements ResourceManagerReloadListener {
                 this.modelRenderer.renderModel(poseStack.last(), multiBufferSource.getBuffer(ItemBlockRenderTypes.getRenderType(blockState, false)), blockState, bakedModel, f, f2, f3, i, i2);
                 break;
             case ENTITYBLOCK_ANIMATED:
-                this.blockEntityRenderer.renderByItem(new ItemStack(blockState.getBlock()), ItemTransforms.TransformType.NONE, poseStack, multiBufferSource, i, i2);
+                // Blueberry start
+                ItemStack stack = new ItemStack(blockState.getBlock());
+                stack.getItem().getRenderer().renderByItem(stack, ItemTransforms.TransformType.NONE, poseStack, multiBufferSource, i, i2);
+                // Blueberry end
             }
 
         }
@@ -105,5 +108,6 @@ public class BlockRenderDispatcher implements ResourceManagerReloadListener {
 
     public void onResourceManagerReload(ResourceManager resourceManager) {
         this.liquidBlockRenderer.setupSprites();
+        net.blueberrymc.client.world.level.fluid.FluidSpriteManager.setupSprites(); // Blueberry
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java b/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
index b1a6665b6167212e4b5ac81da490089e526b1a0c..19d71b147f6d67d9ce96e0beaafad45872647c5e 100644
--- a/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -67,6 +67,10 @@ public class LiquidBlockRenderer {
         boolean flag = fluidState.is(FluidTags.LAVA);
         TextureAtlasSprite[] textureAtlasSprites = flag ? this.lavaIcons : this.waterIcons;
         int i = flag ? 16777215 : BiomeColors.getAverageWaterColor(blockAndTintGetter, blockPos);
+        // Blueberry start
+        net.blueberrymc.client.event.render.LiquidBlockRenderEvent event = net.blueberrymc.client.event.ClientEventFactory.callLiquidBlockRenderEvent(fluidState, blockPos, i);
+        i = event.getColor();
+        // Blueberry end
         float f = (float)(i >> 16 & 255) / 255.0F;
         float f2 = (float)(i >> 8 & 255) / 255.0F;
         float f3 = (float)(i & 255) / 255.0F;
diff --git a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
index 5481adf92979505840ccbfa1ae721e92198e0030..70811ee71d702e913a8d4ac1ddce6256d9708ca3 100644
--- a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
+++ b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
@@ -24,7 +24,7 @@ import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.phys.HitResult;
 
-public class BlockEntityRenderDispatcher implements ResourceManagerReloadListener {
+public class BlockEntityRenderDispatcher implements ResourceManagerReloadListener, net.blueberrymc.client.renderer.blockentity.MinecraftBlockEntityRenderDispatcher { // Blueberry - implement MinecraftBlockEntityRenderDispatcher
     private Map<BlockEntityType<?>, BlockEntityRenderer<?>> renderers = ImmutableMap.of();
     private final Font font;
     private final EntityModelSet entityModelSet;
@@ -113,4 +113,11 @@ public class BlockEntityRenderDispatcher implements ResourceManagerReloadListene
         BlockEntityRendererProvider.Context context = new BlockEntityRendererProvider.Context(this, (BlockRenderDispatcher)this.blockRenderDispatcher.get(), (ItemRenderer)this.itemRenderer.get(), (EntityRenderDispatcher)this.entityRenderer.get(), this.entityModelSet, this.font);
         this.renderers = BlockEntityRenderers.createEntityRenderers(context);
     }
+
+    // Blueberry start
+    @Override
+    public final void registerSpecialRenderer(BlockEntityType<?> blockEntityType, BlockEntityRenderer<?> blockEntityRenderer) {
+        this.renderers.put(blockEntityType, blockEntityRenderer);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
index 7436fe8d37a6f914cf5e86d733d63a26a33b2507..626e6ac85f6083aad6f93e631aa953486eeb6eb5 100644
--- a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
+++ b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
@@ -11,7 +11,8 @@ import net.minecraft.world.level.block.entity.BlockEntityType;
 public class BlockEntityRenderers {
     private static final Map<BlockEntityType<?>, BlockEntityRendererProvider<?>> PROVIDERS = Maps.newHashMap();
 
-    private static <T extends BlockEntity> void register(BlockEntityType<? extends T> blockEntityType, BlockEntityRendererProvider<T> blockEntityRendererProvider) {
+    // Blueberry - private -> public
+    public static <T extends BlockEntity> void register(BlockEntityType<? extends T> blockEntityType, BlockEntityRendererProvider<T> blockEntityRendererProvider) {
         PROVIDERS.put(blockEntityType, blockEntityRendererProvider);
     }
 
diff --git a/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java b/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
index 7bd250de3badc39488d46a254b9198d51fc3e7d9..0bdb0694a4288bc0d3a0ba5e677163d7db2952d0 100644
--- a/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
@@ -23,6 +23,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.Vec3;
 import org.joml.Matrix4f;
+import static net.blueberrymc.common.bml.InternalBlueberryModConfig.Debug.DebugRenderers.*; // Blueberry
 
 public class DebugRenderer {
     public final PathfindingRenderer pathfindingRenderer = new PathfindingRenderer();
@@ -93,7 +94,25 @@ public class DebugRenderer {
 
     public void render(PoseStack poseStack, MultiBufferSource.BufferSource bufferSource, double d, double d2, double d3) {
         if (this.renderChunkborder && !Minecraft.getInstance().showOnlyReducedInfo()) {
-            this.chunkBorderRenderer.render(poseStack, bufferSource, d, d2, d3);
+            // Blueberry start
+            if (pathfinding) this.pathfindingRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (waterDebug) this.waterDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (chunkBorder) this.chunkBorderRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (heightMap) this.heightMapRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (collisionBox) this.collisionBoxRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (neighborsUpdate) this.neighborsUpdateRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (structure) this.structureRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (lightDebug) this.lightDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (worldGenAttempt) this.worldGenAttemptRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (solidFace) this.solidFaceRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (chunk) this.chunkRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (brainDebug) this.brainDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (villageSectionsDebug) this.villageSectionsDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (beeDebug) this.beeDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (raidDebug) this.raidDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (goalSelector) this.goalSelectorRenderer.render(poseStack, bufferSource, d, d2, d3);
+            if (gameTestDebug) this.gameTestDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
+            // Blueberry end
         }
 
         this.gameTestDebugRenderer.render(poseStack, bufferSource, d, d2, d3);
diff --git a/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java b/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java
index cd387f8dd3609f80ae3db9e194b167fabf51eef9..f18f8891d608e0b8e18690e2af1f15faec624721 100644
--- a/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java
@@ -159,7 +159,7 @@ public class ItemRenderer implements ResourceManagerReloadListener {
 
                 this.renderModelLists(bakedModel, itemStack, i, i2, poseStack, vertexConsumer);
             } else {
-                this.blockEntityRenderer.renderByItem(itemStack, transformType, poseStack, multiBufferSource, i, i2);
+                itemStack.getItem().getRenderer().renderByItem(itemStack, transformType, poseStack, multiBufferSource, i, i2); // Blueberry
             }
 
             poseStack.popPose();
diff --git a/src/main/java/net/minecraft/client/resources/model/ModelBakery.java b/src/main/java/net/minecraft/client/resources/model/ModelBakery.java
index 8b2736c26a54ede6bcc4c6f8674959ced5e0d9ce..f64cf6413a13e1bece4f3376533f20af01978714 100644
--- a/src/main/java/net/minecraft/client/resources/model/ModelBakery.java
+++ b/src/main/java/net/minecraft/client/resources/model/ModelBakery.java
@@ -103,6 +103,13 @@ public class ModelBakery {
     private final Map<ResourceLocation, BakedModel> bakedTopLevelModels = Maps.newHashMap();
     private int nextModelGroup = 1;
     private final Object2IntMap<BlockState> modelGroups = Util.make(new Object2IntOpenHashMap(), (object2IntOpenHashMap) -> object2IntOpenHashMap.defaultReturnValue(-1));
+    // Blueberry start
+    private static final Set<ResourceLocation> specialModels = new java.util.HashSet<>();
+
+    public static void addSpecialModel(ResourceLocation resourceLocation) {
+        specialModels.add(resourceLocation);
+    }
+    // Blueberry end
 
     public ModelBakery(BlockColors blockColors, ProfilerFiller profilerFiller, Map<ResourceLocation, BlockModel> map, Map<ResourceLocation, List<ModelBakery.LoadedJson>> map2) {
         this.blockColors = blockColors;
@@ -133,8 +140,10 @@ public class ModelBakery {
         }
 
         profilerFiller.popPush("special");
+        new net.blueberrymc.client.event.SpecialModelRegistryEvent().callEvent(); // Blueberry
         this.loadTopLevel(ItemRenderer.TRIDENT_IN_HAND_MODEL);
         this.loadTopLevel(ItemRenderer.SPYGLASS_IN_HAND_MODEL);
+        for (ResourceLocation resourceLocation : getSpecialModels()) this.addModelToCache(resourceLocation); // Blueberry
         this.topLevelModels.values().forEach((unbakedModel) -> unbakedModel.resolveParents(this::getModel));
         profilerFiller.pop();
     }
@@ -344,6 +353,18 @@ public class ModelBakery {
         this.topLevelModels.put(modelResourceLocation, unbakedModel);
     }
 
+    // Blueberry start
+    private void addModelToCache(ResourceLocation resourceLocation) {
+        UnbakedModel unbakedModel = this.getModel(resourceLocation);
+        this.unbakedCache.put(resourceLocation, unbakedModel);
+        this.topLevelModels.put(resourceLocation, unbakedModel);
+    }
+
+    public Set<ResourceLocation> getSpecialModels() {
+        return specialModels;
+    }
+    // Blueberry end
+
     private void registerModelGroup(Iterable<BlockState> iterable) {
         int i = this.nextModelGroup++;
         iterable.forEach((blockState) -> this.modelGroups.put(blockState, i));
diff --git a/src/main/java/net/minecraft/network/ConnectionProtocol.java b/src/main/java/net/minecraft/network/ConnectionProtocol.java
index 80b4319e3f8b82eaa13a8c84d501fc4f0025f2ca..fb39ded4adfc93e229b5c9824382b87ea6c1757b 100644
--- a/src/main/java/net/minecraft/network/ConnectionProtocol.java
+++ b/src/main/java/net/minecraft/network/ConnectionProtocol.java
@@ -198,6 +198,12 @@ public enum ConnectionProtocol {
     STATUS(1, protocol().addFlow(PacketFlow.SERVERBOUND, (new ConnectionProtocol.PacketSet<ServerStatusPacketListener>()).addPacket(ServerboundStatusRequestPacket.class, ServerboundStatusRequestPacket::new).addPacket(ServerboundPingRequestPacket.class, ServerboundPingRequestPacket::new)).addFlow(PacketFlow.CLIENTBOUND, (new ConnectionProtocol.PacketSet<ClientStatusPacketListener>()).addPacket(ClientboundStatusResponsePacket.class, ClientboundStatusResponsePacket::new).addPacket(ClientboundPongResponsePacket.class, ClientboundPongResponsePacket::new))),
     LOGIN(2, protocol().addFlow(PacketFlow.CLIENTBOUND, (new ConnectionProtocol.PacketSet<ClientLoginPacketListener>()).addPacket(ClientboundLoginDisconnectPacket.class, ClientboundLoginDisconnectPacket::new).addPacket(ClientboundHelloPacket.class, ClientboundHelloPacket::new).addPacket(ClientboundGameProfilePacket.class, ClientboundGameProfilePacket::new).addPacket(ClientboundLoginCompressionPacket.class, ClientboundLoginCompressionPacket::new).addPacket(ClientboundCustomQueryPacket.class, ClientboundCustomQueryPacket::new)).addFlow(PacketFlow.SERVERBOUND, (new ConnectionProtocol.PacketSet<ServerLoginPacketListener>()).addPacket(ServerboundHelloPacket.class, ServerboundHelloPacket::new).addPacket(ServerboundKeyPacket.class, ServerboundKeyPacket::new).addPacket(ServerboundCustomQueryPacket.class, ServerboundCustomQueryPacket::new)));
 
+    // Blueberry start
+    static {
+        ((PacketSet<ClientStatusPacketListener>) STATUS.flows.get(PacketFlow.CLIENTBOUND)).addPacket(net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket.class, net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket::new);
+    }
+    // Blueberry end
+
     private static final int MIN_PROTOCOL_ID = -1;
     private static final int MAX_PROTOCOL_ID = 2;
     private static final ConnectionProtocol[] LOOKUP = new ConnectionProtocol[4];
diff --git a/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java b/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
index 43f39473c5ef3de9e53e6722bea6a0a58382526e..f5254cecd40e9d643cb59316e929560adb08c85d 100644
--- a/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
+++ b/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
@@ -2,7 +2,7 @@ package net.minecraft.network.protocol.status;
 
 import net.minecraft.network.PacketListener;
 
-public interface ClientStatusPacketListener extends PacketListener {
+public interface ClientStatusPacketListener extends PacketListener, net.blueberrymc.network.client.handshake.ClientBlueberryHandshakePacketListener { // Blueberry
     void handleStatusResponse(ClientboundStatusResponsePacket var1);
 
     void handlePongResponse(ClientboundPongResponsePacket var1);
diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index 3a5da7cd65866508ed4843d5c64edf5036462a63..ee8aa2dd61ce1fef6d17aa3224e1925d57093572 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -46,6 +46,7 @@ public class Bootstrap {
                     EntitySelectorOptions.bootStrap();
                     DispenseItemBehavior.bootStrap();
                     CauldronInteraction.bootStrap();
+                    net.blueberrymc.common.event.lifecycle.RegistryBootstrappedEvent.fire(); // Blueberry
                     BuiltInRegistries.bootStrap();
                     wrapStreams();
                 }
diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 6598410c9a23b4b58ca67b9418d63955006a8145..bf50f1e8ecf4e79662c14bea5017cf628aee63e1 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -62,7 +62,7 @@ public class Main {
 
     @DontObfuscate
     public static void main(String[] args) {
-        SharedConstants.tryDetectVersion();
+        //SharedConstants.tryDetectVersion(); // Blueberry - moved below
         OptionParser optionParser = new OptionParser();
         OptionSpec<Void> optionSpec = optionParser.accepts("nogui");
         OptionSpec<Void> optionSpec2 = optionParser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
@@ -79,6 +79,7 @@ public class Main {
         OptionSpec<String> optionSpec13 = optionParser.accepts("serverId").withRequiredArg();
         OptionSpec<Void> optionSpec14 = optionParser.accepts("jfrProfile");
         OptionSpec<String> optionSpec15 = optionParser.nonOptions();
+        optionParser.allowsUnrecognizedOptions(); // Blueberry
 
         try {
             OptionSet optionSet = optionParser.parse(args);
@@ -86,6 +87,11 @@ public class Main {
                 optionParser.printHelpOn(System.err);
                 return;
             }
+            // Blueberry start
+            net.blueberrymc.common.Blueberry.preBootstrap();
+            net.blueberrymc.common.Blueberry.bootstrap(new net.blueberrymc.server.BlueberryServerImpl());
+            SharedConstants.tryDetectVersion();
+            // Blueberry end
 
             CrashReport.preload();
             if (optionSet.has(optionSpec14)) {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 515b94a262495060e46b8a7515d5d94ee380d2b1..67cd86b8bada4c35dc541bceee487a53e9b96008 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -270,6 +270,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.proxy = proxy;
             this.packRepository = packRepository;
             this.resources = new MinecraftServer.ReloadableResources(worldStem.resourceManager(), worldStem.dataPackResources());
+            // Blueberry start
+            if (net.blueberrymc.common.Blueberry.isServer()) {
+                ((net.blueberrymc.server.BlueberryServer) net.blueberrymc.common.Blueberry.getUtil()).setServer(this);
+                net.blueberrymc.common.Blueberry.getModLoader().callPreInit();
+            }
+            // Blueberry end
             this.services = services;
             if (services.profileCache() != null) {
                 services.profileCache().setExecutor(this);
@@ -285,6 +291,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.structureTemplateManager = new StructureTemplateManager(worldStem.resourceManager(), levelStorageAccess, dataFixer, holderGetter);
             this.serverThread = thread;
             this.executor = Util.backgroundExecutor();
+            if (net.blueberrymc.common.Blueberry.isServer()) net.blueberrymc.common.Blueberry.getModLoader().callInit(); // Blueberry
         }
     }
 
@@ -624,6 +631,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.status.setVersion(new ServerStatus.Version(SharedConstants.getCurrentVersion().getName(), SharedConstants.getCurrentVersion().getProtocolVersion()));
             this.status.setEnforcesSecureChat(this.enforceSecureProfile());
             this.updateStatusIcon(this.status);
+            if (net.blueberrymc.common.Blueberry.isServer()) net.blueberrymc.common.Blueberry.getModLoader().callPostInit(); // Blueberry
 
             while(this.running) {
                 long l = Util.getMillis() - this.nextTickTime;
@@ -818,6 +826,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             LOGGER.debug("Autosave finished");
         }
 
+        // Blueberry start
+        this.profiler.push("blueberryServerScheduler");
+        net.blueberrymc.common.Blueberry.getUtil().getServerScheduler().tick();
+        this.profiler.pop();
+        // Blueberry end
         this.profiler.push("tallying");
         long l2 = this.tickTimes[this.tickCount % 100] = Util.getNanos() - l;
         this.averageTickTime = this.averageTickTime * 0.8F + (float)l2 / 1000000.0F * 0.19999999F;
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 804cd72fe4e39bda42460a4ff77f31440bb8b297..c924aa21e3030ea9f3e0a11c3c4fe5a67c769860 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -497,6 +497,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
     }
 
     public void stopServer() {
+        net.blueberrymc.common.Blueberry.shutdown(); // Blueberry
         super.stopServer();
         Util.shutdownExecutors();
         SkullBlockEntity.clear();
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index f852c606b306cc0b727775e5b1b021887aa48e89..7cf6600d006c9d8413a6ff67d2dc1b82598dbf0c 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -181,7 +181,7 @@ import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.slf4j.Logger;
 
-public class ServerGamePacketListenerImpl implements ServerPlayerConnection, TickablePacketListener, ServerGamePacketListener {
+public class ServerGamePacketListenerImpl implements ServerPlayerConnection, TickablePacketListener, ServerGamePacketListener, net.blueberrymc.network.server.ServerBlueberryPacketListener { // Blueberry
     static final Logger LOGGER = LogUtils.getLogger();
     private static final int LATENCY_CHECK_INTERVAL = 15000;
     public static final double MAX_INTERACTION_DISTANCE = Mth.square(6.0D);
@@ -1653,6 +1653,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     }
 
     public void handleCustomPayload(ServerboundCustomPayloadPacket serverboundCustomPayloadPacket) {
+        if (net.blueberrymc.network.server.ServerBlueberryPacketListener.super.handleBlueberryCustomPayload(serverboundCustomPayloadPacket)) return; // Blueberry
     }
 
     public void handleChangeDifficulty(ServerboundChangeDifficultyPacket serverboundChangeDifficultyPacket) {
diff --git a/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
index f228eb5e7e177214baf08d7e82760723f9e8277d..7d986dbb3b3f98914aaf8ba56021b6b45772ae15 100644
--- a/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -33,6 +33,7 @@ public class ServerStatusPacketListenerImpl implements ServerStatusPacketListene
         } else {
             this.hasRequestedStatus = true;
             this.connection.send(new ClientboundStatusResponsePacket(this.server.getStatus()));
+            this.connection.send(new net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket(net.blueberrymc.common.Blueberry.getModLoader().getModInfos())); // Blueberry
         }
     }
 
diff --git a/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java b/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java
index db249dc4ac6eb34f54054dc186f71b44dc0b39bf..856e539ef71747fcb5cb91c16fbde61ae1bb7fa4 100644
--- a/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java
+++ b/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java
@@ -39,6 +39,8 @@ public abstract class BuiltInPackSource implements RepositorySource {
         }
 
         this.listBundledPacks(consumer);
+
+        net.blueberrymc.common.Blueberry.getModLoader().loadPacks(consumer); // Blueberry
     }
 
     @Nullable
diff --git a/src/main/java/net/minecraft/world/inventory/MenuType.java b/src/main/java/net/minecraft/world/inventory/MenuType.java
index d4744adaae53a9a832a1733a57b05aacc4d48531..f4c2e17bdf7a71efa6d1b50e202864c846dec5da 100644
--- a/src/main/java/net/minecraft/world/inventory/MenuType.java
+++ b/src/main/java/net/minecraft/world/inventory/MenuType.java
@@ -35,7 +35,8 @@ public class MenuType<T extends AbstractContainerMenu> {
         return Registry.register(BuiltInRegistries.MENU, s, new MenuType<>(menuSupplier));
     }
 
-    private MenuType(MenuType.MenuSupplier<T> menuSupplier) {
+    // Blueberry - private -> public
+    public MenuType(MenuType.MenuSupplier<T> menuSupplier) {
         this.constructor = menuSupplier;
     }
 
@@ -43,6 +44,7 @@ public class MenuType<T extends AbstractContainerMenu> {
         return this.constructor.create(i, inventory);
     }
 
+    public // Blueberry
     interface MenuSupplier<T extends AbstractContainerMenu> {
         T create(int var1, Inventory var2);
     }
diff --git a/src/main/java/net/minecraft/world/item/Item.java b/src/main/java/net/minecraft/world/item/Item.java
index 16a48c059aabce692e9bc75e445bff1915c574f5..0f35341d7c59e7c3b759809fdfa02bc911b2611e 100644
--- a/src/main/java/net/minecraft/world/item/Item.java
+++ b/src/main/java/net/minecraft/world/item/Item.java
@@ -99,6 +99,7 @@ public class Item implements FeatureElement, ItemLike {
             }
         }
 
+        this.renderer = properties.renderer; // Blueberry
     }
 
     /** @deprecated */
@@ -370,6 +371,15 @@ public class Item implements FeatureElement, ItemLike {
         return this.requiredFeatures;
     }
 
+    // Blueberry start
+    private final java.util.function.Supplier<net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer> renderer;
+
+    public final net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer getRenderer() {
+        net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer renderer = this.renderer != null ? this.renderer.get() : null;
+        return renderer != null ? renderer : net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer.instance;
+    }
+    // Blueberry end
+
     public static class Properties {
         int maxStackSize = 64;
         int maxDamage;
@@ -380,6 +390,7 @@ public class Item implements FeatureElement, ItemLike {
         FoodProperties foodProperties;
         boolean isFireResistant;
         FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
+        java.util.function.Supplier<net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer> renderer; // Blueberry
 
         public Item.Properties food(FoodProperties foodProperties) {
             this.foodProperties = foodProperties;
@@ -424,5 +435,12 @@ public class Item implements FeatureElement, ItemLike {
             this.requiredFeatures = FeatureFlags.REGISTRY.subset(featureFlags);
             return this;
         }
+
+        // Blueberry start
+        public Item.Properties renderer(java.util.function.Supplier<net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer> renderer) {
+            this.renderer = renderer;
+            return this;
+        }
+        // Blueberry end
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java
index eafd47f72f0a581d559229fb74f1618ab1637cfe..390d3afb1905ac3d6fcdcb0e28cff129b0e908a3 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -98,6 +98,7 @@ public class BlockEntityType<T extends BlockEntity> {
     }
 
     @FunctionalInterface
+    public // Blueberry
     interface BlockEntitySupplier<T extends BlockEntity> {
         T create(BlockPos var1, BlockState var2);
     }
diff --git a/src/main/resources/log4j2-debug.xml b/src/main/resources/log4j2-debug.xml
index aea01980678d1a4f74982b0c08f42456d557b124..cabc65a1a5255cb445e4e2ac59f83b759eea9979 100644
--- a/src/main/resources/log4j2-debug.xml
+++ b/src/main/resources/log4j2-debug.xml
@@ -2,13 +2,13 @@
 <Configuration status="WARN" packages="com.mojang.util">
     <Appenders>
         <Console name="SysOut" target="SYSTEM_OUT">
-            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
         </Console>
         <Queue name="ServerGuiConsole">
-            <PatternLayout pattern="[%d{HH:mm:ss} %level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss} %level] [%c]: %msg%n" />
         </Queue>
         <RollingRandomAccessFile name="File" fileName="logs/latest.log" filePattern="logs/%d{yyyy-MM-dd}-%i.log.gz">
-            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level]: %msg%n" />
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
             <Policies>
                 <TimeBasedTriggeringPolicy />
                 <OnStartupTriggeringPolicy />
diff --git a/src/main/resources/log4j2.xml b/src/main/resources/log4j2.xml
new file mode 100644
index 0000000000000000000000000000000000000000..8b00a0d606dae68533536e2eb8a6b10030a9e098
--- /dev/null
+++ b/src/main/resources/log4j2.xml
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="WARN">
+    <Appenders>
+        <Console name="SysOut" target="SYSTEM_OUT">
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
+        </Console>
+        <Queue name="ServerGuiConsole">
+            <PatternLayout pattern="[%d{HH:mm:ss} %level] [%c]: %msg%n" />
+        </Queue>
+        <RollingRandomAccessFile name="File" fileName="logs/latest.log" filePattern="logs/%d{yyyy-MM-dd}-%i.log.gz">
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
+            <Policies>
+                <TimeBasedTriggeringPolicy />
+                <OnStartupTriggeringPolicy />
+            </Policies>
+        </RollingRandomAccessFile>
+    </Appenders>
+    <Loggers>
+        <Root level="info">
+            <filters>
+                <MarkerFilter marker="NETWORK_PACKETS" onMatch="DENY" onMismatch="NEUTRAL" />
+            </filters>
+            <AppenderRef ref="SysOut"/>
+            <AppenderRef ref="File"/>
+            <AppenderRef ref="ServerGuiConsole"/>
+        </Root>
+    </Loggers>
+</Configuration>
