From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 Jan 2021 19:48:53 +0900
Subject: [PATCH] Client-side command support


diff --git a/src/main/java/net/minecraft/client/player/LocalPlayer.java b/src/main/java/net/minecraft/client/player/LocalPlayer.java
index c39a723d08d567bf6cc6e60faef13bb83fa08a4c..c69267a322a5fac34eb2c3f5a9ebfeb19cf9a6e7 100644
--- a/src/main/java/net/minecraft/client/player/LocalPlayer.java
+++ b/src/main/java/net/minecraft/client/player/LocalPlayer.java
@@ -268,6 +268,7 @@ public class LocalPlayer extends AbstractClientPlayer {
    }
 
    public void chat(String s) {
+      if (net.blueberrymc.client.commands.ClientCommandManager.hasCommand(s)) net.blueberrymc.client.commands.ClientCommandManager.performCommand(this.createCommandSourceStack(), s); // Blueberry
       this.connection.send(new ServerboundChatPacket(s));
    }
 
