From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 3 Mar 2021 11:59:03 +0900
Subject: [PATCH] CommandRegistrationEvent to register commands easily


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 2eb447b33b70a043c6fa22e9343396a7c105b11f..7b103f6cbec47145bfa415687b2b45f627b794b0 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -173,6 +173,7 @@ public class Commands {
         TriggerCommand.register(this.dispatcher);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        net.blueberrymc.common.event.EventFactory.callCommandRegistrationEvent(this.dispatcher, commandSelection); // Blueberry
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
