From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <ilyy@outlook.jp>
Date: Fri, 26 Feb 2021 09:22:23 +0900
Subject: [PATCH] Discord RPC Support


diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index bac6905be23e7359a388618ec0a1b109e6178640..317b080ec094f6025fb3439752384cbe450b8890 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -684,6 +684,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
       File file = new File(instance == null ? new File(".") : getInstance().gameDirectory, "crash-reports"); // Blueberry
       File file2 = new File(file, "crash-" + (new SimpleDateFormat("yyyy-MM-dd_HH.mm.ss")).format(new Date()) + "-client.txt");
       Bootstrap.realStdoutPrintln(crashReport.getFriendlyReport());
+      net.blueberrymc.common.util.DiscordRPCTaskExecutor.shutdownNow(); // Blueberry
       if (crashReport.getSaveFile() != null) {
          Bootstrap.realStdoutPrintln("#@!@# Game crashed! Crash report saved to: #@!@# " + crashReport.getSaveFile());
          System.exit(-1);
@@ -808,6 +809,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
       }
 
       this.screen = screen;
+      net.blueberrymc.client.event.ClientEventFactory.callScreenChangedEvent(this.screen); // Blueberry
       if (screen != null) {
          this.mouseHandler.releaseMouse();
          KeyMapping.releaseAll();
@@ -824,9 +826,11 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 
    public void setOverlay(@Nullable Overlay overlay) {
       this.overlay = overlay;
+      net.blueberrymc.client.event.ClientEventFactory.callOverlayChangedEvent(overlay); // Blueberry
    }
 
    public void destroy() {
+      net.blueberrymc.common.util.DiscordRPCTaskExecutor.shutdownNow(); // Blueberry
       try {
          LOGGER.info("Stopping!");
 
diff --git a/src/main/java/net/minecraft/client/main/Main.java b/src/main/java/net/minecraft/client/main/Main.java
index 420cb78609ca13b7e6c97edb6ad5f3cfb0543271..914bb3da9d50eb56744e81d0776f0828ed95d94b 100644
--- a/src/main/java/net/minecraft/client/main/Main.java
+++ b/src/main/java/net/minecraft/client/main/Main.java
@@ -34,8 +34,18 @@ import org.apache.logging.log4j.Logger;
 public class Main {
    private static final Logger LOGGER = LogManager.getLogger();
 
-   // MagmaCube start - decompile error
+   // Blueberry start
    public static void main(String[] args) {
+      try {
+         main0(args);
+      } finally {
+         net.blueberrymc.common.util.DiscordRPCTaskExecutor.shutdown();
+      }
+   }
+   // Blueberry end
+
+   // MagmaCube start - decompile error
+   public static void main0(String[] args) { // Blueberry
       if (net.minecraft.SharedConstants.IS_RUNNING_IN_IDE) org.lwjgl.glfw.GLFW.glfwInit(); // MagmaCube - lwjgl debug agent causes glGetError to crash when we didn't call glfwInit
       OptionParser optionparser = new OptionParser();
       optionparser.allowsUnrecognizedOptions();
