From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 15 Jan 2021 19:48:53 +0900
Subject: [PATCH] Client-side command support


diff --git a/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java b/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java
index 93fc0e72637df24b24c526583d9b7d8cfc750bff..94291d5dddb394e939b3e8688540cb1bfc855248 100644
--- a/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java
+++ b/src/main/java/net/minecraft/client/gui/components/CommandSuggestions.java
@@ -172,7 +172,13 @@ public class CommandSuggestions {
         boolean flag2 = this.commandsOnly || flag;
         int i = this.input.getCursorPosition();
         if (flag2) {
-            CommandDispatcher<SharedSuggestionProvider> commandDispatcher = this.minecraft.player.connection.getCommands();
+            // Blueberry start - client-side command completion
+            CommandDispatcher<SharedSuggestionProvider> commandDispatcher = net.blueberrymc.client.commands.ClientCommandManager.getRoot(this.minecraft.player);
+            CommandDispatcher<SharedSuggestionProvider> mojangCommandDispatcher = this.minecraft.player.connection.getCommands();
+            for (CommandNode<SharedSuggestionProvider> child : mojangCommandDispatcher.getRoot().getChildren()) {
+                commandDispatcher.getRoot().addChild(child);
+            }
+            // Blueberry end - client-side command completion
             if (this.currentParse == null) {
                 this.currentParse = commandDispatcher.parse(stringReader, this.minecraft.player.connection.getSuggestionsProvider());
             }
diff --git a/src/main/java/net/minecraft/client/gui/screens/ChatScreen.java b/src/main/java/net/minecraft/client/gui/screens/ChatScreen.java
index d7d57ec4dd16e8e9a6a570aaa1526295e5d856f8..3b8fb37505494b016748719a1abc9e4d8b3a7e4f 100644
--- a/src/main/java/net/minecraft/client/gui/screens/ChatScreen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/ChatScreen.java
@@ -218,6 +218,12 @@ public class ChatScreen extends Screen {
             }
 
             if (s.startsWith("/")) {
+                // Blueberry start
+                if (net.blueberrymc.client.commands.ClientCommandManager.hasCommand(s)) {
+                    net.blueberrymc.client.commands.ClientCommandManager.performCommand(this.minecraft.player.createCommandSourceStack(), s);
+                    return true;
+                }
+                // Blueberry end
                 this.minecraft.player.connection.sendCommand(s.substring(1));
             } else {
                 this.minecraft.player.connection.sendChat(s);
