From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 14 Jan 2021 01:44:46 +0900
Subject: [PATCH] Mods support


diff --git a/src/main/java/com/mojang/blaze3d/platform/Window.java b/src/main/java/com/mojang/blaze3d/platform/Window.java
index d68b1c0e36835cceed57cfd7ecb3c4c757fa8047..5122a248127d3e5e1f81e183cd6d1386d4be8023 100644
--- a/src/main/java/com/mojang/blaze3d/platform/Window.java
+++ b/src/main/java/com/mojang/blaze3d/platform/Window.java
@@ -82,7 +82,7 @@ public final class Window implements AutoCloseable {
 		GLFW.glfwWindowHint(139267, 2);
 		GLFW.glfwWindowHint(139272, 204801);
 		GLFW.glfwWindowHint(139270, 1);
-		this.window = GLFW.glfwCreateWindow(this.width, this.height, string2, this.fullscreen && monitor != null ? monitor.getMonitor() : 0L, 0L);
+        this.window = net.blueberrymc.client.EarlyLoadingScreen.getInstance().acquireWindowOrGet(() -> GLFW.glfwCreateWindow(this.width, this.height, string2, this.fullscreen && monitor != null ? monitor.getMonitor() : 0L, 0L));
 		if (monitor != null) {
 			VideoMode videoMode = monitor.getPreferredVidMode(this.fullscreen ? this.preferredFullscreenVideoMode : Optional.empty());
 			this.windowedX = this.x = monitor.getX() + videoMode.getWidth() / 2 - this.width / 2;
diff --git a/src/main/java/net/blueberrymc/client/BlueberryClientImpl.java b/src/main/java/net/blueberrymc/client/BlueberryClientImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..82f560b47ec3def0c2ddf248a4005dfb28538ded
--- /dev/null
+++ b/src/main/java/net/blueberrymc/client/BlueberryClientImpl.java
@@ -0,0 +1,21 @@
+package net.blueberrymc.client;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.screens.MenuScreens;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.gui.screens.inventory.MenuAccess;
+import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.inventory.MenuType;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import org.jetbrains.annotations.NotNull;
+
+public class BlueberryClientImpl extends BlueberryClient {
+    public void registerSpecialBlockEntityRenderer(@NotNull BlockEntityType<?> blockEntityType, @NotNull BlockEntityRenderer<?> blockEntityRenderer) {
+        Minecraft.getInstance().getBlockEntityRenderDispatcher().registerSpecialRenderer(blockEntityType, blockEntityRenderer);
+    }
+
+    public <M extends AbstractContainerMenu, U extends Screen & MenuAccess<M>> void registerMenuScreensFactory(@NotNull MenuType<? extends M> menuType, @NotNull ScreenConstructor<M, U> screenConstructor) {
+        MenuScreens.register(menuType, screenConstructor::create);
+    }
+}
diff --git a/src/main/java/net/blueberrymc/server/BlueberryServerImpl.java b/src/main/java/net/blueberrymc/server/BlueberryServerImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..7b1e2686690e9d7df32131010891223b24eb568f
--- /dev/null
+++ b/src/main/java/net/blueberrymc/server/BlueberryServerImpl.java
@@ -0,0 +1,4 @@
+package net.blueberrymc.server;
+
+public class BlueberryServerImpl extends BlueberryServer {
+}
diff --git a/src/main/java/net/minecraft/SystemReport.java b/src/main/java/net/minecraft/SystemReport.java
index 96aca9627fa13af5491f482b93e03285ff5a21a2..6069e177b571e2dfbe182dac486b353a07c01d32 100644
--- a/src/main/java/net/minecraft/SystemReport.java
+++ b/src/main/java/net/minecraft/SystemReport.java
@@ -56,6 +56,26 @@ public class SystemReport {
 			List<String> list = (List<String>)Util.getVmArguments().collect(Collectors.toList());
 			return String.format(Locale.ROOT, "%d total; %s", list.size(), String.join(" ", list));
 		}));
+        // Blueberry start
+        this.setDetail("Blueberry Version", net.blueberrymc.common.util.Versioning.getVersion().getFullyQualifiedVersion());
+        this.setDetail("Blueberry commit", net.blueberrymc.common.util.Versioning.getVersion().getCommit());
+        // TODO: we need better layout
+        this.setDetail("Mods", () -> {
+            StringBuilder sb = new StringBuilder("\n");
+            sb.append("      Status:\n");
+            sb.append("        L = Loaded\n");
+            sb.append("        P = Pre Init\n");
+            sb.append("        I = Init\n");
+            sb.append("        J = Post Init\n");
+            sb.append("        A = Available\n");
+            sb.append("        E = Errored\n");
+            sb.append("        U = Unloaded\n");
+            for (net.blueberrymc.common.bml.BlueberryMod mod : net.blueberrymc.common.Blueberry.getModLoader().getLoadedMods()) {
+                sb.append("      ").append(mod.getName()).append(" (").append(mod.getDescription().getModId()).append(") [").append(mod.getDescription().getVersion()).append("] - ").append(mod.getStateList().toString()).append("\n");
+            }
+            return sb.toString();
+        });
+        // Blueberry end
 	}
 
 	public void setDetail(String string, String string2) {
diff --git a/src/main/java/net/minecraft/client/Minecraft.java b/src/main/java/net/minecraft/client/Minecraft.java
index 0c1d00140bdfebc8a65735f686332906d16dbac1..7af799e68d141d02fb379e02bd912a258dfc609a 100644
--- a/src/main/java/net/minecraft/client/Minecraft.java
+++ b/src/main/java/net/minecraft/client/Minecraft.java
@@ -485,7 +485,9 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 		this.mainRenderTarget = new MainTarget(this.window.getWidth(), this.window.getHeight());
 		this.mainRenderTarget.setClearColor(0.0F, 0.0F, 0.0F, 0.0F);
 		this.mainRenderTarget.clear(ON_OSX);
+        net.blueberrymc.client.EarlyLoadingScreen.getInstance().startRender(false); // Blueberry
 		this.resourceManager = new ReloadableResourceManager(PackType.CLIENT_RESOURCES);
+        net.blueberrymc.common.Blueberry.getModLoader().callPreInit(); // Blueberry
 		this.resourcePackRepository.reload();
 		this.options.loadSelectedResourcePacks(this.resourcePackRepository);
 		this.languageManager = new LanguageManager(this.options.languageCode);
@@ -500,6 +502,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 		this.splashManager = new SplashManager(this.user);
 		this.resourceManager.registerReloadListener(this.splashManager);
 		this.musicManager = new MusicManager(this);
+        net.blueberrymc.client.EarlyLoadingScreen.getInstance().blockUntilFinish(); // Blueberry
 		this.fontManager = new FontManager(this.textureManager);
 		this.font = this.fontManager.createFont();
 		this.fontFilterFishy = this.fontManager.createFontFilterFishy();
@@ -521,6 +524,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 		);
 		this.resourceManager.registerReloadListener(this.blockEntityRenderDispatcher);
 		BlockEntityWithoutLevelRenderer blockEntityWithoutLevelRenderer = new BlockEntityWithoutLevelRenderer(this.blockEntityRenderDispatcher, this.entityModels);
+        BlockEntityWithoutLevelRenderer.instance = blockEntityWithoutLevelRenderer; // Blueberry
 		this.resourceManager.registerReloadListener(blockEntityWithoutLevelRenderer);
 		this.itemRenderer = new ItemRenderer(this, this.textureManager, this.modelManager, this.itemColors, blockEntityWithoutLevelRenderer);
 		this.resourceManager.registerReloadListener(this.itemRenderer);
@@ -605,6 +609,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 		LoadingOverlay.registerTextures(this);
 		this.setScreen(new GenericMessageScreen(Component.translatable("gui.loadingMinecraft")));
 		List<PackResources> list = this.resourcePackRepository.openAllSelected();
+        net.blueberrymc.common.Blueberry.getModLoader().callInit(); // Blueberry
 		this.reloadStateTracker.startReload(ResourceLoadStateTracker.ReloadReason.INITIAL, list);
 		ReloadInstance reloadInstance = this.resourceManager.createReload(Util.backgroundExecutor(), this, RESOURCE_RELOAD_INITIAL_TASK, list);
 		GameLoadTimesEvent.INSTANCE.beginStep(TelemetryProperty.LOAD_TIME_LOADING_OVERLAY_MS);
@@ -648,7 +653,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 			if (gameLoadCookie != null && gameLoadCookie.quickPlayData().isEnabled()) {
 				QuickPlay.connect(this, gameLoadCookie.quickPlayData(), gameLoadCookie.realmsClient());
 			} else {
-				this.setScreen(new TitleScreen(true));
+                this.setScreen(new net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen(new TitleScreen(/*true*/))); // Blueberry
 			}
 		};
 
@@ -688,6 +693,11 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 				list.add(BanNoticeScreens::createSkinBan);
 			}
 		}
+        // Blueberry start
+        if (net.blueberrymc.common.bml.loading.ModLoadingErrors.hasErrorOrWarning()) {
+            list.add(net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen::new);
+        }
+        // Blueberry end
 	}
 
 	private static boolean countryEqualsISO3(Object object) {
@@ -904,6 +914,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 		File file2 = new File(file, "crash-reports");
 		File file3 = new File(file2, "crash-" + Util.getFilenameFormattedDateTime() + "-client.txt");
 		Bootstrap.realStdoutPrintln(crashReport.getFriendlyReport());
+        net.blueberrymc.common.util.DiscordRPCTaskExecutor.shutdownNow(); // Blueberry
 		if (minecraft != null) {
 			minecraft.soundManager.emergencyShutdown();
 		}
@@ -1070,6 +1081,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 			}
 
 			BufferUploader.reset();
+            net.blueberrymc.client.event.ClientEventFactory.callScreenChangedEvent(this.screen); // Blueberry
 			if (screen != null) {
 				this.mouseHandler.releaseMouse();
 				KeyMapping.releaseAll();
@@ -1086,6 +1098,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 
 	public void setOverlay(@Nullable Overlay overlay) {
 		this.overlay = overlay;
+        net.blueberrymc.client.event.ClientEventFactory.callOverlayChangedEvent(overlay); // Blueberry
 	}
 
 	public void destroy() {
@@ -1110,6 +1123,7 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 				this.screen.removed();
 			}
 
+            net.blueberrymc.common.Blueberry.shutdown();
 			this.close();
 		} finally {
 			Util.timeSource = System::nanoTime;
@@ -1917,6 +1931,10 @@ public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements
 			this.pendingConnection.tick();
 		}
 
+        // Blueberry start
+        this.profiler.popPush("blueberryClientScheduler");
+        net.blueberrymc.common.Blueberry.getUtil().getClientScheduler().tick();
+        // Blueberry end
 		this.profiler.popPush("keyboard");
 		this.keyboardHandler.tick();
 		this.profiler.pop();
diff --git a/src/main/java/net/minecraft/client/gui/components/ChatComponent.java b/src/main/java/net/minecraft/client/gui/components/ChatComponent.java
index 282c08851e754302239182a22c1d4755391de8a6..55f2953ef67546f9838cde2cae0f452d393ac461 100644
--- a/src/main/java/net/minecraft/client/gui/components/ChatComponent.java
+++ b/src/main/java/net/minecraft/client/gui/components/ChatComponent.java
@@ -447,15 +447,27 @@ public class ChatComponent {
 	}
 
 	public static int getWidth(double d) {
-		int i = 320;
-		int j = 40;
-		return Mth.floor(d * 280.0 + 40.0);
+        // Blueberry start
+        int width = 320;
+        if (net.blueberrymc.common.bml.InternalBlueberryModConfig.Misc.ChatSettings.extendedWidth && Minecraft.getInstance().getWindow() != null) {
+            width = Minecraft.getInstance().getWindow().getWidth() / 2 - 8;
+        }
+        // Blueberry end
+        int i = 320;
+        int i2 = 40;
+        return Mth.floor(d * width); // Blueberry
 	}
 
 	public static int getHeight(double d) {
-		int i = 180;
-		int j = 20;
-		return Mth.floor(d * 160.0 + 20.0);
+        // Blueberry start
+        int height = 180;
+        if (net.blueberrymc.common.bml.InternalBlueberryModConfig.Misc.ChatSettings.extendedHeight && Minecraft.getInstance().getWindow() != null) {
+            height = Minecraft.getInstance().getWindow().getHeight() / 2 - 40;
+        }
+        // Blueberry end
+        int i = 180;
+        int i2 = 20;
+        return Mth.floor(d * height); // Blueberry
 	}
 
 	public static double defaultUnfocusedPct() {
diff --git a/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java b/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
index 50e3df6b74e84962a18908aeae4825b3120dbc06..0fb2a102e19f0760ec33f8f6c337695907e11be7 100644
--- a/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
+++ b/src/main/java/net/minecraft/client/gui/components/DebugScreenOverlay.java
@@ -70,6 +70,7 @@ import net.minecraft.world.level.levelgen.RandomState;
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
+import net.blueberrymc.common.Blueberry; // Blueberry
 import org.jetbrains.annotations.Nullable;
 
 @Environment(EnvType.CLIENT)
@@ -498,6 +499,11 @@ public class DebugScreenOverlay {
 			GlUtil.getRenderer(),
 			GlUtil.getOpenGLVersion()
 		);
+        // Blueberry start
+        list.add("");
+        list.add("Blueberry " + Blueberry.getVersion().getFullyQualifiedVersion());
+        list.add(Blueberry.getModLoader().getLoadedMods().size() + " mods loaded, " + Blueberry.getModLoader().getActiveMods().size() + " mods active");
+        // Blueberry end
 		if (this.minecraft.showOnlyReducedInfo()) {
 			return list;
 		} else {
diff --git a/src/main/java/net/minecraft/client/gui/font/FontManager.java b/src/main/java/net/minecraft/client/gui/font/FontManager.java
index 19777b2f9a16fdf2f66cd33861e08ed5312ced27..3ee604d2a242a5164dbf585d63dd1958e0b1971c 100644
--- a/src/main/java/net/minecraft/client/gui/font/FontManager.java
+++ b/src/main/java/net/minecraft/client/gui/font/FontManager.java
@@ -211,6 +211,7 @@ public class FontManager implements PreparableReloadListener, AutoCloseable {
 		if (!this.fontSets.containsKey(Minecraft.DEFAULT_FONT)) {
 			throw new IllegalStateException("Default font failed to load");
 		}
+        net.minecraft.client.gui.screens.LoadingOverlay.isFontReady = true; // Blueberry
 	}
 
 	public void updateOptions(Options options) {
diff --git a/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java b/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
index 9399d6c675e182419cd78d987382354a0e5d2218..8100974f2ca05c8aa74184e23afce94248ee7f26 100644
--- a/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/src/main/java/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -28,6 +28,7 @@ import net.minecraft.util.Mth;
 
 @Environment(EnvType.CLIENT)
 public class LoadingOverlay extends Overlay {
+    public static boolean isFontReady = false; // Blueberry
 	static final ResourceLocation MOJANG_STUDIOS_LOGO_LOCATION = new ResourceLocation("textures/gui/title/mojangstudios.png");
 	private static final int LOGO_BACKGROUND_COLOR = FastColor.ARGB32.color(255, 239, 50, 61);
 	private static final int LOGO_BACKGROUND_COLOR_DARK = FastColor.ARGB32.color(255, 0, 0, 0);
@@ -122,6 +123,7 @@ public class LoadingOverlay extends Overlay {
 		RenderSystem.depthMask(true);
 		RenderSystem.enableDepthTest();
 		int v = (int)((double)guiGraphics.guiHeight() * 0.8325);
+        net.blueberrymc.client.EarlyLoadingScreen.getInstance().renderMessagesFromGUI(guiGraphics); // Blueberry
 		float w = this.reload.getActualProgress();
 		this.currentProgress = Mth.clamp(this.currentProgress * 0.95F + w * 0.050000012F, 0.0F, 1.0F);
 		if (g < 1.0F) {
@@ -129,6 +131,7 @@ public class LoadingOverlay extends Overlay {
 		}
 
 		if (g >= 2.0F) {
+            net.blueberrymc.common.Blueberry.getModLoader().callPostInit(); // Blueberry
 			this.minecraft.setOverlay(null);
 		}
 
@@ -142,6 +145,7 @@ public class LoadingOverlay extends Overlay {
 
 			this.fadeOutStart = Util.getMillis();
 			if (this.minecraft.screen != null) {
+                if (this.minecraft.screen instanceof net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen) ((net.blueberrymc.common.bml.client.gui.screens.ModLoadingProblemScreen) minecraft.screen).refresh(); // Blueberry
 				this.minecraft.screen.init(this.minecraft, guiGraphics.guiWidth(), guiGraphics.guiHeight());
 			}
 		}
diff --git a/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java b/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
index f2866234a3a3964b5437fe5fc8b444da395639e6..4b30d5e51b5bcf9b5f4967bead3adc71c509c7ea 100644
--- a/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/PauseScreen.java
@@ -95,6 +95,7 @@ public class PauseScreen extends Screen {
 		gridLayout.arrangeElements();
 		FrameLayout.alignInRectangle(gridLayout, 0, 0, this.width, this.height, 0.5F, 0.25F);
 		gridLayout.visitWidgets(this::addRenderableWidget);
+        this.addRenderableWidget(Button.builder(net.blueberrymc.common.resources.BlueberryText.text("blueberry", "gui.screens.mods"), (buttonx) -> net.blueberrymc.client.gui.screens.ModListScreen.switchToModListScreen()).bounds(this.width / 2 - 102, this.height / 4 + 144 - 16, 204, 20).build()); // Blueberry (TODO)
 	}
 
 	private void onDisconnect() {
diff --git a/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java b/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
index bff59290b676b800fbbc889b706b83e6422b796a..73901a1697845c78add0de3090b56d9b2d6c1156 100644
--- a/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/src/main/java/net/minecraft/client/gui/screens/TitleScreen.java
@@ -40,6 +40,8 @@ import net.minecraft.util.Mth;
 import net.minecraft.world.level.levelgen.WorldOptions;
 import net.minecraft.world.level.levelgen.presets.WorldPresets;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import net.blueberrymc.client.gui.screens.ModListScreen; // Blueberry
+import net.blueberrymc.common.resources.BlueberryText; // Blueberry
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -165,13 +167,10 @@ public class TitleScreen extends Screen {
 			Screen screen = (Screen)(this.minecraft.options.skipMultiplayerWarning ? new JoinMultiplayerScreen(this) : new SafetyScreen(this));
 			this.minecraft.setScreen(screen);
 		}).bounds(this.width / 2 - 100, i + j * 1, 200, 20).tooltip(tooltip).build()).active = bl;
-		this.addRenderableWidget(
-				Button.builder(Component.translatable("menu.online"), button -> this.minecraft.setScreen(new RealmsMainScreen(this)))
-					.bounds(this.width / 2 - 100, i + j * 2, 200, 20)
-					.tooltip(tooltip)
-					.build()
-			)
-			.active = bl;
+        // Blueberry start
+        (this.addRenderableWidget(Button.builder(BlueberryText.text("blueberry", "gui.screens.mods"), (button) -> ModListScreen.switchToModListScreen()).bounds(this.width / 2 - 100, i + j * 2, 98, 20).tooltip(tooltip).build())).active = bl;
+        (this.addRenderableWidget(Button.builder(Component.translatable("menu.online"), (button) -> this.minecraft.setScreen(new RealmsMainScreen(this))).bounds(this.width / 2 + 2, i + j * 2, 98, 20).tooltip(tooltip).build())).active = bl;
+        // Blueberry end
 	}
 
 	@Nullable
@@ -301,7 +300,12 @@ public class TitleScreen extends Screen {
 				string = string + I18n.get("menu.modded");
 			}
 
-			guiGraphics.drawString(this.font, string, 2, this.height - 10, 16777215 | k);
+            // Blueberry start
+            guiGraphics.drawString(this.font, "Blueberry " + net.blueberrymc.common.Blueberry.getVersion().getFullyQualifiedVersion(), 2, this.height - 40, 0xFFFFFF | k);
+            guiGraphics.drawString(this.font, "MagmaCube " + net.blueberrymc.common.util.Versioning.getVersion().getShortMagmaCubeCommit(), 2, this.height - 30, 0xFFFFFF | k); // MagmaCube // Blueberry - offset 20 -> 30
+            guiGraphics.drawString(this.font, string, 2, this.height - 20, 16777215 | k); // offset 10 -> 20
+            guiGraphics.drawString(this.font, net.blueberrymc.common.Blueberry.getModLoader().getLoadedMods().size() + " mods loaded, " + net.blueberrymc.common.Blueberry.getModLoader().getActiveMods().size() + " mods active", 2, this.height - 10, 0xFFFFFF | k);
+            // Blueberry end
 			if (this.realmsNotificationsEnabled() && g >= 1.0F) {
 				RenderSystem.enableDepthTest();
 				this.realmsNotificationsScreen.render(guiGraphics, i, j, f);
diff --git a/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java b/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
index c4f7ad8de9b05716ce9782e15ec24123f5b7c167..f25c75bef47f6f2f2908d6f69a7e73d94fabe9d7 100644
--- a/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
+++ b/src/main/java/net/minecraft/client/gui/screens/multiplayer/ServerSelectionList.java
@@ -312,6 +312,13 @@ public class ServerSelectionList extends ObjectSelectionList<ServerSelectionList
 				guiGraphics.blitSprite(this.statusIcon, p, j, 10, 8);
 			}
 
+			// Blueberry start
+			if (serverData.serverType != null) {
+//                RenderSystem.setShader(net.minecraft.client.renderer.GameRenderer::getPositionTexColorShader);
+				// TODO: p?
+				guiGraphics.blit(net.blueberrymc.client.gui.BlueberryGuiComponents.GUI_ICONS_LOCATION, k + l - 18, j + 8, (float) 0.0F, (float) (serverData.serverType.getOffset() * 16), 16, 16, 256, 256);
+			}
+			// Blueberry end
 			byte[] bs = this.serverData.getIconBytes();
 			if (!Arrays.equals(bs, this.lastIconBytes)) {
 				if (this.uploadServerIcon(bs)) {
@@ -330,6 +337,12 @@ public class ServerSelectionList extends ObjectSelectionList<ServerSelectionList
 			guiGraphics.drawString(this.minecraft.font, component, r, j + 1, -8355712, false);
 			if (this.statusIconTooltip != null && n >= p && n <= p + 10 && o >= j && o <= j + 8) {
 				this.screen.setTooltipForNextRenderPass(this.statusIconTooltip);
+                // Blueberry start
+            } else if (serverData.serverType != null && n >= p - 18 && n <= p - 5 && o >= 0 && o <= 22) {
+                String text = serverData.serverType.getBlueberryText().getContents();
+                if (text.contains("%d")) text = String.format(text, serverData.modsCount);
+                this.screen.setTooltipForNextRenderPass(java.util.Collections.singletonList((FormattedCharSequence) Component.literal(text)));
+                // Blueberry end
 			} else if (this.onlinePlayersTooltip != null && n >= r && n <= r + q && o >= j && o <= j - 1 + 9) {
 				this.screen.setTooltipForNextRenderPass(Lists.transform(this.onlinePlayersTooltip, Component::getVisualOrderText));
 			}
diff --git a/src/main/java/net/minecraft/client/main/Main.java b/src/main/java/net/minecraft/client/main/Main.java
index 251c4dda28c48b8c38480e44a8f3ee19268b0e1f..e4e864b97d6de2df587703de9a47f28e0f349b5e 100644
--- a/src/main/java/net/minecraft/client/main/Main.java
+++ b/src/main/java/net/minecraft/client/main/Main.java
@@ -57,8 +57,8 @@ public class Main {
 		Stopwatch stopwatch2 = Stopwatch.createStarted(Ticker.systemTicker());
 		GameLoadTimesEvent.INSTANCE.beginStep(TelemetryProperty.LOAD_TIME_TOTAL_TIME_MS, stopwatch);
 		GameLoadTimesEvent.INSTANCE.beginStep(TelemetryProperty.LOAD_TIME_PRE_WINDOW_MS, stopwatch2);
-		SharedConstants.tryDetectVersion();
-		SharedConstants.enableDataFixerOptimizations();
+//		SharedConstants.tryDetectVersion();
+//		SharedConstants.enableDataFixerOptimizations();
 		OptionParser optionParser = new OptionParser();
 		optionParser.allowsUnrecognizedOptions();
 		optionParser.accepts("demo");
@@ -96,6 +96,12 @@ public class Main {
 		OptionSpec<String> optionSpec28 = optionParser.nonOptions();
 		OptionSet optionSet = optionParser.parse(strings);
 		List<String> list = optionSet.valuesOf(optionSpec28);
+		// Blueberry start
+		net.blueberrymc.common.Blueberry.preBootstrap();
+		net.blueberrymc.common.Blueberry.bootstrap(new net.blueberrymc.server.BlueberryServerImpl());
+		SharedConstants.tryDetectVersion();
+		SharedConstants.enableDataFixerOptimizations();
+		// Blueberry end
 		if (!list.isEmpty()) {
 			LOGGER.info("Completely ignored arguments: {}", list);
 		}
diff --git a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
index 883d0705fddfdcc2b762138ee67e167e9cdb44ed..d7df20b40c4585b84658c830aef66e56ace0ac82 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -307,7 +307,7 @@ import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 
 @Environment(EnvType.CLIENT)
-public class ClientPacketListener extends ClientCommonPacketListenerImpl implements ClientGamePacketListener, TickablePacketListener {
+public class ClientPacketListener extends ClientCommonPacketListenerImpl implements ClientGamePacketListener, TickablePacketListener, net.blueberrymc.network.client.ClientBlueberryPacketListener { // Blueberry
 	private static final Logger LOGGER = LogUtils.getLogger();
 	private static final Component UNSECURE_SERVER_TOAST_TITLE = Component.translatable("multiplayer.unsecureserver.toast.title");
 	private static final Component UNSERURE_SERVER_TOAST = Component.translatable("multiplayer.unsecureserver.toast");
@@ -1954,6 +1954,7 @@ public class ClientPacketListener extends ClientCommonPacketListenerImpl impleme
 
 	@Override
 	public void handleCustomPayload(CustomPacketPayload customPacketPayload) {
+        if (net.blueberrymc.network.client.ClientBlueberryPacketListener.super.handleBlueberryCustomPayload(customPacketPayload)) return; // Blueberry
 		if (customPacketPayload instanceof PathfindingDebugPayload pathfindingDebugPayload) {
 			this.minecraft
 				.debugRenderer
diff --git a/src/main/java/net/minecraft/client/multiplayer/ServerData.java b/src/main/java/net/minecraft/client/multiplayer/ServerData.java
index dc02847e4570954e114502b7319c58bded293d3a..0444f9b7125e0679b1be1a605f4eaa9d1719dff7 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ServerData.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ServerData.java
@@ -34,6 +34,10 @@ public class ServerData {
 	private byte[] iconBytes;
 	private ServerData.Type type;
 	private ServerData.State state = ServerData.State.INITIAL;
+    // Blueberry start
+    public @Nullable net.blueberrymc.network.ServerType serverType = null;
+    public int modsCount = 0;
+    // Blueberry end
 
 	public ServerData(String string, String string2, ServerData.Type type) {
 		this.name = string;
diff --git a/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java b/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
index f96392e533af20d1afbd088ed9f754ef1c3d7c2d..62d86be0ba2e1f35c408f1a7bb0d320e869ff9ee 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ServerStatusPinger.java
@@ -65,6 +65,7 @@ public class ServerStatusPinger {
 						connection.disconnect(Component.translatable("multiplayer.status.unrequested"));
 					} else {
 						this.receivedPing = true;
+                        serverData.serverType = net.blueberrymc.network.ServerType.VANILLA; // Blueberry
 						ServerStatus serverStatus = clientboundStatusResponsePacket.status();
 						serverData.motd = serverStatus.description();
 						serverStatus.version().ifPresentOrElse(version -> {
@@ -122,7 +123,20 @@ public class ServerStatusPinger {
 					}
 				}
 
-				@Override
+                // Blueberry start
+                @Override
+                public void handleBlueberryHandshakeResponse(net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket packet) {
+                    boolean compatible = net.blueberrymc.common.util.ListUtils.isCompatible(packet.getModInfos(), net.blueberrymc.common.Blueberry.getModLoader().getModInfos());
+                    if (compatible) {
+                        serverData.serverType = net.blueberrymc.network.ServerType.BLUEBERRY_GOOD;
+                    } else {
+                        serverData.serverType = net.blueberrymc.network.ServerType.BLUEBERRY_BAD;
+                    }
+                    serverData.modsCount = packet.getModInfos().size();
+                }
+                // Blueberry end
+
+                @Override
 				public boolean isAcceptingMessages() {
 					return connection.isConnected();
 				}
diff --git a/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java b/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java
index 3446f9a5b6218cea338232c2e279f02146cc2057..5eb80bfd41781d1e9a03d95265713a16dacc11e0 100644
--- a/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/BlockEntityWithoutLevelRenderer.java
@@ -57,6 +57,7 @@ public class BlockEntityWithoutLevelRenderer implements ResourceManagerReloadLis
 		.map(dyeColor -> new ShulkerBoxBlockEntity(dyeColor, BlockPos.ZERO, Blocks.SHULKER_BOX.defaultBlockState()))
 		.toArray(ShulkerBoxBlockEntity[]::new);
 	private static final ShulkerBoxBlockEntity DEFAULT_SHULKER_BOX = new ShulkerBoxBlockEntity(BlockPos.ZERO, Blocks.SHULKER_BOX.defaultBlockState());
+    public static BlockEntityWithoutLevelRenderer instance; // Blueberry
 	private final ChestBlockEntity chest = new ChestBlockEntity(BlockPos.ZERO, Blocks.CHEST.defaultBlockState());
 	private final ChestBlockEntity trappedChest = new TrappedChestBlockEntity(BlockPos.ZERO, Blocks.TRAPPED_CHEST.defaultBlockState());
 	private final EnderChestBlockEntity enderChest = new EnderChestBlockEntity(BlockPos.ZERO, Blocks.ENDER_CHEST.defaultBlockState());
diff --git a/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java b/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
index f0b255adc2679638e5e84089a09433b2e4a4334b..8127c83a7edbe4504297bed0e9639f702e244514 100644
--- a/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
+++ b/src/main/java/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
@@ -123,7 +123,10 @@ public class BlockRenderDispatcher implements ResourceManagerReloadListener {
 						.renderModel(poseStack.last(), multiBufferSource.getBuffer(ItemBlockRenderTypes.getRenderType(blockState, false)), blockState, bakedModel, f, g, h, i, j);
 					break;
 				case ENTITYBLOCK_ANIMATED:
-					this.blockEntityRenderer.renderByItem(new ItemStack(blockState.getBlock()), ItemDisplayContext.NONE, poseStack, multiBufferSource, i, j);
+                    // Blueberry start
+                    ItemStack stack = new ItemStack(blockState.getBlock());
+                    stack.getItem().getRenderer().renderByItem(stack, ItemDisplayContext.NONE, poseStack, multiBufferSource, i, j);
+                    // Blueberry end
 			}
 		}
 	}
@@ -131,5 +134,6 @@ public class BlockRenderDispatcher implements ResourceManagerReloadListener {
 	@Override
 	public void onResourceManagerReload(ResourceManager resourceManager) {
 		this.liquidBlockRenderer.setupSprites();
+        net.blueberrymc.client.world.level.fluid.FluidSpriteManager.setupSprites(); // Blueberry
 	}
 }
diff --git a/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java b/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
index a5e9a2934db6fcb3a3087fc5615789f71f106f7a..2c494a7d5499e5e67ac94a85fadddee4a3e8b436 100644
--- a/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -72,6 +72,10 @@ public class LiquidBlockRenderer {
 		boolean bl = fluidState.is(FluidTags.LAVA);
 		TextureAtlasSprite[] textureAtlasSprites = bl ? this.lavaIcons : this.waterIcons;
 		int i = bl ? 16777215 : BiomeColors.getAverageWaterColor(blockAndTintGetter, blockPos);
+        // Blueberry start
+        net.blueberrymc.client.event.render.LiquidBlockRenderEvent event = net.blueberrymc.client.event.ClientEventFactory.callLiquidBlockRenderEvent(fluidState, blockPos, i);
+        i = event.getColor();
+        // Blueberry end
 		float f = (float)(i >> 16 & 0xFF) / 255.0F;
 		float g = (float)(i >> 8 & 0xFF) / 255.0F;
 		float h = (float)(i & 0xFF) / 255.0F;
diff --git a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
index 52e9afac9a377a5e50d7bf6c835adcc576932435..cd98788515f67c5ba75814b8854e4dc85d2147f8 100644
--- a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
+++ b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
@@ -27,7 +27,7 @@ import net.minecraft.world.phys.HitResult;
 import org.jetbrains.annotations.Nullable;
 
 @Environment(EnvType.CLIENT)
-public class BlockEntityRenderDispatcher implements ResourceManagerReloadListener {
+public class BlockEntityRenderDispatcher implements ResourceManagerReloadListener, net.blueberrymc.client.renderer.blockentity.MinecraftBlockEntityRenderDispatcher { // Blueberry - implement MinecraftBlockEntityRenderDispatcher
 	private Map<BlockEntityType<?>, BlockEntityRenderer<?>> renderers = ImmutableMap.of();
 	private final Font font;
 	private final EntityModelSet entityModelSet;
@@ -131,4 +131,11 @@ public class BlockEntityRenderDispatcher implements ResourceManagerReloadListene
 		);
 		this.renderers = BlockEntityRenderers.createEntityRenderers(context);
 	}
+
+    // Blueberry start
+    @Override
+    public final void registerSpecialRenderer(BlockEntityType<?> blockEntityType, BlockEntityRenderer<?> blockEntityRenderer) {
+        this.renderers.put(blockEntityType, blockEntityRenderer);
+    }
+    // Blueberry end
 }
diff --git a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
index 238a5d74bfa9c217aeb3ff3e88313639f85f1d63..a441a0bdbe9898b4d762ae1a71ecc95b9539cf97 100644
--- a/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
+++ b/src/main/java/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
@@ -14,7 +14,8 @@ import net.minecraft.world.level.block.entity.BlockEntityType;
 public class BlockEntityRenderers {
 	private static final Map<BlockEntityType<?>, BlockEntityRendererProvider<?>> PROVIDERS = Maps.<BlockEntityType<?>, BlockEntityRendererProvider<?>>newHashMap();
 
-	private static <T extends BlockEntity> void register(BlockEntityType<? extends T> blockEntityType, BlockEntityRendererProvider<T> blockEntityRendererProvider) {
+    // Blueberry - private -> public
+	public static <T extends BlockEntity> void register(BlockEntityType<? extends T> blockEntityType, BlockEntityRendererProvider<T> blockEntityRendererProvider) {
 		PROVIDERS.put(blockEntityType, blockEntityRendererProvider);
 	}
 
diff --git a/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java b/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
index 205b1ea11cc0b5bf255f72bbd24e689e6701c7c1..79ea9d39b4a03275a5c35fc096003f0615d5a1cf 100644
--- a/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/debug/DebugRenderer.java
@@ -20,6 +20,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.Vec3;
 import org.jetbrains.annotations.Nullable;
+import static net.blueberrymc.common.bml.InternalBlueberryModConfig.Debug.DebugRenderers.*; // Blueberry
 
 @Environment(EnvType.CLIENT)
 public class DebugRenderer {
@@ -100,7 +101,25 @@ public class DebugRenderer {
 
 	public void render(PoseStack poseStack, MultiBufferSource.BufferSource bufferSource, double d, double e, double f) {
 		if (this.renderChunkborder && !Minecraft.getInstance().showOnlyReducedInfo()) {
-			this.chunkBorderRenderer.render(poseStack, bufferSource, d, e, f);
+            // Blueberry start
+            if (pathfinding) this.pathfindingRenderer.render(poseStack, bufferSource, d, e, f);
+            if (waterDebug) this.waterDebugRenderer.render(poseStack, bufferSource, d, e, f);
+            if (chunkBorder) this.chunkBorderRenderer.render(poseStack, bufferSource, d, e, f);
+            if (heightMap) this.heightMapRenderer.render(poseStack, bufferSource, d, e, f);
+            if (collisionBox) this.collisionBoxRenderer.render(poseStack, bufferSource, d, e, f);
+            if (neighborsUpdate) this.neighborsUpdateRenderer.render(poseStack, bufferSource, d, e, f);
+            if (structure) this.structureRenderer.render(poseStack, bufferSource, d, e, f);
+            if (lightDebug) this.lightDebugRenderer.render(poseStack, bufferSource, d, e, f);
+            if (worldGenAttempt) this.worldGenAttemptRenderer.render(poseStack, bufferSource, d, e, f);
+            if (solidFace) this.solidFaceRenderer.render(poseStack, bufferSource, d, e, f);
+            if (chunk) this.chunkRenderer.render(poseStack, bufferSource, d, e, f);
+            if (brainDebug) this.brainDebugRenderer.render(poseStack, bufferSource, d, e, f);
+            if (villageSectionsDebug) this.villageSectionsDebugRenderer.render(poseStack, bufferSource, d, e, f);
+            if (beeDebug) this.beeDebugRenderer.render(poseStack, bufferSource, d, e, f);
+            if (raidDebug) this.raidDebugRenderer.render(poseStack, bufferSource, d, e, f);
+            if (goalSelector) this.goalSelectorRenderer.render(poseStack, bufferSource, d, e, f);
+            if (gameTestDebug) this.gameTestDebugRenderer.render(poseStack, bufferSource, d, e, f);
+            // Blueberry end
 		}
 
 		this.gameTestDebugRenderer.render(poseStack, bufferSource, d, e, f);
diff --git a/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java b/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java
index 4b955f506f0e4869ab93987870b015555770c4de..2b6e6c734f4f41c1cc5c4e415527be8d6a102f11 100644
--- a/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java
+++ b/src/main/java/net/minecraft/client/renderer/entity/ItemRenderer.java
@@ -160,7 +160,7 @@ public class ItemRenderer implements ResourceManagerReloadListener {
 
 				this.renderModelLists(bakedModel, itemStack, i, j, poseStack, vertexConsumer);
 			} else {
-				this.blockEntityRenderer.renderByItem(itemStack, itemDisplayContext, poseStack, multiBufferSource, i, j);
+                itemStack.getItem().getRenderer().renderByItem(itemStack, itemDisplayContext, poseStack, multiBufferSource, i, j); // Blueberry
 			}
 
 			poseStack.popPose();
diff --git a/src/main/java/net/minecraft/client/resources/model/ModelBakery.java b/src/main/java/net/minecraft/client/resources/model/ModelBakery.java
index 33827434b04f08893431f622c5aeed89b6471c74..d45f7e26f20272b0bfc02f0ca821ed3ad48edd2f 100644
--- a/src/main/java/net/minecraft/client/resources/model/ModelBakery.java
+++ b/src/main/java/net/minecraft/client/resources/model/ModelBakery.java
@@ -125,6 +125,13 @@ public class ModelBakery {
 	private final Object2IntMap<BlockState> modelGroups = Util.make(
 		new Object2IntOpenHashMap<>(), object2IntOpenHashMap -> object2IntOpenHashMap.defaultReturnValue(-1)
 	);
+    // Blueberry start
+    private static final Set<ResourceLocation> specialModels = new java.util.HashSet<>();
+
+    public static void addSpecialModel(ResourceLocation resourceLocation) {
+        specialModels.add(resourceLocation);
+    }
+    // Blueberry end
 
 	public ModelBakery(
 		BlockColors blockColors, ProfilerFiller profilerFiller, Map<ResourceLocation, BlockModel> map, Map<ResourceLocation, List<ModelBakery.LoadedJson>> map2
@@ -160,8 +167,10 @@ public class ModelBakery {
 		}
 
 		profilerFiller.popPush("special");
+        new net.blueberrymc.client.event.SpecialModelRegistryEvent().callEvent(); // Blueberry
 		this.loadTopLevel(ItemRenderer.TRIDENT_IN_HAND_MODEL);
 		this.loadTopLevel(ItemRenderer.SPYGLASS_IN_HAND_MODEL);
+        for (ResourceLocation resourceLocation : getSpecialModels()) this.addModelToCache(resourceLocation); // Blueberry
 		this.topLevelModels.values().forEach(unbakedModel -> unbakedModel.resolveParents(this::getModel));
 		profilerFiller.pop();
 	}
@@ -403,7 +412,19 @@ public class ModelBakery {
 		this.topLevelModels.put(modelResourceLocation, unbakedModel);
 	}
 
-	private void registerModelGroup(Iterable<BlockState> iterable) {
+    // Blueberry start
+    private void addModelToCache(ResourceLocation resourceLocation) {
+        UnbakedModel unbakedModel = this.getModel(resourceLocation);
+        this.unbakedCache.put(resourceLocation, unbakedModel);
+        this.topLevelModels.put(resourceLocation, unbakedModel);
+    }
+
+    public Set<ResourceLocation> getSpecialModels() {
+        return specialModels;
+    }
+    // Blueberry end
+
+    private void registerModelGroup(Iterable<BlockState> iterable) {
 		int i = this.nextModelGroup++;
 		iterable.forEach(blockState -> this.modelGroups.put(blockState, i));
 	}
diff --git a/src/main/java/net/minecraft/network/protocol/common/custom/DiscardedPayload.java b/src/main/java/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
index 5fda67ba06ee75fe0c62a00e0b9ca2a82827e97c..ee8e117c6d2e11f55d86a3c4817bd665cb214551 100644
--- a/src/main/java/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
+++ b/src/main/java/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
@@ -4,7 +4,24 @@ import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.ResourceLocation;
 
-public record DiscardedPayload(ResourceLocation id) implements CustomPacketPayload {
+// Blueberry start
+public record DiscardedPayload(ResourceLocation id, byte[] payload) implements CustomPacketPayload, net.blueberrymc.network.BlueberryCustomPayload {
+    @Deprecated
+    public DiscardedPayload(ResourceLocation id) {
+        this(id, new byte[0]);
+    }
+
+    public DiscardedPayload(ResourceLocation id, FriendlyByteBuf buf) {
+        this(id, readRemainingBytes(buf));
+    }
+
+    private static byte[] readRemainingBytes(FriendlyByteBuf buf) {
+        byte[] bytes = new byte[buf.readableBytes()];
+        buf.readBytes(bytes);
+        return bytes;
+    }
+    // Blueberry end
+
 	public static <T extends FriendlyByteBuf> StreamCodec<T, DiscardedPayload> codec(ResourceLocation resourceLocation, int i) {
 		return CustomPacketPayload.codec((discardedPayload, friendlyByteBuf) -> {
 		}, friendlyByteBuf -> {
diff --git a/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java b/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
index 8a52addea7666e52c7e396cde2d69b69d5da11c4..445b9aa0e77bfee21057cfd1c43a2a23aac67f4e 100644
--- a/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
+++ b/src/main/java/net/minecraft/network/protocol/status/ClientStatusPacketListener.java
@@ -4,7 +4,7 @@ import net.minecraft.network.ClientboundPacketListener;
 import net.minecraft.network.ConnectionProtocol;
 import net.minecraft.network.protocol.ping.ClientPongPacketListener;
 
-public interface ClientStatusPacketListener extends ClientPongPacketListener, ClientboundPacketListener {
+public interface ClientStatusPacketListener extends ClientPongPacketListener, ClientboundPacketListener, net.blueberrymc.network.client.handshake.ClientBlueberryHandshakePacketListener { // Blueberry
 	@Override
 	default ConnectionProtocol protocol() {
 		return ConnectionProtocol.STATUS;
diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index 348ad247505f9768ad10f5d6874a7c45e35daf3e..dfa9889bc853516ec2743f9fb3fbcd5c5d5d662a 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -50,6 +50,7 @@ public class Bootstrap {
 					EntitySelectorOptions.bootStrap();
 					DispenseItemBehavior.bootStrap();
 					CauldronInteraction.bootStrap();
+                    net.blueberrymc.common.event.lifecycle.RegistryBootstrappedEvent.fire(); // Blueberry
 					BuiltInRegistries.bootStrap();
 					CreativeModeTabs.validate();
 					wrapStreams();
diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 444c99d3a8c646a583e84d5408da934059b2afac..a41a37c79740dddc4c98ce690a24a85b699bd81e 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -64,7 +64,7 @@ public class Main {
 
 	@DontObfuscate
 	public static void main(String[] strings) {
-		SharedConstants.tryDetectVersion();
+        //SharedConstants.tryDetectVersion(); // Blueberry - moved below
 		OptionParser optionParser = new OptionParser();
 		OptionSpec<Void> optionSpec = optionParser.accepts("nogui");
 		OptionSpec<Void> optionSpec2 = optionParser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
@@ -82,6 +82,7 @@ public class Main {
 		OptionSpec<Void> optionSpec14 = optionParser.accepts("jfrProfile");
 		OptionSpec<Path> optionSpec15 = optionParser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
 		OptionSpec<String> optionSpec16 = optionParser.nonOptions();
+        optionParser.allowsUnrecognizedOptions(); // Blueberry
 
 		try {
 			OptionSet optionSet = optionParser.parse(strings);
@@ -89,6 +90,12 @@ public class Main {
 				optionParser.printHelpOn(System.err);
 				return;
 			}
+            // Blueberry start
+            net.blueberrymc.common.Blueberry.preBootstrap();
+            net.blueberrymc.common.Blueberry.bootstrap(new net.blueberrymc.server.BlueberryServerImpl());
+            SharedConstants.tryDetectVersion();
+            // SharedConstants.enableDataFixerOptimizations();
+            // Blueberry end
 
 			Path path = optionSet.valueOf(optionSpec15);
 			if (path != null) {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index b5d8aa2a3bdb72ab2ad1dcafd30ec9b3ff37a80f..1aa50770bb4b3b7238412331cf8f795a1c6d3e1e 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -292,6 +292,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 			this.proxy = proxy;
 			this.packRepository = packRepository;
 			this.resources = new MinecraftServer.ReloadableResources(worldStem.resourceManager(), worldStem.dataPackResources());
+            // Blueberry start
+            if (net.blueberrymc.common.Blueberry.isServer()) {
+                ((net.blueberrymc.server.BlueberryServer) net.blueberrymc.common.Blueberry.getUtil()).setServer(this);
+                net.blueberrymc.common.Blueberry.getModLoader().callPreInit();
+            }
+            // Blueberry end
 			this.services = services;
 			if (services.profileCache() != null) {
 				services.profileCache().setExecutor(this);
@@ -313,6 +319,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 			this.serverThread = thread;
 			this.executor = Util.backgroundExecutor();
 			this.potionBrewing = PotionBrewing.bootstrap(this.worldData.enabledFeatures());
+            if (net.blueberrymc.common.Blueberry.isServer()) net.blueberrymc.common.Blueberry.getModLoader().callInit(); // Blueberry
 		}
 	}
 
@@ -668,6 +675,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 			this.nextTickTimeNanos = Util.getNanos();
 			this.statusIcon = (ServerStatus.Favicon)this.loadStatusIcon().orElse(null);
 			this.status = this.buildServerStatus();
+            if (net.blueberrymc.common.Blueberry.isServer()) net.blueberrymc.common.Blueberry.getModLoader().callPostInit(); // Blueberry
 
 			while (this.running) {
 				long l;
@@ -898,6 +906,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 			LOGGER.debug("Autosave finished");
 		}
 
+        // Blueberry start
+        this.profiler.push("blueberryServerScheduler");
+        net.blueberrymc.common.Blueberry.getUtil().getServerScheduler().tick();
+        this.profiler.pop();
+        // Blueberry end
 		this.profiler.push("tallying");
 		long m = Util.getNanos() - l;
 		int i = this.tickCount % 100;
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 3ac7cfed2bf3a1424b846ac061700d03342d255f..7ed31606b4a22972b1c267fc7db632f8d381601f 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -557,6 +557,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
 	@Override
 	public void stopServer() {
+        net.blueberrymc.common.Blueberry.shutdown(); // Blueberry
 		super.stopServer();
 		Util.shutdownExecutors();
 		SkullBlockEntity.clear();
diff --git a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index fc56dacca848882eec073c29d3aef53d090071af..307b6f7304f63e06efbebfe79df0df75335f7280 100644
--- a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -26,7 +26,7 @@ import net.minecraft.util.VisibleForDebug;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 
-public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener {
+public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener, net.blueberrymc.network.server.ServerBlueberryPacketListener { // Blueberry
 	private static final Logger LOGGER = LogUtils.getLogger();
 	public static final int LATENCY_CHECK_INTERVAL = 15000;
 	private static final int CLOSED_LISTENER_TIMEOUT = 15000;
@@ -83,6 +83,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
 	@Override
 	public void handleCustomPayload(ServerboundCustomPayloadPacket serverboundCustomPayloadPacket) {
+        if (net.blueberrymc.network.server.ServerBlueberryPacketListener.super.handleBlueberryCustomPayload(serverboundCustomPayloadPacket)) return; // Blueberry
 	}
 
 	@Override
diff --git a/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
index 77912be1e8fb133bbfeb6c799094bb6eb72647a3..e2b104e610c4ac4145dc815bafc6dd369cdf81d3 100644
--- a/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -36,6 +36,7 @@ public class ServerStatusPacketListenerImpl implements ServerStatusPacketListene
 		} else {
 			this.hasRequestedStatus = true;
 			this.connection.send(new ClientboundStatusResponsePacket(this.status));
+            //this.connection.send(new net.blueberrymc.network.client.handshake.ClientboundBlueberryHandshakePacket(net.blueberrymc.common.Blueberry.getModLoader().getModInfos())); // Blueberry
 		}
 	}
 
diff --git a/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java b/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java
index cf4210ec95f894f0e612f8a07d57af66e7f27e35..2b612f9f482c17e9cd06c6564d02b4203f342a6b 100644
--- a/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java
+++ b/src/main/java/net/minecraft/server/packs/repository/BuiltInPackSource.java
@@ -47,6 +47,8 @@ public abstract class BuiltInPackSource implements RepositorySource {
 		}
 
 		this.listBundledPacks(consumer);
+
+        net.blueberrymc.common.Blueberry.getModLoader().loadPacks(consumer); // Blueberry
 	}
 
 	@Nullable
diff --git a/src/main/java/net/minecraft/world/inventory/MenuType.java b/src/main/java/net/minecraft/world/inventory/MenuType.java
index 0e042052eebc2b67b9f61dd4faed65da56915076..44dd77d2bc9912ebc553d0735527852453b5112e 100644
--- a/src/main/java/net/minecraft/world/inventory/MenuType.java
+++ b/src/main/java/net/minecraft/world/inventory/MenuType.java
@@ -45,7 +45,8 @@ public class MenuType<T extends AbstractContainerMenu> implements FeatureElement
 		return Registry.register(BuiltInRegistries.MENU, string, new MenuType<>(menuSupplier, FeatureFlags.REGISTRY.subset(featureFlags)));
 	}
 
-	private MenuType(MenuType.MenuSupplier<T> menuSupplier, FeatureFlagSet featureFlagSet) {
+    // Blueberry - private -> public
+	public MenuType(MenuType.MenuSupplier<T> menuSupplier, FeatureFlagSet featureFlagSet) {
 		this.constructor = menuSupplier;
 		this.requiredFeatures = featureFlagSet;
 	}
@@ -59,6 +60,7 @@ public class MenuType<T extends AbstractContainerMenu> implements FeatureElement
 		return this.requiredFeatures;
 	}
 
+    public // Blueberry
 	interface MenuSupplier<T extends AbstractContainerMenu> {
 		T create(int i, Inventory inventory);
 	}
diff --git a/src/main/java/net/minecraft/world/item/Item.java b/src/main/java/net/minecraft/world/item/Item.java
index f40a150328520dadb16f190a203bd71325cca47a..f297304fa19394f24be3ec9e563a6fbd7fd79918 100644
--- a/src/main/java/net/minecraft/world/item/Item.java
+++ b/src/main/java/net/minecraft/world/item/Item.java
@@ -93,6 +93,7 @@ public class Item implements FeatureElement, ItemLike {
 				LOGGER.error("Item classes should end with Item and {} doesn't.", string);
 			}
 		}
+        this.renderer = properties.renderer; // Blueberry
 	}
 
 	@Deprecated
@@ -332,13 +333,23 @@ public class Item implements FeatureElement, ItemLike {
 		return this.requiredFeatures;
 	}
 
-	public static class Properties {
+    // Blueberry start
+    private final java.util.function.Supplier<net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer> renderer;
+
+    public final net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer getRenderer() {
+        net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer renderer = this.renderer != null ? this.renderer.get() : null;
+        return renderer != null ? renderer : net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer.instance;
+    }
+    // Blueberry end
+
+    public static class Properties {
 		private static final Interner<DataComponentMap> COMPONENT_INTERNER = Interners.newStrongInterner();
 		@Nullable
 		private DataComponentMap.Builder components;
 		@Nullable
 		Item craftingRemainingItem;
 		FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
+        java.util.function.Supplier<net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer> renderer; // Blueberry
 
 		public Item.Properties food(FoodProperties foodProperties) {
 			return this.component(DataComponents.FOOD, foodProperties);
@@ -373,6 +384,13 @@ public class Item implements FeatureElement, ItemLike {
 			return this;
 		}
 
+        // Blueberry start
+        public Item.Properties renderer(java.util.function.Supplier<net.minecraft.client.renderer.BlockEntityWithoutLevelRenderer> renderer) {
+            this.renderer = renderer;
+            return this;
+        }
+        // Blueberry end
+
 		public <T> Item.Properties component(DataComponentType<T> dataComponentType, T object) {
 			if (this.components == null) {
 				this.components = DataComponentMap.builder().addAll(DataComponents.COMMON_ITEM_COMPONENTS);
diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java
index 4aaa7f53f0aa845de30a4f292106d515dd1f5b55..a1aa6f34730094695ec85cdebec409c1cf18a653 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -313,6 +313,7 @@ public class BlockEntityType<T extends BlockEntity> {
 	}
 
 	@FunctionalInterface
+    public // Blueberry
 	interface BlockEntitySupplier<T extends BlockEntity> {
 		T create(BlockPos blockPos, BlockState blockState);
 	}
diff --git a/src/main/resources/log4j2.xml b/src/main/resources/log4j2.xml
new file mode 100644
index 0000000000000000000000000000000000000000..8b00a0d606dae68533536e2eb8a6b10030a9e098
--- /dev/null
+++ b/src/main/resources/log4j2.xml
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="WARN">
+    <Appenders>
+        <Console name="SysOut" target="SYSTEM_OUT">
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
+        </Console>
+        <Queue name="ServerGuiConsole">
+            <PatternLayout pattern="[%d{HH:mm:ss} %level] [%c]: %msg%n" />
+        </Queue>
+        <RollingRandomAccessFile name="File" fileName="logs/latest.log" filePattern="logs/%d{yyyy-MM-dd}-%i.log.gz">
+            <PatternLayout pattern="[%d{HH:mm:ss}] [%t/%level] [%c]: %msg%n" />
+            <Policies>
+                <TimeBasedTriggeringPolicy />
+                <OnStartupTriggeringPolicy />
+            </Policies>
+        </RollingRandomAccessFile>
+    </Appenders>
+    <Loggers>
+        <Root level="info">
+            <filters>
+                <MarkerFilter marker="NETWORK_PACKETS" onMatch="DENY" onMismatch="NEUTRAL" />
+            </filters>
+            <AppenderRef ref="SysOut"/>
+            <AppenderRef ref="File"/>
+            <AppenderRef ref="ServerGuiConsole"/>
+        </Root>
+    </Loggers>
+</Configuration>
