From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 3 Mar 2021 11:59:03 +0900
Subject: [PATCH] CommandRegistrationEvent to register commands easily


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index f52297f2f6b46ba2a9f6890f659c1b08e9056e90..8e6177844b2092d22cf365d3bed5b8c67d8a64b7 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -190,6 +190,7 @@ public class Commands {
         TriggerCommand.register(this.dispatcher);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        net.blueberrymc.common.event.EventFactory.callCommandRegistrationEvent(this.dispatcher, commandSelection); // Blueberry
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
