From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 3 Mar 2021 11:59:03 +0900
Subject: [PATCH] CommandRegistrationEvent to register commands easily


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 34bc3b5a4126b1e55de0122f3a7dae800c1dd537..97d8ef9de863ef267edef5bb11a890d68a4bc4b3 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -183,6 +183,7 @@ public class Commands {
         TriggerCommand.register(this.dispatcher);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        net.blueberrymc.common.event.EventFactory.callCommandRegistrationEvent(this.dispatcher, commandSelection); // Blueberry
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
