From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Thu, 4 Mar 2021 22:06:05 +0900
Subject: [PATCH] Implement addRecipe/removeRecipe API


diff --git a/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java b/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
index 7793801375bb24745e2ee9bfd688a9d4d950dbd4..b1fa6c96a742387c9f8b4c42516c243766263bf1 100644
--- a/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/src/main/java/net/minecraft/world/item/crafting/RecipeManager.java
@@ -33,12 +33,12 @@ import net.minecraft.world.level.Level;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class RecipeManager extends SimpleJsonResourceReloadListener {
+public class RecipeManager extends SimpleJsonResourceReloadListener implements net.blueberrymc.world.item.crafting.BlueberryRecipeManager { // Blueberry
 	private static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
 	private static final Logger LOGGER = LogUtils.getLogger();
 	private final HolderLookup.Provider registries;
-	private Multimap<RecipeType<?>, RecipeHolder<?>> byType = ImmutableMultimap.of();
-	private Map<ResourceLocation, RecipeHolder<?>> byName = ImmutableMap.of();
+	private Multimap<RecipeType<?>, RecipeHolder<?>> byType = com.google.common.collect.Multimaps.newListMultimap(com.google.common.collect.Maps.newHashMap(), java.util.ArrayList::new);
+	private Map<ResourceLocation, RecipeHolder<?>> byName = com.google.common.collect.Maps.newHashMap(); // Blueberry - use HashMap instead
 	private boolean hasErrors;
 
 	public RecipeManager(HolderLookup.Provider provider) {
@@ -58,15 +58,16 @@ public class RecipeManager extends SimpleJsonResourceReloadListener {
 			try {
 				Recipe<?> recipe = Recipe.CODEC.parse(registryOps, (JsonElement)entry.getValue()).getOrThrow(JsonParseException::new);
 				RecipeHolder<?> recipeHolder = new RecipeHolder<>(resourceLocation, recipe);
-				builder.put(recipe.getType(), recipeHolder);
-				builder2.put(resourceLocation, recipeHolder);
+				byType.put(recipe.getType(), recipeHolder); // Blueberry
+				byName.put(resourceLocation, recipeHolder); // Blueberry
 			} catch (IllegalArgumentException | JsonParseException var12) {
 				LOGGER.error("Parsing error loading recipe {}", resourceLocation, var12);
 			}
 		}
 
-		this.byType = builder.build();
-		this.byName = builder2.build();
+//		this.byType = builder.build(); // Blueberry
+//		this.byName = builder2.build(); // Blueberry
+		net.blueberrymc.world.item.crafting.RecipeBuilder.getRecipes().forEach((id, recipe) -> addRecipe(recipe)); // Blueberry - add recipes now
 		LOGGER.info("Loaded {} recipes", this.byType.size());
 	}
 
@@ -185,6 +186,26 @@ public class RecipeManager extends SimpleJsonResourceReloadListener {
 		};
 	}
 
+	// Blueberry start - implement addRecipe/removeRecipe
+	@Override
+	public void addRecipe(RecipeHolder<?> recipe) {
+		if (!net.blueberrymc.world.item.crafting.RecipeBuilder.knownRecipeManagers.contains(this))
+			net.blueberrymc.world.item.crafting.RecipeBuilder.knownRecipeManagers.add(this);
+		if (byName.containsKey(recipe.id())) {
+			throw new IllegalStateException("Duplicate recipe ignored with ID " + recipe.id());
+		} else {
+			byType.put(recipe.value().getType(), recipe);
+			byName.put(recipe.id(), recipe);
+		}
+	}
+
+	@Override
+	public void removeRecipe(RecipeType<?> type, ResourceLocation id) {
+		var holder = this.byName.remove(id);
+		this.byType.get(type).remove(holder);
+	}
+	// Blueberry end - implement addRecipe/removeRecipe
+
 	public interface CachedCheck<C extends Container, T extends Recipe<C>> {
 		Optional<RecipeHolder<T>> getRecipeFor(C container, Level level);
 	}
