From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 5 Mar 2021 21:37:57 +0900
Subject: [PATCH] MenuScreens#register PAIL private -> public


diff --git a/src/main/java/net/minecraft/client/gui/screens/MenuScreens.java b/src/main/java/net/minecraft/client/gui/screens/MenuScreens.java
index 20f9815a9981e24bcdf379f799a5be9f0f6499ce..610bbd3f103f2f339ca91440ba51f7d8e467f976 100644
--- a/src/main/java/net/minecraft/client/gui/screens/MenuScreens.java
+++ b/src/main/java/net/minecraft/client/gui/screens/MenuScreens.java
@@ -54,7 +54,8 @@ public class MenuScreens {
 		return (MenuScreens.ScreenConstructor<T, ?>)SCREENS.get(menuType);
 	}
 
-	private static <M extends AbstractContainerMenu, U extends Screen & MenuAccess<M>> void register(
+	// Blueberry - private -> public
+	public static <M extends AbstractContainerMenu, U extends Screen & MenuAccess<M>> void register(
 		MenuType<? extends M> menuType, MenuScreens.ScreenConstructor<M, U> screenConstructor
 	) {
 		MenuScreens.ScreenConstructor<?, ?> screenConstructor2 = (MenuScreens.ScreenConstructor<?, ?>)SCREENS.put(menuType, screenConstructor);
