From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 5 Mar 2021 21:38:13 +0900
Subject: [PATCH] Fire simple events


diff --git a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
index d7df20b40c4585b84658c830aef66e56ace0ac82..1d630e0c19b9985ab7cf82f9af1cfae5da3d0db2 100644
--- a/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/src/main/java/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -2445,6 +2445,13 @@ public class ClientPacketListener extends ClientCommonPacketListenerImpl impleme
 	}
 
 	public void sendChat(String string) {
+		// Blueberry start - fire ClientLocalPlayerChatEvent
+		var event = new net.blueberrymc.client.event.player.ClientLocalPlayerChatEvent(string);
+		if (!event.callEvent()) {
+			return;
+		}
+		string = event.getMessage();
+		// Blueberry end - fire ClientLocalPlayerChatEvent
 		Instant instant = Instant.now();
 		long l = Crypt.SaltSupplier.getLong();
 		LastSeenMessagesTracker.Update update = this.lastSeenMessages.generateAndApplyUpdate();
diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 5d79c2e7b393c212bca8a191b8e839072ca63694..c28469432fb30297f4ab161bb63c9b3ee948483b 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -118,7 +118,10 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 		this.address = this.channel.remoteAddress();
 		if (this.delayedDisconnect != null) {
 			this.disconnect(this.delayedDisconnect);
+            return; // Blueberry
 		}
+
+        new net.blueberrymc.common.event.network.ConnectionInitEvent(this.channel, this.receiving).callEvent(); // Blueberry
 	}
 
 	@Override
@@ -585,6 +588,8 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 				this.channel.pipeline().remove("compress");
 			}
 		}
+
+        new net.blueberrymc.common.event.network.ConnectionSetupCompressionEvent(this.channel, this.receiving, i).callEvent(); // Blueberry
 	}
 
 	public void handleDisconnection() {
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 31d70416e30c7262e329725ba5948c7913ec55bc..7ea6a5f39a921831ba6ca809f40dfa1b7664f6ad 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -225,6 +225,7 @@ public abstract class PlayerList {
 		this.broadcastAll(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(serverPlayer)));
 		this.sendLevelInfo(serverPlayer, serverLevel2);
 		serverLevel2.addNewPlayer(serverPlayer);
+        new net.blueberrymc.common.event.player.PlayerJoinEvent(serverPlayer).callEvent(); // Blueberry
 		this.server.getCustomBossEvents().onPlayerConnect(serverPlayer);
 
 		for (MobEffectInstance mobEffectInstance : serverPlayer.getActiveEffects()) {
diff --git a/src/main/java/net/minecraft/world/level/biome/Biome.java b/src/main/java/net/minecraft/world/level/biome/Biome.java
index ddcc9705c6ec74350d45b72da81676c98032beb2..93d31033ff8d051ceed723b3a11d46286ea94d0c 100644
--- a/src/main/java/net/minecraft/world/level/biome/Biome.java
+++ b/src/main/java/net/minecraft/world/level/biome/Biome.java
@@ -312,12 +312,12 @@ public final class Biome {
 
 		public Biome build() {
 			if (this.temperature != null && this.downfall != null && this.specialEffects != null && this.mobSpawnSettings != null && this.generationSettings != null) {
-				return new Biome(
+				return net.blueberrymc.common.bml.event.Event.callEvent(new net.blueberrymc.common.event.biome.BiomeInitEvent(this, () -> new Biome(
 					new Biome.ClimateSettings(this.hasPrecipitation, this.temperature, this.temperatureModifier, this.downfall),
 					this.specialEffects,
 					this.generationSettings,
 					this.mobSpawnSettings
-				);
+				))).getBiome().get();
 			} else {
 				throw new IllegalStateException("You are missing parameters to build a proper biome\n" + this);
 			}
diff --git a/src/main/java/net/minecraft/world/level/biome/BiomeGenerationSettings.java b/src/main/java/net/minecraft/world/level/biome/BiomeGenerationSettings.java
index cd24cf0696e48bed5804e808a2a394d72f96d2d7..4788c58c68552ee911e99b323c3f9290ae25ce78 100644
--- a/src/main/java/net/minecraft/world/level/biome/BiomeGenerationSettings.java
+++ b/src/main/java/net/minecraft/world/level/biome/BiomeGenerationSettings.java
@@ -129,13 +129,14 @@ public class BiomeGenerationSettings {
 		}
 
 		public BiomeGenerationSettings build() {
-			return new BiomeGenerationSettings(
-				(Map<GenerationStep.Carving, HolderSet<ConfiguredWorldCarver<?>>>)this.carvers
+            // Blueberry - call BiomeGenerationSettingsInitEvent
+            return net.blueberrymc.common.bml.event.Event.callEvent(new net.blueberrymc.common.event.biome.BiomeGenerationSettingsInitEvent(this, () -> new BiomeGenerationSettings(
+				this.carvers
 					.entrySet()
 					.stream()
 					.collect(ImmutableMap.toImmutableMap(Entry::getKey, entry -> HolderSet.direct(entry.getValue()))),
-				(List<HolderSet<PlacedFeature>>)this.features.stream().map(HolderSet::direct).collect(ImmutableList.toImmutableList())
-			);
+				this.features.stream().map(HolderSet::direct).collect(ImmutableList.toImmutableList())
+			))).getBiomeGenerationSettings().get();
 		}
 	}
 }
