From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 17 Mar 2021 21:21:48 +0900
Subject: [PATCH] BungeeCord ip_forward support


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index c28469432fb30297f4ab161bb63c9b3ee948483b..387db9deb36be0c7878868be55868a3ce277636c 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -87,7 +87,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 	private volatile boolean sendLoginDisconnect = true;
 	private final Queue<Consumer<Connection>> pendingActions = Queues.<Consumer<Connection>>newConcurrentLinkedQueue();
 	private Channel channel;
-	private SocketAddress address;
+	public SocketAddress address; // Blueberry - private -> public
 	@Nullable
 	private volatile PacketListener disconnectListener;
 	@Nullable
@@ -106,6 +106,11 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 	private volatile Component delayedDisconnect;
 	@Nullable
 	BandwidthDebugMonitor bandwidthDebugMonitor;
+    // Blueberry start
+    public @Nullable java.util.UUID spoofedUUID;
+    public @Nullable com.mojang.authlib.properties.Property[] spoofedProfile;
+    public boolean preparing = true;
+    // Blueberry end
 
 	public Connection(PacketFlow packetFlow) {
 		this.receiving = packetFlow;
@@ -116,6 +121,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 		super.channelActive(channelHandlerContext);
 		this.channel = channelHandlerContext.channel();
 		this.address = this.channel.remoteAddress();
+        this.preparing = false; // Blueberry
 		if (this.delayedDisconnect != null) {
 			this.disconnect(this.delayedDisconnect);
             return; // Blueberry
@@ -414,6 +420,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 	}
 
 	public void disconnect(Component component) {
+        this.preparing = false; // Blueberry
 		if (this.channel == null) {
 			this.delayedDisconnect = component;
 		}
diff --git a/src/main/java/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java b/src/main/java/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
index 5a51dc1cae49489ed67530c8abb8f6ace2bb1348..625bbbfd4f6cbd821f536535d00ba9c6d2753d94 100644
--- a/src/main/java/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
@@ -4,18 +4,23 @@ import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
+import java.util.Objects; // Blueberry
 
-public record ClientIntentionPacket(int protocolVersion, String hostName, int port, ClientIntent intention) implements Packet<ServerHandshakePacketListener> {
+public final class ClientIntentionPacket implements Packet<ServerHandshakePacketListener> {
 	public static final StreamCodec<FriendlyByteBuf, ClientIntentionPacket> STREAM_CODEC = Packet.codec(ClientIntentionPacket::write, ClientIntentionPacket::new);
 	private static final int MAX_HOST_LENGTH = 255;
+    private final int protocolVersion;
+    public String hostName;
+    private final int port;
+    private final ClientIntent intention;
 
-	@Deprecated
-	public ClientIntentionPacket(int protocolVersion, String hostName, int port, ClientIntent intention) {
-		this.protocolVersion = protocolVersion;
-		this.hostName = hostName;
-		this.port = port;
-		this.intention = intention;
-	}
+    public ClientIntentionPacket(int protocolVersion, String hostName, int port, ClientIntent intention) {
+        this.protocolVersion = protocolVersion;
+        this.hostName = hostName;
+        this.port = port;
+        this.intention = intention;
+    }
+    // Blueberry end - convert record to class
 
 	private ClientIntentionPacket(FriendlyByteBuf friendlyByteBuf) {
 		this(friendlyByteBuf.readVarInt(), friendlyByteBuf.readUtf(255), friendlyByteBuf.readUnsignedShort(), ClientIntent.byId(friendlyByteBuf.readVarInt()));
@@ -41,4 +46,47 @@ public record ClientIntentionPacket(int protocolVersion, String hostName, int po
 	public boolean isTerminal() {
 		return true;
 	}
+
+    // Blueberry start - convert record to class
+    public int protocolVersion() {
+        return protocolVersion;
+    }
+
+    public String hostName() {
+        return hostName;
+    }
+
+    public int port() {
+        return port;
+    }
+
+    public ClientIntent intention() {
+        return intention;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) return true;
+        if (obj == null || obj.getClass() != this.getClass()) return false;
+        var that = (ClientIntentionPacket) obj;
+        return this.protocolVersion == that.protocolVersion &&
+                Objects.equals(this.hostName, that.hostName) &&
+                this.port == that.port &&
+                Objects.equals(this.intention, that.intention);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(protocolVersion, hostName, port, intention);
+    }
+
+    @Override
+    public String toString() {
+        return "ClientIntentionPacket[" +
+                "protocolVersion=" + protocolVersion + ", " +
+                "hostName=" + hostName + ", " +
+                "port=" + port + ", " +
+                "intention=" + intention + ']';
+    }
+    // Blueberry end - convert record to class
 }
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 7ed31606b4a22972b1c267fc7db632f8d381601f..372a17f8f30b5b63ea00a0d58e16609dfba376f7 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -160,9 +160,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 		if (!this.usesAuthentication()) {
 			LOGGER.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
 			LOGGER.warn("The server will make no attempt to authenticate usernames. Beware.");
-			LOGGER.warn(
-				"While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose."
-			);
+			// Blueberry start
+			if (net.blueberrymc.common.bml.InternalBlueberryModConfig.Server.bungee) {
+				LOGGER.warn("Whilst this makes it possible to use BungeeCord, unless access to your server is properly restricted, it also opens up the ability for hackers to connect with any username they choose.");
+				LOGGER.warn("Please see https://www.spigotmc.org/wiki/firewall-guide/ for further information.");
+			} else {
+				LOGGER.warn("While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose.");
+			}
+			// Blueberry end
 			LOGGER.warn("To change this, set \"online-mode\" to \"true\" in the server.properties file.");
 		}
 
diff --git a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
index fa25ebe58fff31d898e493b758dff6d8b5b7e536..19b764b3945d8422d78136d8ea50d94998654334 100644
--- a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
+++ b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
@@ -155,6 +155,7 @@ public class ServerConnectionListener {
 							connection.setReadOnly();
 						}
 					} else {
+                        if (connection.preparing) continue; // Blueberry
 						iterator.remove();
 						connection.handleDisconnection();
 					}
diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index ec92e0c5229913805c929cce6116cbfbf65523ec..ff3d6e0331b55f9ab83ca22af25d78387f80692f 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -65,6 +65,24 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
 			this.connection.disconnect(component);
 		} else {
 			this.connection.setupInboundProtocol(LoginProtocols.SERVERBOUND, new ServerLoginPacketListenerImpl(this.server, this.connection, bl));
+			// Blueberry start
+			if (net.blueberrymc.common.bml.InternalBlueberryModConfig.Server.bungee) {
+				String[] split = clientIntentionPacket.hostName().split("\00");
+				if (split.length == 3 || split.length == 4) {
+					clientIntentionPacket.hostName = split[0];
+					connection.address = new java.net.InetSocketAddress(split[1], ((java.net.InetSocketAddress) connection.getRemoteAddress()).getPort());
+					connection.spoofedUUID = com.mojang.util.UndashedUuid.fromStringLenient(split[2]);
+				} else {
+					Component text = Component.literal("If you wish to use IP forwarding, please enable it in your BungeeCord config as well!");
+					connection.send(new ClientboundLoginDisconnectPacket(text));
+					connection.disconnect(text);
+					return;
+				}
+				if (split.length == 4) {
+					connection.spoofedProfile = new com.google.gson.Gson().fromJson(split[3], com.mojang.authlib.properties.Property[].class);
+				}
+			}
+// Blueberry end
 		}
 	}
 
diff --git a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index f2a70756215545426b43fab8e8455cafaa30ee18..109713ec3ee8502fdc7b7ed52988f348e9cff1b8 100644
--- a/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -130,7 +130,19 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
 	}
 
 	void startClientVerification(GameProfile gameProfile) {
-		this.authenticatedProfile = gameProfile;
+        // Blueberry start
+        if (net.blueberrymc.common.bml.InternalBlueberryModConfig.Server.bungee) {
+            java.util.UUID uuid = connection.spoofedUUID != null ? connection.spoofedUUID : UUIDUtil.createOfflinePlayerUUID(gameProfile.getName());
+            this.authenticatedProfile = new GameProfile(uuid, gameProfile.getName());
+            if (connection.spoofedProfile != null) {
+                for (com.mojang.authlib.properties.Property property : connection.spoofedProfile) {
+                    this.authenticatedProfile.getProperties().put(property.name(), property);
+                }
+            }
+        } else {
+            this.authenticatedProfile = gameProfile;
+        }
+        // Blueberry end
 		this.state = ServerLoginPacketListenerImpl.State.VERIFYING;
 	}
 
diff --git a/src/main/java/net/minecraft/server/players/OldUsersConverter.java b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
index 716c0ba5367106bc2b09c01336963a09f4ab4d66..f37cbd2f63a4ac2e8de8ed981e49ce8cc84d32b9 100644
--- a/src/main/java/net/minecraft/server/players/OldUsersConverter.java
+++ b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
@@ -49,8 +49,8 @@ public class OldUsersConverter {
 
 	private static void lookupPlayers(MinecraftServer minecraftServer, Collection<String> collection, ProfileLookupCallback profileLookupCallback) {
 		String[] strings = (String[])collection.stream().filter(stringx -> !StringUtil.isNullOrEmpty(stringx)).toArray(String[]::new);
-		if (minecraftServer.usesAuthentication()) {
-			minecraftServer.getProfileRepository().findProfilesByNames(strings, profileLookupCallback);
+        if (minecraftServer.usesAuthentication() || net.blueberrymc.common.bml.InternalBlueberryModConfig.Server.bungee) { // Blueberry - bungee = online mode, for now
+            minecraftServer.getProfileRepository().findProfilesByNames(strings, profileLookupCallback);
 		} else {
 			for (String string : strings) {
 				profileLookupCallback.onProfileLookupSucceeded(UUIDUtil.createOfflineProfile(string));
