From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Mon, 5 Apr 2021 17:17:17 +0900
Subject: [PATCH] Add ResourceManager#remove(PackResources)


diff --git a/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java
index 545dd6362633a1fabf73cbcc68e87bd90df23b17..e2ba61d1ecf81c74bd7be2a4cccd0daf65640a42 100644
--- a/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java
@@ -57,6 +57,8 @@ public class FallbackResourceManager implements ResourceManager {
         this.fallbacks.add(new FallbackResourceManager.PackEntry(s, packResources, predicate));
     }
 
+    @Override public boolean remove(PackResources packResources) { return this.fallbacks.remove(packResources); } // Blueberry
+
     public Set<String> getNamespaces() {
         return ImmutableSet.of(this.namespace);
     }
diff --git a/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java
index 62e8b49a734836c9ba3511b31ab5c37c93c8c568..a1e674d1e9a63e2ac4c97c15c9e40fc59f1ed3e6 100644
--- a/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java
@@ -22,7 +22,7 @@ public class MultiPackResourceManager implements CloseableResourceManager {
     private final List<PackResources> packs;
 
     public MultiPackResourceManager(PackType packType, List<PackResources> list) {
-        this.packs = List.copyOf(list);
+        this.packs = new java.util.ArrayList<>(list); // Blueberry
         Map<String, FallbackResourceManager> map = new HashMap();
         List<String> list2 = list.stream().flatMap((packResourcesx) -> packResourcesx.getNamespaces(packType).stream()).distinct().toList();
 
@@ -106,4 +106,6 @@ public class MultiPackResourceManager implements CloseableResourceManager {
     public void close() {
         this.packs.forEach(PackResources::close);
     }
+
+    @Override public boolean remove(PackResources packResources) { return this.packs.remove(packResources); } // Blueberry
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java
index 7896ee38c0a600e126f82a607d7559b92abbdbef..e9ab3be3d475b29e06a9eea6c871fa74c91c02f3 100644
--- a/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java
@@ -66,4 +66,6 @@ public class ReloadableResourceManager implements ResourceManager, AutoCloseable
     public Stream<PackResources> listPacks() {
         return this.resources.listPacks();
     }
+
+    @Override public boolean remove(PackResources packResources) {return this.resources.remove(packResources); } // Blueberry
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java
index dc2dcc0936329e5da2ae898ec6c6442cc23ad479..59fe1f014517e1e0dbd1d9fa99b8a9f794eca8cf 100644
--- a/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java
@@ -9,7 +9,7 @@ import java.util.stream.Stream;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.PackResources;
 
-public interface ResourceManager extends ResourceProvider {
+public interface ResourceManager extends ResourceProvider, net.blueberrymc.server.packs.resources.BlueberryResourceProvider { // Blueberry
     Set<String> getNamespaces();
 
     List<Resource> getResourceStack(ResourceLocation var1);
