From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Mon, 5 Apr 2021 17:17:17 +0900
Subject: [PATCH] Add ResourceManager#remove(PackResources)


diff --git a/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java
index c49797d8daeda8956e64050ecdbaa22642b4bf9b..a25fa4d33d643ad1fa1baaa2cc77095602644e19 100644
--- a/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/FallbackResourceManager.java
@@ -55,6 +55,8 @@ public class FallbackResourceManager implements ResourceManager {
         this.fallbacks.add(new FallbackResourceManager.PackEntry(s, packResources, predicate));
     }
 
+    @Override public boolean remove(PackResources packResources) { return this.fallbacks.remove(packResources); } // Blueberry
+
     public Set<String> getNamespaces() {
         return ImmutableSet.of(this.namespace);
     }
diff --git a/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java
index d0c8c9373b12db197cb556ddb16753abb8e70e15..52554f30c14fa7ea5e705a44c23f92344d6a0353 100644
--- a/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/MultiPackResourceManager.java
@@ -22,7 +22,7 @@ public class MultiPackResourceManager implements CloseableResourceManager {
     private final List<PackResources> packs;
 
     public MultiPackResourceManager(PackType packType, List<PackResources> list) {
-        this.packs = List.copyOf(list);
+        this.packs = new java.util.ArrayList<>(list); // Blueberry
         Map<String, FallbackResourceManager> map = new HashMap();
         List<String> list2 = list.stream().flatMap((packResourcesx) -> packResourcesx.getNamespaces(packType).stream()).distinct().toList();
 
@@ -114,4 +114,6 @@ public class MultiPackResourceManager implements CloseableResourceManager {
     public void close() {
         this.packs.forEach(PackResources::close);
     }
+
+    @Override public boolean remove(PackResources packResources) { return this.packs.remove(packResources); } // Blueberry
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java
index 5284e16bc26f60f35c5b718da8483c0c7c91ad8b..66d4351b28b6292990cc879c297ad2a3709d7e58 100644
--- a/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/ReloadableResourceManager.java
@@ -66,4 +66,6 @@ public class ReloadableResourceManager implements ResourceManager, AutoCloseable
     public Stream<PackResources> listPacks() {
         return this.resources.listPacks();
     }
+
+    @Override public boolean remove(PackResources packResources) {return this.resources.remove(packResources); } // Blueberry
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java b/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java
index dc2dcc0936329e5da2ae898ec6c6442cc23ad479..59fe1f014517e1e0dbd1d9fa99b8a9f794eca8cf 100644
--- a/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java
+++ b/src/main/java/net/minecraft/server/packs/resources/ResourceManager.java
@@ -9,7 +9,7 @@ import java.util.stream.Stream;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.packs.PackResources;
 
-public interface ResourceManager extends ResourceProvider {
+public interface ResourceManager extends ResourceProvider, net.blueberrymc.server.packs.resources.BlueberryResourceProvider { // Blueberry
     Set<String> getNamespaces();
 
     List<Resource> getResourceStack(ResourceLocation var1);
