From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Sun, 7 Nov 2021 21:40:40 +0900
Subject: [PATCH] Add workaround for leaves not ticking


diff --git a/src/main/java/net/minecraft/world/level/block/LeavesBlock.java b/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
index 2c053a6a280163be98242000655e9cfe756b1890..01eec1c62c998f04086696dd7126160c5d5f8aa2 100644
--- a/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
@@ -68,7 +68,11 @@ public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
 
         int i = getDistanceAt(blockState2) + 1;
         if (i != 1 || blockState.getValue(DISTANCE) != i) {
-            levelAccessor.scheduleTick(blockPos, this, 1);
+            // Blueberry start
+            //levelAccessor.scheduleTick(blockPos, this, 1);
+            // TODO: workaround for leaves not ticking
+            if (levelAccessor instanceof ServerLevel sl) tick(blockState, sl, blockPos, sl.random);
+            // Blueberry end
         }
 
         return blockState;
