From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Sun, 7 Nov 2021 21:40:40 +0900
Subject: [PATCH] Add workaround for leaves not ticking


diff --git a/src/main/java/net/minecraft/world/level/block/LeavesBlock.java b/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
index 4816ccff6413f2a5e96b09fc79aeb1c0e03a2544..4701adedbaf94fa3f4a54972fa3af9bd7627258d 100644
--- a/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LeavesBlock.java
@@ -69,7 +69,11 @@ public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
 
         int i = getDistanceAt(blockState2) + 1;
         if (i != 1 || blockState.getValue(DISTANCE) != i) {
-            levelAccessor.scheduleTick(blockPos, this, 1);
+            // Blueberry start
+            //levelAccessor.scheduleTick(blockPos, this, 1);
+            // TODO: workaround for leaves not ticking
+            if (levelAccessor instanceof ServerLevel sl) tick(blockState, sl, blockPos, sl.random);
+            // Blueberry end
         }
 
         return blockState;
