From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Wed, 10 Nov 2021 14:38:53 +0900
Subject: [PATCH] Fire network events


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index dba3ba3bd696eb6707f9f5670f42e318561cc1df..f93f1dc7c545a164d3e500da57c1326cff9f01f2 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -97,6 +97,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             LOGGER.error(LogUtils.FATAL_MARKER, "Failed to change protocol to handshake", var3);
         }
 
+        new net.blueberrymc.common.event.network.ConnectionInitEvent(this.channel, this.receiving).callEvent(); // Blueberry
     }
 
     public void setProtocol(ConnectionProtocol connectionProtocol) {
@@ -368,6 +369,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             }
         }
 
+        new net.blueberrymc.common.event.network.ConnectionSetupCompressionEvent(this.channel, this.receiving, i).callEvent(); // Blueberry
     }
 
     public void handleDisconnection() {
