From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 12 Nov 2021 14:27:56 +0900
Subject: [PATCH] Add workaround for race condition

CrudeIncrementalIntIdentityHashBiMap#byId returns null... for some reason

diff --git a/src/main/java/net/minecraft/world/level/chunk/HashMapPalette.java b/src/main/java/net/minecraft/world/level/chunk/HashMapPalette.java
index 02dbc3888b7c0f7a2d84636574c09c91c03d0bcd..7bf544cfc8ec91ab7edfc2780e8f26f5a81d62a6 100644
--- a/src/main/java/net/minecraft/world/level/chunk/HashMapPalette.java
+++ b/src/main/java/net/minecraft/world/level/chunk/HashMapPalette.java
@@ -58,6 +58,17 @@ public class HashMapPalette<T> implements Palette<T> {
     public T valueFor(int i) {
         T object = this.values.byId(i);
         if (object == null) {
+            // Blueberry start
+            // TODO: figure out why Thread.sleep(1) works; maybe race condition?
+            try {
+                Thread.sleep(1);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            object = this.values.byId(i);
+            if (object != null) return object;
+            try { return (T) net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(); } catch (ClassCastException ignore) {}
+            // Blueberry end
             throw new MissingPaletteEntryException(i);
         } else {
             return object;
