From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Fri, 26 Nov 2021 19:11:37 +0900
Subject: [PATCH] Add ability to ignore server render distance


diff --git a/src/main/java/net/minecraft/client/Options.java b/src/main/java/net/minecraft/client/Options.java
index 77b71788ce7c006d58bb52efb86aa25c13859e83..8327413ee44d20362ad2990a581e37239e03dd26 100644
--- a/src/main/java/net/minecraft/client/Options.java
+++ b/src/main/java/net/minecraft/client/Options.java
@@ -698,7 +698,8 @@ public class Options {
     }
 
     public int getEffectiveRenderDistance() {
-        return this.serverRenderDistance > 0 ? Math.min(this.renderDistance, this.serverRenderDistance) : this.renderDistance;
+        // Blueberry - ability to ignore server render distance
+        return !net.blueberrymc.common.bml.InternalBlueberryModConfig.Multiplayer.ignoreServerViewDistance && this.serverRenderDistance > 0 ? Math.min(this.renderDistance, this.serverRenderDistance) : this.renderDistance;
     }
 
     interface FieldAccess {
