From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Tue, 5 Apr 2022 15:14:42 +0900
Subject: [PATCH] Implement permissions API


diff --git a/src/main/java/net/minecraft/commands/CommandSource.java b/src/main/java/net/minecraft/commands/CommandSource.java
index da359dc9037eae090646ebb56831b0d54fec2d87..9b09393e545cae57248e6eb7a0c7041c040788f9 100644
--- a/src/main/java/net/minecraft/commands/CommandSource.java
+++ b/src/main/java/net/minecraft/commands/CommandSource.java
@@ -2,7 +2,7 @@ package net.minecraft.commands;
 
 import net.minecraft.network.chat.Component;
 
-public interface CommandSource {
+public interface CommandSource extends net.blueberrymc.common.permission.PermissionHolder { // Blueberry
     CommandSource NULL = new CommandSource() {
         public void sendSystemMessage(Component component) {
         }
diff --git a/src/main/java/net/minecraft/commands/CommandSourceStack.java b/src/main/java/net/minecraft/commands/CommandSourceStack.java
index 85f645e6aff5a44f2ad5d1c982fe08fc5fb2dcf7..b431d4e026ac87046b00243b9c0d248f6ab39797 100644
--- a/src/main/java/net/minecraft/commands/CommandSourceStack.java
+++ b/src/main/java/net/minecraft/commands/CommandSourceStack.java
@@ -40,7 +40,7 @@ import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
-public class CommandSourceStack implements SharedSuggestionProvider {
+public class CommandSourceStack implements SharedSuggestionProvider, net.blueberrymc.common.permission.PermissionHolder { // Blueberry
     public static final SimpleCommandExceptionType ERROR_NOT_PLAYER = new SimpleCommandExceptionType(Component.translatable("permissions.requires.player"));
     public static final SimpleCommandExceptionType ERROR_NOT_ENTITY = new SimpleCommandExceptionType(Component.translatable("permissions.requires.entity"));
     private final CommandSource source;
@@ -360,4 +360,16 @@ public class CommandSourceStack implements SharedSuggestionProvider {
     public FeatureFlagSet enabledFeatures() {
         return this.level.enabledFeatures();
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return source.hasPermission(permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return source.getPermissionState(permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index c2aa67412e8efdf9537d85c4fa83bc792f136219..3d888854195acea47db758e2335ada71386ae43d 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1833,4 +1833,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             };
         }
     }
-}
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForConsole(permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForConsole(permission);
+    }
+    // Blueberry end
+}
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java b/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java
index 836b4209ce8be8c0234feed4fa8a9a96086b94cd..105e0ed8d8dc66bb9d0ecd50660da61c137f8769 100644
--- a/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java
@@ -47,4 +47,16 @@ public class RconConsoleSource implements CommandSource {
     public boolean shouldInformAdmins() {
         return this.server.shouldRconBroadcast();
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForRcon(permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForRcon(permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index bfdc113a9d9e3168e48f4d25c9ab2e1a146d66cd..111287c14bec9bead205878b560ebade117c5bff 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3413,4 +3413,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             return this.save;
         }
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForEntity(this, permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForEntity(this, permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index a0feaab81526d96e02717008cdaf09390e2d13a5..2eb0341f607d4a6fee0b963f7eda269c8d758ece 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -2043,4 +2043,16 @@ public abstract class Player extends LivingEntity {
             return this.message;
         }
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForPlayer(getUUID(), permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForPlayer(getUUID(), permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/level/BaseCommandBlock.java b/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
index c4aeef7a468b4912742bb17cf812f24782d22e99..057de8df8f1e1477b72ab265a604b66b3335b5bf 100644
--- a/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
+++ b/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
@@ -209,4 +209,16 @@ public abstract class BaseCommandBlock implements CommandSource {
     }
 
     public abstract boolean isValid();
-}
\ No newline at end of file
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForCommandBlock(this, permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForCommandBlock(this, permission);
+    }
+    // Blueberry end
+}
