From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Tue, 5 Apr 2022 15:14:42 +0900
Subject: [PATCH] Implement permissions API


diff --git a/src/main/java/net/minecraft/commands/CommandSource.java b/src/main/java/net/minecraft/commands/CommandSource.java
index da359dc9037eae090646ebb56831b0d54fec2d87..9b09393e545cae57248e6eb7a0c7041c040788f9 100644
--- a/src/main/java/net/minecraft/commands/CommandSource.java
+++ b/src/main/java/net/minecraft/commands/CommandSource.java
@@ -2,7 +2,7 @@ package net.minecraft.commands;
 
 import net.minecraft.network.chat.Component;
 
-public interface CommandSource {
+public interface CommandSource extends net.blueberrymc.common.permission.PermissionHolder { // Blueberry
     CommandSource NULL = new CommandSource() {
         public void sendSystemMessage(Component component) {
         }
diff --git a/src/main/java/net/minecraft/commands/CommandSourceStack.java b/src/main/java/net/minecraft/commands/CommandSourceStack.java
index b54b175e7816b7758bcad03d58aeb38f8af52209..faf0921815dd203b76fe3646035248ad9489b400 100644
--- a/src/main/java/net/minecraft/commands/CommandSourceStack.java
+++ b/src/main/java/net/minecraft/commands/CommandSourceStack.java
@@ -38,7 +38,7 @@ import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
-public class CommandSourceStack implements SharedSuggestionProvider {
+public class CommandSourceStack implements SharedSuggestionProvider, net.blueberrymc.common.permission.PermissionHolder { // Blueberry
     public static final SimpleCommandExceptionType ERROR_NOT_PLAYER = new SimpleCommandExceptionType(Component.translatable("permissions.requires.player"));
     public static final SimpleCommandExceptionType ERROR_NOT_ENTITY = new SimpleCommandExceptionType(Component.translatable("permissions.requires.entity"));
     private final CommandSource source;
@@ -346,4 +346,16 @@ public class CommandSourceStack implements SharedSuggestionProvider {
     public FeatureFlagSet enabledFeatures() {
         return this.level.enabledFeatures();
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return source.hasPermission(permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return source.getPermissionState(permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 2b4c28c628dae74261845b43578a8d62cccf0406..70afffca9f20358f2324703ad6046099c9b68c95 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1835,4 +1835,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             };
         }
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForConsole(permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForConsole(permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java b/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java
index 836b4209ce8be8c0234feed4fa8a9a96086b94cd..105e0ed8d8dc66bb9d0ecd50660da61c137f8769 100644
--- a/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/src/main/java/net/minecraft/server/rcon/RconConsoleSource.java
@@ -47,4 +47,16 @@ public class RconConsoleSource implements CommandSource {
     public boolean shouldInformAdmins() {
         return this.server.shouldRconBroadcast();
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForRcon(permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForRcon(permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index e4f22f8f9a525590f9c37e89af73ff9cfad6d17b..4b4f6131cd695701e263d64a6c029276ed180080 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3247,4 +3247,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             return this.save;
         }
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForEntity(this, permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForEntity(this, permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 5561d50920b2e0a5a54cfbe1f40cfda121358300..f67802ab04bbece5631c37ffd29cf4bf4c834860 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -2018,4 +2018,16 @@ public abstract class Player extends LivingEntity {
             return this.message;
         }
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForPlayer(getUUID(), permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForPlayer(getUUID(), permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/level/BaseCommandBlock.java b/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
index 010b13fcef9fa6f1c6b9fcd33651e69424bb39f8..77b96d2c0dee85db4577d03b8bf936c86d2064aa 100644
--- a/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
+++ b/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
@@ -207,4 +207,16 @@ public abstract class BaseCommandBlock implements CommandSource {
     public boolean shouldInformAdmins() {
         return this.getLevel().getGameRules().getBoolean(GameRules.RULE_COMMANDBLOCKOUTPUT);
     }
+
+    // Blueberry start
+    @Override
+    public boolean hasPermission(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().hasPermissionForCommandBlock(this, permission);
+    }
+
+    @Override
+    public net.blueberrymc.common.permission.PermissionState getPermissionState(String permission) {
+        return net.blueberrymc.common.Blueberry.getPermissionProvider().getPermissionStateForCommandBlock(this, permission);
+    }
+    // Blueberry end
 }
\ No newline at end of file
