From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Tue, 15 Feb 2022 02:15:35 +0900
Subject: [PATCH] Prevent freezing builtin registries

I still haven't found a way around this

diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index 02f7c1a286a7b04be87bfca8bd7f2dbd186b0879..2a06964f5bdf762d5e8917a2c5a4d4679e59b443 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -52,7 +52,7 @@ public class Bootstrap {
                     EntitySelectorOptions.bootStrap();
                     DispenseItemBehavior.bootStrap();
                     CauldronInteraction.bootStrap();
-                    Registry.freezeBuiltins();
+                    //Registry.freezeBuiltins(); // Blueberry - don't freeze registries
                     wrapStreams();
                 }
             }
