From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: acrylic-style <me@acrylicstyle.xyz>
Date: Tue, 9 Aug 2022 17:11:40 +0900
Subject: [PATCH] Bring vanilla pack on top


diff --git a/src/main/java/net/minecraft/client/resources/ClientPackSource.java b/src/main/java/net/minecraft/client/resources/ClientPackSource.java
index f37af77e9f0752eb2dcfa9757931dc7d09668760..e714e59b8ad30b3918e950810bdba54406ed9e1a 100644
--- a/src/main/java/net/minecraft/client/resources/ClientPackSource.java
+++ b/src/main/java/net/minecraft/client/resources/ClientPackSource.java
@@ -77,7 +77,7 @@ public class ClientPackSource implements RepositorySource {
     }
 
     public void loadPacks(Consumer<Pack> consumer, Pack.PackConstructor packConstructor) {
-        Pack pack = Pack.create("vanilla", true, () -> this.vanillaPack, packConstructor, Pack.Position.BOTTOM, PackSource.BUILT_IN);
+        Pack pack = Pack.create("vanilla", true, () -> this.vanillaPack, packConstructor, Pack.Position.TOP, PackSource.BUILT_IN);
         if (pack != null) {
             consumer.accept(pack);
         }
